{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "// Immediately-invoked-function-expression to avoid global variables.\n",
       "(function() {\n",
       "    var warning_div = document.getElementById(\"webio-warning-13170384587329148168\");\n",
       "    var hide = function () {\n",
       "        var script = document.getElementById(\"webio-setup-12459878090760114824\");\n",
       "        var parent = script && script.parentElement;\n",
       "        var grandparent = parent && parent.parentElement;\n",
       "        if (grandparent) {\n",
       "            grandparent.style.display = \"none\";\n",
       "        }\n",
       "        warning_div.style.display = \"none\";\n",
       "    };\n",
       "    if (typeof Jupyter !== \"undefined\") {\n",
       "        console.log(\"WebIO detected Jupyter notebook environment.\");\n",
       "        // Jupyter notebook.\n",
       "        var extensions = (\n",
       "            Jupyter\n",
       "            && Jupyter.notebook.config.data\n",
       "            && Jupyter.notebook.config.data.load_extensions\n",
       "        );\n",
       "        if (extensions && extensions[\"webio-jupyter-notebook\"]) {\n",
       "            // Extension already loaded.\n",
       "            console.log(\"Jupyter WebIO nbextension detected; not loading ad-hoc.\");\n",
       "            hide();\n",
       "            return;\n",
       "        }\n",
       "    } else if (window.location.pathname.includes(\"/lab\")) {\n",
       "        // Guessing JupyterLa\n",
       "        console.log(\"Jupyter Lab detected; make sure the @webio/jupyter-lab-provider labextension is installed.\");\n",
       "        hide();\n",
       "        return;\n",
       "    }\n",
       "})();\n",
       "\n",
       "</script>\n",
       "<p\n",
       "    id=\"webio-warning-13170384587329148168\"\n",
       "    class=\"output_text output_stderr\"\n",
       "    style=\"padding: 1em; font-weight: bold;\"\n",
       ">\n",
       "    Unable to load WebIO. Please make sure WebIO works for your Jupyter client.\n",
       "    For troubleshooting, please see <a href=\"https://juliagizmos.github.io/WebIO.jl/latest/providers/ijulia/\">\n",
       "    the WebIO/IJulia documentation</a>.\n",
       "    <!-- TODO: link to installation docs. -->\n",
       "</p>\n"
      ],
      "text/plain": [
       "HTML{String}(\"<script>\\n// Immediately-invoked-function-expression to avoid global variables.\\n(function() {\\n    var warning_div = document.getElementById(\\\"webio-warning-13170384587329148168\\\");\\n    var hide = function () {\\n        var script = document.getElementById(\\\"webio-setup-12459878090760114824\\\");\\n        var parent = script && script.parentElement;\\n        var grandparent = parent && parent.parentElement;\\n        if (grandparent) {\\n            grandparent.style.display = \\\"none\\\";\\n        }\\n        warning_div.style.display = \\\"none\\\";\\n    };\\n    if (typeof Jupyter !== \\\"undefined\\\") {\\n        console.log(\\\"WebIO detected Jupyter notebook environment.\\\");\\n        // Jupyter notebook.\\n        var extensions = (\\n            Jupyter\\n            && Jupyter.notebook.config.data\\n            && Jupyter.notebook.config.data.load_extensions\\n        );\\n        if (extensions && extensions[\\\"webio-jupyter-notebook\\\"]) {\\n            // Extension already loaded.\\n            console.log(\\\"Jupyter WebIO nbextension detected; not loading ad-hoc.\\\");\\n            hide();\\n            return;\\n        }\\n    } else if (window.location.pathname.includes(\\\"/lab\\\")) {\\n        // Guessing JupyterLa\\n        console.log(\\\"Jupyter Lab detected; make sure the @webio/jupyter-lab-provider labextension is installed.\\\");\\n        hide();\\n        return;\\n    }\\n})();\\n\\n</script>\\n<p\\n    id=\\\"webio-warning-13170384587329148168\\\"\\n    class=\\\"output_text output_stderr\\\"\\n    style=\\\"padding: 1em; font-weight: bold;\\\"\\n>\\n    Unable to load WebIO. Please make sure WebIO works for your Jupyter client.\\n    For troubleshooting, please see <a href=\\\"https://juliagizmos.github.io/WebIO.jl/latest/providers/ijulia/\\\">\\n    the WebIO/IJulia documentation</a>.\\n    <!-- TODO: link to installation docs. -->\\n</p>\\n\")"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "┌ Warning: Error requiring WebSockets from WebIO:\n",
      "│ LoadError: IOError: open: operation not permitted (EPERM)\n",
      "│ Stacktrace:\n",
      "│  [1] uv_error at .\\libuv.jl:85 [inlined]\n",
      "│  [2] open(::String, ::UInt16, ::UInt16) at .\\filesystem.jl:81\n",
      "│  [3] tryopen_exclusive(::String, ::UInt16) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:152\n",
      "│  [4] #open_exclusive#4(::UInt16, ::Int64, ::Int64, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:185\n",
      "│  [5] #open_exclusive at .\\none:0 [inlined]\n",
      "│  [6] (::getfield(Pidfile, Symbol(\"##mkpidlock#1#2\")))(::Base.Iterators.Pairs{Symbol,Int64,Tuple{Symbol},NamedTuple{(:stale_age,),Tuple{Int64}}}, ::Function, ::String, ::Int32) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:42\n",
      "│  [7] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [8] #mkpidlock#3 at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:56 [inlined]\n",
      "│  [9] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [10] withlock at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:10 [inlined]\n",
      "│  [11] #register#1(::String, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:49\n",
      "│  [12] register(::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:29\n",
      "│  [13] top-level scope at none:0\n",
      "│  [14] include at .\\boot.jl:317 [inlined]\n",
      "│  [15] include_relative(::Module, ::String) at .\\loading.jl:1044\n",
      "│  [16] include at .\\sysimg.jl:29 [inlined]\n",
      "│  [17] include(::String) at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1\n",
      "│  [18] top-level scope at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:111\n",
      "│  [19] eval at .\\boot.jl:319 [inlined]\n",
      "│  [20] eval at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1 [inlined]\n",
      "│  [21] (::getfield(WebIO, Symbol(\"##79#91\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:67\n",
      "│  [22] err(::getfield(WebIO, Symbol(\"##79#91\")), ::Module, ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:38\n",
      "│  [23] #78 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:66 [inlined]\n",
      "│  [24] withpath(::getfield(WebIO, Symbol(\"##78#90\")), ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:28\n",
      "│  [25] (::getfield(WebIO, Symbol(\"##77#89\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:65\n",
      "│  [26] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [27] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [28] #3 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19 [inlined]\n",
      "│  [29] iterate at .\\generator.jl:47 [inlined]\n",
      "│  [30] _collect(::Array{Function,1}, ::Base.Generator{Array{Function,1},getfield(Requires, Symbol(\"##3#4\"))}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at .\\array.jl:632\n",
      "│  [31] map at .\\array.jl:561 [inlined]\n",
      "│  [32] loadpkg(::Base.PkgId) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19\n",
      "│  [33] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [34] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [35] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:651\n",
      "│  [36] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [37] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [38] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [39] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [40] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [41] _require(::Base.PkgId) at .\\loading.jl:937\n",
      "│  [42] require(::Base.PkgId) at .\\loading.jl:858\n",
      "│  [43] (::getfield(Distributed, Symbol(\"##2#4\")){Base.PkgId})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\Distributed.jl:77\n",
      "│  [44] (::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:276\n",
      "│  [45] run_work_thunk(::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}, ::Bool) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:56\n",
      "│  [46] run_work_thunk(::Distributed.RemoteValue, ::Function) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:65\n",
      "│  [47] (::getfield(Distributed, Symbol(\"##102#104\")){Distributed.RemoteValue,getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}})() at .\\task.jl:259\n",
      "│ in expression starting at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\providers\\generic_http.jl:21\n",
      "└ @ Requires C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:40\n",
      "┌ Warning: Error requiring WebSockets from WebIO:\n",
      "│ LoadError: IOError: open: operation not permitted (EPERM)\n",
      "│ Stacktrace:\n",
      "│  [1] uv_error at .\\libuv.jl:85 [inlined]\n",
      "│  [2] open(::String, ::UInt16, ::UInt16) at .\\filesystem.jl:81\n",
      "│  [3] tryopen_exclusive(::String, ::UInt16) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:152\n",
      "│  [4] #open_exclusive#4(::UInt16, ::Int64, ::Int64, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:185\n",
      "│  [5] #open_exclusive at .\\none:0 [inlined]\n",
      "│  [6] (::getfield(Pidfile, Symbol(\"##mkpidlock#1#2\")))(::Base.Iterators.Pairs{Symbol,Int64,Tuple{Symbol},NamedTuple{(:stale_age,),Tuple{Int64}}}, ::Function, ::String, ::Int32) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:42\n",
      "│  [7] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [8] #mkpidlock#3 at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:56 [inlined]\n",
      "│  [9] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [10] withlock at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:10 [inlined]\n",
      "│  [11] #register#1(::String, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:49\n",
      "│  [12] register(::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:29\n",
      "│  [13] top-level scope at none:0\n",
      "│  [14] include at .\\boot.jl:317 [inlined]\n",
      "│  [15] include_relative(::Module, ::String) at .\\loading.jl:1044\n",
      "│  [16] include at .\\sysimg.jl:29 [inlined]\n",
      "│  [17] include(::String) at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1\n",
      "│  [18] top-level scope at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:111\n",
      "│  [19] eval at .\\boot.jl:319 [inlined]\n",
      "│  [20] eval at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1 [inlined]\n",
      "│  [21] (::getfield(WebIO, Symbol(\"##79#91\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:67\n",
      "│  [22] err(::getfield(WebIO, Symbol(\"##79#91\")), ::Module, ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:38\n",
      "│  [23] #78 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:66 [inlined]\n",
      "│  [24] withpath(::getfield(WebIO, Symbol(\"##78#90\")), ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:28\n",
      "│  [25] (::getfield(WebIO, Symbol(\"##77#89\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:65\n",
      "│  [26] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [27] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [28] #3 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19 [inlined]\n",
      "│  [29] iterate at .\\generator.jl:47 [inlined]\n",
      "│  [30] _collect(::Array{Function,1}, ::Base.Generator{Array{Function,1},getfield(Requires, Symbol(\"##3#4\"))}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at .\\array.jl:632\n",
      "│  [31] map at .\\array.jl:561 [inlined]\n",
      "│  [32] loadpkg(::Base.PkgId) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19\n",
      "│  [33] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [34] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [35] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:651\n",
      "│  [36] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [37] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [38] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [39] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [40] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [41] _require(::Base.PkgId) at .\\loading.jl:937\n",
      "│  [42] require(::Base.PkgId) at .\\loading.jl:858\n",
      "│  [43] (::getfield(Distributed, Symbol(\"##2#4\")){Base.PkgId})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\Distributed.jl:77\n",
      "│  [44] (::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:276\n",
      "│  [45] run_work_thunk(::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}, ::Bool) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:56\n",
      "│  [46] run_work_thunk(::Distributed.RemoteValue, ::Function) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:65\n",
      "│  [47] (::getfield(Distributed, Symbol(\"##102#104\")){Distributed.RemoteValue,getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}})() at .\\task.jl:259\n",
      "│ in expression starting at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\providers\\generic_http.jl:21\n",
      "└ @ Requires C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:40\n",
      "┌ Warning: Error requiring WebSockets from WebIO:\n",
      "│ LoadError: IOError: open: operation not permitted (EPERM)\n",
      "│ Stacktrace:\n",
      "│  [1] uv_error at .\\libuv.jl:85 [inlined]\n",
      "│  [2] open(::String, ::UInt16, ::UInt16) at .\\filesystem.jl:81\n",
      "│  [3] tryopen_exclusive(::String, ::UInt16) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:152\n",
      "│  [4] #open_exclusive#4(::UInt16, ::Int64, ::Int64, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:185\n",
      "│  [5] #open_exclusive at .\\none:0 [inlined]\n",
      "│  [6] (::getfield(Pidfile, Symbol(\"##mkpidlock#1#2\")))(::Base.Iterators.Pairs{Symbol,Int64,Tuple{Symbol},NamedTuple{(:stale_age,),Tuple{Int64}}}, ::Function, ::String, ::Int32) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:42\n",
      "│  [7] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [8] #mkpidlock#3 at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:56 [inlined]\n",
      "│  [9] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [10] withlock at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:10 [inlined]\n",
      "│  [11] #register#1(::String, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:49\n",
      "│  [12] register(::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:29\n",
      "│  [13] top-level scope at none:0\n",
      "│  [14] include at .\\boot.jl:317 [inlined]\n",
      "│  [15] include_relative(::Module, ::String) at .\\loading.jl:1044\n",
      "│  [16] include at .\\sysimg.jl:29 [inlined]\n",
      "│  [17] include(::String) at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1\n",
      "│  [18] top-level scope at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:111\n",
      "│  [19] eval at .\\boot.jl:319 [inlined]\n",
      "│  [20] eval at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1 [inlined]\n",
      "│  [21] (::getfield(WebIO, Symbol(\"##79#91\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:67\n",
      "│  [22] err(::getfield(WebIO, Symbol(\"##79#91\")), ::Module, ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:38\n",
      "│  [23] #78 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:66 [inlined]\n",
      "│  [24] withpath(::getfield(WebIO, Symbol(\"##78#90\")), ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:28\n",
      "│  [25] (::getfield(WebIO, Symbol(\"##77#89\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:65\n",
      "│  [26] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [27] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [28] #3 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19 [inlined]\n",
      "│  [29] iterate at .\\generator.jl:47 [inlined]\n",
      "│  [30] _collect(::Array{Function,1}, ::Base.Generator{Array{Function,1},getfield(Requires, Symbol(\"##3#4\"))}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at .\\array.jl:632\n",
      "│  [31] map at .\\array.jl:561 [inlined]\n",
      "│  [32] loadpkg(::Base.PkgId) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19\n",
      "│  [33] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [34] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [35] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:651\n",
      "│  [36] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [37] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [38] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [39] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [40] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [41] _require(::Base.PkgId) at .\\loading.jl:937\n",
      "│  [42] require(::Base.PkgId) at .\\loading.jl:858\n",
      "│  [43] (::getfield(Distributed, Symbol(\"##2#4\")){Base.PkgId})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\Distributed.jl:77\n",
      "│  [44] (::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:276\n",
      "│  [45] run_work_thunk(::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}, ::Bool) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:56\n",
      "│  [46] run_work_thunk(::Distributed.RemoteValue, ::Function) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:65\n",
      "│  [47] (::getfield(Distributed, Symbol(\"##102#104\")){Distributed.RemoteValue,getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}})() at .\\task.jl:259\n",
      "│ in expression starting at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\providers\\generic_http.jl:21\n",
      "└ @ Requires C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:40\n",
      "┌ Warning: Error requiring WebSockets from WebIO:\n",
      "│ LoadError: IOError: open: operation not permitted (EPERM)\n",
      "│ Stacktrace:\n",
      "│  [1] uv_error at .\\libuv.jl:85 [inlined]\n",
      "│  [2] open(::String, ::UInt16, ::UInt16) at .\\filesystem.jl:81\n",
      "│  [3] tryopen_exclusive(::String, ::UInt16) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:152\n",
      "│  [4] #open_exclusive#4(::UInt16, ::Int64, ::Int64, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:185\n",
      "│  [5] #open_exclusive at .\\none:0 [inlined]\n",
      "│  [6] (::getfield(Pidfile, Symbol(\"##mkpidlock#1#2\")))(::Base.Iterators.Pairs{Symbol,Int64,Tuple{Symbol},NamedTuple{(:stale_age,),Tuple{Int64}}}, ::Function, ::String, ::Int32) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:42\n",
      "│  [7] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [8] #mkpidlock#3 at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:56 [inlined]\n",
      "│  [9] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [10] withlock at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:10 [inlined]\n",
      "│  [11] #register#1(::String, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:49\n",
      "│  [12] register(::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:29\n",
      "│  [13] top-level scope at none:0\n",
      "│  [14] include at .\\boot.jl:317 [inlined]\n",
      "│  [15] include_relative(::Module, ::String) at .\\loading.jl:1044\n",
      "│  [16] include at .\\sysimg.jl:29 [inlined]\n",
      "│  [17] include(::String) at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1\n",
      "│  [18] top-level scope at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:111\n",
      "│  [19] eval at .\\boot.jl:319 [inlined]\n",
      "│  [20] eval at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1 [inlined]\n",
      "│  [21] (::getfield(WebIO, Symbol(\"##79#91\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:67\n",
      "│  [22] err(::getfield(WebIO, Symbol(\"##79#91\")), ::Module, ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:38\n",
      "│  [23] #78 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:66 [inlined]\n",
      "│  [24] withpath(::getfield(WebIO, Symbol(\"##78#90\")), ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:28\n",
      "│  [25] (::getfield(WebIO, Symbol(\"##77#89\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:65\n",
      "│  [26] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [27] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [28] #3 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19 [inlined]\n",
      "│  [29] iterate at .\\generator.jl:47 [inlined]\n",
      "│  [30] _collect(::Array{Function,1}, ::Base.Generator{Array{Function,1},getfield(Requires, Symbol(\"##3#4\"))}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at .\\array.jl:632\n",
      "│  [31] map at .\\array.jl:561 [inlined]\n",
      "│  [32] loadpkg(::Base.PkgId) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19\n",
      "│  [33] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [34] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [35] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:651\n",
      "│  [36] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [37] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [38] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [39] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [40] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [41] _require(::Base.PkgId) at .\\loading.jl:937\n",
      "│  [42] require(::Base.PkgId) at .\\loading.jl:858\n",
      "│  [43] (::getfield(Distributed, Symbol(\"##2#4\")){Base.PkgId})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\Distributed.jl:77\n",
      "│  [44] (::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:276\n",
      "│  [45] run_work_thunk(::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}, ::Bool) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:56\n",
      "│  [46] run_work_thunk(::Distributed.RemoteValue, ::Function) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:65\n",
      "│  [47] (::getfield(Distributed, Symbol(\"##102#104\")){Distributed.RemoteValue,getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}})() at .\\task.jl:259\n",
      "│ in expression starting at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\providers\\generic_http.jl:21\n",
      "└ @ Requires C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:40\n",
      "┌ Info: Recompiling stale cache file C:\\Users\\alech\\.julia\\compiled\\v1.0\\cElegansFunctions.ji for cElegansFunctions [top-level]\n",
      "└ @ Base loading.jl:1190\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "R version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "Copyright (C) 2019 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Method definition computeBarcodesNullModels(Any, Any) in module cElegansFunctions at E:functions\\cElegansFunctions.jl:173 overwritten at E:functions\\cElegansFunctions.jl:377.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "R version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "Copyright (C) 2019 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "      From worker 4:\t\n",
      "      From worker 5:\t\n",
      "      From worker 5:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 5:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 5:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 5:\t\n",
      "      From worker 5:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 5:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 5:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 5:\t\n",
      "      From worker 5:\tR is a collaborative project with many contributors.\n",
      "      From worker 5:\tType 'contributors()' for more information and\n",
      "      From worker 5:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 5:\t\n",
      "      From worker 5:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 5:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 5:\tType 'q()' to quit R.\n",
      "      From worker 5:\t\n",
      "      From worker 6:\t\n",
      "      From worker 6:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 6:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 6:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 6:\t\n",
      "      From worker 6:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 6:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 6:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 6:\t\n",
      "      From worker 6:\tR is a collaborative project with many contributors.\n",
      "      From worker 6:\tType 'contributors()' for more information and\n",
      "      From worker 6:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 6:\t\n",
      "      From worker 2:\t\n",
      "      From worker 2:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 2:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 2:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 2:\t\n",
      "      From worker 2:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 2:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 2:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 2:\t\n",
      "      From worker 2:\tR is a collaborative project with many contributors.\n",
      "      From worker 2:\tType 'contributors()' for more information and\n",
      "      From worker 2:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 2:\t\n",
      "      From worker 2:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 2:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 2:\tType 'q()' to quit R.\n",
      "      From worker 2:\t\n",
      "      From worker 4:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 4:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 4:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 4:\t\n",
      "      From worker 4:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 4:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 4:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 4:\t\n",
      "      From worker 4:\tR is a collaborative project with many contributors.\n",
      "      From worker 4:\tType 'contributors()' for more information and\n",
      "      From worker 4:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 4:\t\n",
      "      From worker 4:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 4:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 4:\tType 'q()' to quit R.\n",
      "      From worker 4:\t\n",
      "      From worker 7:\t\n",
      "      From worker 7:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 7:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 7:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 7:\t\n",
      "      From worker 7:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 7:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 7:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 7:\t\n",
      "      From worker 7:\tR is a collaborative project with many contributors.\n",
      "      From worker 7:\tType 'contributors()' for more information and\n",
      "      From worker 7:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 7:\t\n",
      "      From worker 7:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 7:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 7:\tType 'q()' to quit R.\n",
      "      From worker 7:\t\n",
      "      From worker 3:\t\n",
      "      From worker 3:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 3:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 3:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 3:\t\n",
      "      From worker 3:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 3:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 3:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 3:\t\n",
      "      From worker 3:\tR is a collaborative project with many contributors.\n",
      "      From worker 3:\tType 'contributors()' for more information and\n",
      "      From worker 3:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 3:\t\n",
      "      From worker 3:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 3:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 3:\tType 'q()' to quit R.\n",
      "      From worker 3:\t\n",
      "      From worker 8:\t\n",
      "      From worker 8:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 8:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 8:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 8:\t\n",
      "      From worker 8:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 8:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 8:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 8:\t\n",
      "      From worker 8:\tR is a collaborative project with many contributors.\n",
      "      From worker 8:\tType 'contributors()' for more information and\n",
      "      From worker 8:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 8:\t\n",
      "      From worker 8:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 8:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 8:\tType 'q()' to quit R.\n",
      "      From worker 8:\t\n",
      "      From worker 6:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 6:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 6:\tType 'q()' to quit R.\n",
      "      From worker 6:\t\n"
     ]
    }
   ],
   "source": [
    "#load neccesary packages\n",
    "using CSV\n",
    "using DataFrames\n",
    "using Plots\n",
    "using Statistics\n",
    "using StatsPlots\n",
    "using RDatasets\n",
    "using Combinatorics\n",
    "using HypothesisTests\n",
    "using StatsBase\n",
    "using VegaLite\n",
    "using Distances\n",
    "using Distributed\n",
    "addprocs(7)\n",
    "using Eirene   \n",
    "@everywhere using SharedArrays\n",
    "\n",
    "\n",
    "@everywhere push!(LOAD_PATH, $\"E:functions/\") \n",
    "@everywhere using cElegansFunctions\n",
    "@everywhere using RCall\n",
    "\n",
    "\n",
    "@everywhere R\"library(TDA)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "#load neccesary data files\n",
    "positions = CSV.read(\"E://From Gene Expression File//279positions.csv\", copycols = true, header = false)\n",
    "positionsLabels = CSV.read(\"E://From Gene Expression File//279positionsLabels.csv\", copycols = true, header = false)\n",
    "birthTimes = CSV.read(\"E://From Gene Expression File//279birthTimes.csv\", copycols = true, header = false)\n",
    "birthTimesLabels = CSV.read(\"E://From Gene Expression File//279birthTimesLabels.csv\", copycols = true, header = false)\n",
    "connectionList = CSV.read(\"E://From Gene Expression File//NeuronConnect_corrected.csv\", copycols = true, header = false)\n",
    "labeledTypeList = CSV.read(\"E://From Gene Expression File//279labeledTypes.csv\", copycols = true, header = false)\n",
    "zeroDimTopSim = CSV.read(\"E://From Gene Expression File//dimensionZeroSimilarity.csv\", copycols = true, header = false)\n",
    "oneDimTopSim = CSV.read(\"E://From Gene Expression File//dimensionOneSimilarity.csv\", copycols = true, header = false)\n",
    "twoDimTopSim = CSV.read(\"E://From Gene Expression File//dimensionTwoSimilarity.csv\", copycols = true, header = false)\n",
    "zeroDimTopSimGD = CSV.read(\"E://From Gene Expression File//dimensionZeroMixedPairSimilarity.csv\", copycols = true, header = false)\n",
    "oneDimTopSimGD = CSV.read(\"E://From Gene Expression File//dimensionOneMixedPairSimilarity.csv\", copycols = true, header = false)\n",
    "twoDimTopSimGD = CSV.read(\"E://From Gene Expression File//dimensionTwoMixedPairSimilarity.csv\", copycols = true, header = false)\n",
    "\n",
    "\n",
    "#done\n",
    "println(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[3]:7\n",
      "└ @ Core .\\In[3]:7\n",
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[3]:16\n",
      "└ @ Core .\\In[3]:16\n",
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[3]:17\n",
      "└ @ Core .\\In[3]:17\n",
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[3]:24\n",
      "└ @ Core .\\In[3]:24\n",
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[3]:25\n",
      "└ @ Core .\\In[3]:25\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: labelsTimesPositionsType, birthSortedAdjacencyMatrix, birthOrder, numValues, and times\n"
     ]
    }
   ],
   "source": [
    "#process all the data into a single array sorted by birth time\n",
    "\n",
    "#creates an array for positions with their labels\n",
    "positionsMatrix = Array{Any, 2}(UndefInitializer(),279,3)\n",
    "\n",
    "for i in 1:279\n",
    "    positionsMatrix[i,1] = positionsLabels[1][i]\n",
    "    positionsMatrix[i,2] = positions[i,1]\n",
    "    positionsMatrix[i,3] = positions[i,2]\n",
    "end\n",
    "\n",
    "#creates an array for positions with their labels\n",
    "birthTimesMatrix = Array{Any, 2}(UndefInitializer(),279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    birthTimesMatrix[i,1] = birthTimesLabels[1][i]\n",
    "    birthTimesMatrix[i,2] = birthTimes[1][i]\n",
    "end\n",
    "\n",
    "#creates an array for types with their labels\n",
    "typeMatrix = Array{Any, 2}(UndefInitializer(), 279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    typeMatrix[i,1] = labeledTypeList[1][i]\n",
    "    typeMatrix[i,2] = labeledTypeList[2][i]\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#sorts all the above matricies by neuron label to make them compatible\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && positionsMatrix[i - movement, 1] < positionsMatrix[i-movement - 1, 1]\n",
    "        shadowCache = positionsMatrix[i - movement, :]\n",
    "        positionsMatrix[i - movement, :] = positionsMatrix[i - movement - 1, :]\n",
    "        positionsMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "    \n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && birthTimesMatrix[i - movement, 1] < birthTimesMatrix[i-movement - 1, 1]\n",
    "        shadowCache = birthTimesMatrix[i - movement, :]\n",
    "        birthTimesMatrix[i - movement, :] = birthTimesMatrix[i - movement - 1, :]\n",
    "        birthTimesMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "    \n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && typeMatrix[i - movement, 1] < typeMatrix[i-movement - 1, 1]\n",
    "        shadowCache = typeMatrix[i - movement, :]\n",
    "        typeMatrix[i - movement, :] = typeMatrix[i - movement - 1, :]\n",
    "        typeMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#stores all these data points into a single concise matrix\n",
    "labelsTimesPositionsType = Array{Any, 2}(UndefInitializer(),279,5)\n",
    "\n",
    "labelsTimesPositionsType[:,1:2] = birthTimesMatrix[:,:]\n",
    "labelsTimesPositionsType[:,3:4] = positionsMatrix[:,2:3]\n",
    "labelsTimesPositionsType[:,5] = typeMatrix[:,2]\n",
    "\n",
    "#alter the positions isometrically to look nicer \n",
    "labelsTimesPositionsType[:,3] *= -1\n",
    "labelsTimesPositionsType[:,4] *= -1\n",
    "\n",
    "\n",
    "#sorts new matrix by birth time\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && labelsTimesPositionsType[i - movement, 2] < labelsTimesPositionsType[i-movement - 1, 2]\n",
    "        shadowCache = labelsTimesPositionsType[i - movement, :]\n",
    "        labelsTimesPositionsType[i - movement, :] = labelsTimesPositionsType[i - movement - 1, :]\n",
    "        labelsTimesPositionsType[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#processes connection list and stores that in its own matrix\n",
    "\n",
    "connectionBase = Array{Any, 2}(UndefInitializer(),6417,4)\n",
    "\n",
    "for i in 1:6417\n",
    "    for j in 1:4\n",
    "        connectionBase[i,j] = connectionList[i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#create adjacency matrix sorted by birth time weighted for a node filtered persistent homology calculation\n",
    "birthSortedAdjacencyMatrix = zeros(279,279)\n",
    "       \n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        birthSortedAdjacencyMatrix[i,j] = 3000\n",
    "    end\n",
    "end\n",
    "\n",
    "for k in 1:6417\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if connectionBase[k,1] == labelsTimesPositionsType[i,1] && connectionBase[k,2] == labelsTimesPositionsType[j,1]\n",
    "                max = maximum([labelsTimesPositionsType[i,2],labelsTimesPositionsType[j,2]])\n",
    "                birthSortedAdjacencyMatrix[i,j] = max\n",
    "            end\n",
    "            if connectionBase[k,1] == labelsTimesPositionsType[j,1] && connectionBase[k,2] == labelsTimesPositionsType[i,1]\n",
    "                max = maximum([labelsTimesPositionsType[i,2],labelsTimesPositionsType[j,2]])\n",
    "                birthSortedAdjacencyMatrix[i,j] = max\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "#add all the mostly trivial self loops, which are neccesary for dimension 0 calculations\n",
    "for i in 1:279\n",
    "    birthSortedAdjacencyMatrix[i,i] = labelsTimesPositionsType[i,2]\n",
    "end       \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#construct an array that is a list of times at which neurons are born, and how many born at that time. also constructs a list providing the number on this list for each neuron\n",
    "birthOrder = Array{Int64, 1}(undef, 279)\n",
    "times = Array{Float64, 2}(undef, 2,2)\n",
    "times[1,:] = [labelsTimesPositionsType[1,2],1]\n",
    "birthOrder[1] = 1\n",
    "times[2,:] = [labelsTimesPositionsType[2,2],1]\n",
    "birthOrder[2] = 2\n",
    "birthOrderCounter = 2\n",
    "\n",
    "for i in 3:279\n",
    "    if labelsTimesPositionsType[i,2] != times[end,1]\n",
    "        times = cat(dims = 1, times[:,:], transpose([labelsTimesPositionsType[i,2],1]))\n",
    "        birthOrderCounter += 1\n",
    "    else\n",
    "        times[end,2] += 1\n",
    "    end\n",
    "    birthOrder[i] = birthOrderCounter\n",
    "end\n",
    "\n",
    "\n",
    "#determine number of points on x-axis\n",
    "numValues = size(times,1)\n",
    "\n",
    "#end step\n",
    "println(\"output: labelsTimesPositionsType, birthSortedAdjacencyMatrix, birthOrder, numValues, and times\")\n",
    "\n",
    "#cleanup\n",
    "positionsMatrix = nothing\n",
    "positions = nothing\n",
    "positionsLabels = nothing\n",
    "birthTimes = nothing\n",
    "birthTimesLabels = nothing\n",
    "connectionList = nothing\n",
    "labeledTypeList = nothing\n",
    "birthTimesMatrix = nothing\n",
    "typeMatrix = nothing\n",
    "connectionBase = nothing\n",
    "birthOrderCounter = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: numValues, BettiCurves\n"
     ]
    }
   ],
   "source": [
    "#computes the betti curves in dimensions 0-3, and birth/death rates for each also\n",
    "eireneMatrix = eirene(birthSortedAdjacencyMatrix, model = \"vr\", minrad = 0.5, maxrad = 3001, maxdim = 3)\n",
    " \n",
    "\n",
    "\n",
    "\n",
    "#make arrays to store betti curves\n",
    "BettiCurves = zeros(numValues,4)\n",
    "BettiRates = zeros(numValues,3,2)\n",
    "barcodes = barcode(eireneMatrix, dim=0)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "\n",
    "#calculate and store dimension 0 betti values\n",
    "for i in 1:numValues\n",
    "    \n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "            BettiRates[i,1,1] += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "            BettiRates[i,1,2] += 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,1] = bettiNumber\n",
    "end\n",
    "\n",
    "\n",
    "#calculate and store dimension 1 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=1)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "            BettiRates[i,2,1] += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "            BettiRates[i,2,2] += 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,2] = bettiNumber\n",
    "end\n",
    "\n",
    "#calculate and store dimension 2 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=2)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "            BettiRates[i,3,1] += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "            BettiRates[i,3,2] += 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,3] = bettiNumber\n",
    "end\n",
    "\n",
    "\n",
    "#calculate and store dimension 3 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=3)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "for i in 1:numValues\n",
    "        \n",
    "\n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,4] = bettiNumber\n",
    "end\n",
    "\n",
    "#end step\n",
    "println(\"output: numValues, BettiCurves\")\n",
    "\n",
    "#cleanup\n",
    "barcodes = nothing\n",
    "bettiNumber = nothing\n",
    "amount = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "UndefVarError",
     "evalue": "UndefVarError: location not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: location not defined",
      "",
      "Stacktrace:",
      " [1] top-level scope at In[6]:40"
     ]
    }
   ],
   "source": [
    "#plot of betti curves\n",
    "\n",
    "\n",
    "#determine percent of neurons born with time\n",
    "percentBorn = zeros(numValues)\n",
    "percentCounter = 1\n",
    "percentBorn[percentCounter] = 1/2.79\n",
    "for i in 2:279\n",
    "    if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[i-1,2]\n",
    "        percentCounter += 1\n",
    "        percentBorn[percentCounter] += percentBorn[percentCounter-1]\n",
    "    end\n",
    "    percentBorn[percentCounter] += 1/2.79  \n",
    "end\n",
    "\n",
    "#determine edge density with time\n",
    "edgeDensity = zeros(numValues,2)\n",
    "edgeCounter = 0\n",
    "\n",
    "for t in 1:numValues\n",
    "    edgeCounter += trunc(Int, times[t,2])\n",
    "    \n",
    "    numberEdges = 0\n",
    "\n",
    "    for i in 1:edgeCounter\n",
    "        for j in 1:i-1\n",
    "            if birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "                numberEdges += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    edgeDensity[t,2] = numberEdges/binomial(edgeCounter,2)\n",
    "    edgeDensity[t,1] = times[t,1]\n",
    "end\n",
    "\n",
    "\n",
    "#do the actual plotting\n",
    "p1 = plot(times[:,1], [BettiCurves[:,2], BettiCurves[:,3]], color = [:red :blue theme(:default)], title = \"C Elegans Betti Curves\", label = [\"dim 1\" \"dim 2\" \"dim 3\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"betti number\", xlims=(0,2400))\n",
    "sticks!(1:1300, location[:], lc = :black, label = \"period of no neuron birth\")\n",
    "p2 = plot(edgeDensity[:,1], edgeDensity[:,2], lc = [:grey], line=(:dot,2), legendfont = font(4), xlabel = \"minutes\", ylabel = \"edge density\", legend = false, label = \"edge density\",  xlims=(0,2400))\n",
    "p3 = plot(times[:,1], percentBorn[:], lc = [:black], line=(:dot,2),  ylabel = \"percent of neurons born\", xlims=(0,2400), legend = false)\n",
    "p4 = plot(times[:,1], [BettiCurves[:,1], BettiCurves[:,4]], color = [:green :purple theme(:default)], title = \"C Elegans Betti Curves\", label = [\"dim 0\" \"dim 3\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"betti number\", ylims = (0,10), xlims=(0,2400))\n",
    "\n",
    "\n",
    "plot(p1,p4,p2,p3, layout = (4,1), size = (800,800))\n",
    "#savefig(\"oneThroughThreeBettiCurves.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotbarcode_pjs(eireneMatrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do the actual plotting\n",
    "p1 = plot(times[:,1], BettiRates[:,1,1], color = [ :blue theme(:default)], title = \"Dimension Zero\", label = [\"birth\" ], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "p2 = plot(times[:,1], BettiRates[:,2,1], color = [ :blue theme(:default)], title = \"Dimension One\", label = [\"birth\" ], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "p3 = plot(times[:,1], BettiRates[:,3,1], color = [ :blue theme(:default)], title = \"Dimension Two\", label = [\"birth\" ], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "\n",
    "p4 = plot(times[:,1], BettiRates[:,1,2], color = [:red  theme(:default)], title = \"Dimension Zero\", label = [\"death\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "p5 = plot(times[:,1], BettiRates[:,2,2], color = [:red  theme(:default)], title = \"Dimension One\", label = [\"death\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "p6 = plot(times[:,1], BettiRates[:,3,2], color = [:red  theme(:default)], title = \"Dimension Two\", label = [\"death\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "\n",
    "\n",
    "plot(p1,p4,p2,p5,p3,p6, layout = (3,2), size = (1400,800))\n",
    "#savefig(\"oneThroughTwoBettiRates.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#post-plotting cleanup\n",
    "location = nothing\n",
    "percentBorn = nothing\n",
    "edgeDensity = nothing\n",
    "numberEdges = nothing\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "p3 = nothing\n",
    "p4 = nothing\n",
    "eireneMatrix = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make 1000 models by varying birth times and adding random edges\n",
    "numTrials = 1000\n",
    "randomModelsWithTimes = zeros(279,280,numTrials)\n",
    "\n",
    "\n",
    "\n",
    "#load all the models with the base\n",
    "for k in 1:numTrials\n",
    "    randomModelsWithTimes[:,1:279,k] = birthSortedAdjacencyMatrix[:,:]\n",
    "\n",
    "\n",
    "    #reconstruct each as a binary matrix to ease calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                randomModelsWithTimes[i,j,k] = 1\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 3000\n",
    "                randomModelsWithTimes[i,j,k] = 0\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "\n",
    "    #for each model, add 810 random connections\n",
    "    for counter in 1:810\n",
    "        initial = rand(1:279)\n",
    "        terminal = rand(1:279)\n",
    "        while terminal == initial\n",
    "            terminal = rand(1:279)\n",
    "        end\n",
    "        \n",
    "        randomModelsWithTimes[initial,terminal,k] = 1\n",
    "        randomModelsWithTimes[terminal,initial,k] = 1\n",
    "    end\n",
    "\n",
    "\n",
    "    #alter the birth times of each neuron\n",
    "    bornTimes = labelsTimesPositionsType[:,2]\n",
    "\n",
    "    for i in 1:201\n",
    "        sign = rand()\n",
    "    \n",
    "        if sign < 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.1\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] += error\n",
    "        end\n",
    "    \n",
    "        if sign >= 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.1\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] -= error\n",
    "        end\n",
    "    end\n",
    "\n",
    "    for i in 202:279\n",
    "        sign = rand()\n",
    "    \n",
    "        if sign < 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.02\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] += error\n",
    "        end\n",
    "    \n",
    "        if sign >= 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.02\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] -= error\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #reorder the current matrix by birth time\n",
    "    for i in 1:279\n",
    "        movement = 0\n",
    "    \n",
    "        while i - movement > 1 && bornTimes[i - movement] < bornTimes[i-movement - 1]\n",
    "            shadowCache = bornTimes[i - movement]\n",
    "            bornTimes[i - movement] = bornTimes[i - movement - 1]\n",
    "            bornTimes[i - movement - 1] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[i - movement, :,k]\n",
    "            randomModelsWithTimes[i - movement, :,k] = randomModelsWithTimes[i - movement - 1, :,k]\n",
    "            randomModelsWithTimes[i - movement - 1, :,k] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[:,i - movement, k]\n",
    "            randomModelsWithTimes[:,i - movement, k] = randomModelsWithTimes[:,i - movement - 1, k]\n",
    "            randomModelsWithTimes[:,i - movement - 1, k] = shadowCache\n",
    "        \n",
    "            movement += 1\n",
    "        end\n",
    "    end\n",
    "    randomModelsWithTimes[:,280,k] = bornTimes[:]\n",
    "    \n",
    "    #re-label the random models for eirene calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] == 1\n",
    "                randomModelsWithTimes[i,j,k] = maximum([bornTimes[i], bornTimes[j]])\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 0\n",
    "                randomModelsWithTimes[i,j,k] = 3000\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "println(\"output: randomModelsWithTimes, numTrials\")\n",
    "\n",
    "#cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "begin\n",
    "#perform computations\n",
    "randomCurves = pmap(i -> computeBarcodes(randomModelsWithTimes[:,1:279,i], randomModelsWithTimes[:,280,i]), 1:numTrials)   \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "RandomizedBettiCurves = zeros(279,3,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    RandomizedBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "\n",
    "\n",
    "edgeDensities = zeros(279, numTrials)\n",
    "for k in 1:numTrials\n",
    "    numberEdges = 0\n",
    "    \n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                numberEdges += 1\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        edgeDensities[i,k] += (numberEdges/binomial(i,2))\n",
    "    end\n",
    "end \n",
    "            \n",
    "    \n",
    "println(\"output: RandomizedBettiCurves, edgeDensities\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(279,3)\n",
    "bettiStd = zeros(279,3)\n",
    "averageTimes = zeros(279)\n",
    "edgeDensityMeans = zeros(279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    bettiMeans[i,1] = mean(RandomizedBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(RandomizedBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(RandomizedBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(RandomizedBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(RandomizedBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(RandomizedBettiCurves[i,3,:])\n",
    "    \n",
    "    averageTimes[i] = mean(randomModelsWithTimes[i,280,:])\n",
    "    \n",
    "    edgeDensityMeans[i,1] = mean(edgeDensities[i,:])\n",
    "    edgeDensityMeans[i,2] = std(edgeDensities[i,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd, averageTimes, edgeDensityMeans\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(averageTimes[:], bettiMeans[:,2:3], xlims = (0,2400), ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:topleft , title = \"Additional Connections and Altered Birth Times\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(averageTimes[:], bettiMeans[:,1], xlims = (0,2400), ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:topright , title = \"Additional Connections and Altered Birth Times\", label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), color = [:green theme(:default)], label = [ \"original dim 0\"])\n",
    "p3 = plot(averageTimes[:,1], edgeDensityMeans[:,1], color = [:gray], ribbon = edgeDensityMeans[:,2], line=(:dot,2),  ylabel = \"edge density\", xlims=(0,2400))\n",
    "\n",
    "plot(p1,p2,p3, layout = (3,1), size = (800,800))\n",
    "\n",
    "#savefig(\"alteredOne.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find birth rate for each dimension also\n",
    "begin\n",
    "#perform computations\n",
    "randomRates = pmap(i -> computeRates(randomModelsWithTimes[:,1:279,i], randomModelsWithTimes[:,280,i]), 1:numTrials)\n",
    "\n",
    "        \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "RandomizedBettiRates = zeros(279,3,2,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    RandomizedBettiRates[:,:,:,k] = randomRates[k]\n",
    "end\n",
    "\n",
    "        \n",
    "    \n",
    "println(\"output: RandomizedBettiCurves\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(279,3,2)\n",
    "bettiStd = zeros(279,3,2)\n",
    "\n",
    "\n",
    "for i in 1:279\n",
    "    bettiMeans[i,1,1] = mean(RandomizedBettiRates[i,1,1,:])\n",
    "    bettiMeans[i,2,1] = mean(RandomizedBettiRates[i,2,1,:])\n",
    "    bettiMeans[i,3,1] = mean(RandomizedBettiRates[i,3,1,:])\n",
    "    bettiMeans[i,1,2] = mean(RandomizedBettiRates[i,1,2,:])\n",
    "    bettiMeans[i,2,2] = mean(RandomizedBettiRates[i,2,2,:])\n",
    "    bettiMeans[i,3,2] = mean(RandomizedBettiRates[i,3,2,:])\n",
    "    \n",
    "    bettiStd[i,1,1] = std(RandomizedBettiRates[i,1,1,:])\n",
    "    bettiStd[i,2,1] = std(RandomizedBettiRates[i,2,1,:])\n",
    "    bettiStd[i,3,1] = std(RandomizedBettiRates[i,3,1,:])\n",
    "    bettiStd[i,1,2] = std(RandomizedBettiRates[i,1,2,:])\n",
    "    bettiStd[i,2,2] = std(RandomizedBettiRates[i,2,2,:])\n",
    "    bettiStd[i,3,2] = std(RandomizedBettiRates[i,3,2,:])\n",
    "\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(averageTimes[:], bettiMeans[:,1,1], xlims = (0,2400), ribbon = bettiStd[:,1,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 0\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,1,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "p2 = plot(averageTimes[:], bettiMeans[:,2,1], xlims = (0,2400), ribbon = bettiStd[:,2,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 1\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,2,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "p3 = plot(averageTimes[:], bettiMeans[:,3,1], xlims = (0,2400), ribbon = bettiStd[:,3,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 2\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,3,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "\n",
    "p4 = plot(averageTimes[:], bettiMeans[:,1,2], xlims = (0,2400), ribbon = bettiStd[:,1,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 0\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,1,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "p5 = plot(averageTimes[:], bettiMeans[:,2,2], xlims = (0,2400), ribbon = bettiStd[:,2,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 1\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,2,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "p6 = plot(averageTimes[:], bettiMeans[:,3,2], xlims = (0,2400), ribbon = bettiStd[:,3,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 2\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,3,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "\n",
    "plot(p1,p4,p2,p5,p3,p6, layout = (3,2), size = (1400,800))\n",
    "\n",
    "#savefig(\"alteredOneRates.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#post plot cleanup\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "bettiMeans = nothing\n",
    "bettiStd = nothing\n",
    "averageTimes = nothing\n",
    "edgeDensities = nothing\n",
    "edgeDensityMeans = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make 1000 models by varying birth times and adding random edges (make average wiring cost is within +/- 15%)\n",
    "numTrials = 1000\n",
    "bound = 0.15\n",
    "randomModelsWithTimes = zeros(279,280,numTrials)\n",
    "\n",
    "\n",
    "\n",
    "#find the latent average wiring cost\n",
    "latentWiringCostInfo = zeros(2)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "            a = labelsTimesPositionsType[i,3]\n",
    "            b = labelsTimesPositionsType[i,4]\n",
    "            c = labelsTimesPositionsType[j,3]\n",
    "            d = labelsTimesPositionsType[j,4]\n",
    "            latentWiringCostInfo[1] += evaluate(Euclidean(),[a,b], [c,d])\n",
    "            latentWiringCostInfo[2] += 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "            \n",
    "#find 100 sets of 810 that are within bound% of wiring cost\n",
    "begin\n",
    "\n",
    "newConnections = pmap(i -> findWiringWithinBounds(labelsTimesPositionsType[:,3:4], bound, latentWiringCostInfo), 1:numTrials)\n",
    "        \n",
    "end\n",
    "connections = zeros(810,2,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    connections[:,:,k] = newConnections[k]\n",
    "end\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#load all the models with the base\n",
    "for k in 1:numTrials\n",
    "    \n",
    "    randomModelsWithTimes[:,1:279,k] = birthSortedAdjacencyMatrix[:,:]\n",
    "    \n",
    "    \n",
    "    #reconstruct each as a binary matrix to ease calculations \n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                randomModelsWithTimes[i,j,k] = 1\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 3000\n",
    "                randomModelsWithTimes[i,j,k] = 0\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    \n",
    "    for i in 1:180\n",
    "        randomModelsWithTimes[trunc(Int,connections[i,1,k]),trunc(Int,connections[i,2,k]),k] = 1\n",
    "        randomModelsWithTimes[trunc(Int,connections[i,2,k]),trunc(Int,connections[i,1,k]),k] = 1\n",
    "    end\n",
    "\n",
    "    #alter the birth times of each neuron\n",
    "    bornTimes = labelsTimesPositionsType[:,2]\n",
    "\n",
    "    for i in 1:201\n",
    "        sign = rand()\n",
    "    \n",
    "        if sign < 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.1\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] += error\n",
    "            \n",
    "        end\n",
    "    \n",
    "        if sign >= 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.1\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] -= error\n",
    "            \n",
    "        end\n",
    "    end\n",
    "\n",
    "    for i in 202:279\n",
    "        sign = rand()\n",
    "    \n",
    "        if sign < 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.02\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] += error\n",
    "           \n",
    "        end\n",
    "    \n",
    "        if sign >= 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.02\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] -= error\n",
    "            \n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #reorder the current matrix by birth time\n",
    "    for i in 1:279\n",
    "        movement = 0\n",
    "    \n",
    "        while i - movement > 1 && bornTimes[i - movement] < bornTimes[i-movement - 1]\n",
    "            shadowCache = bornTimes[i - movement]\n",
    "            bornTimes[i - movement] = bornTimes[i - movement - 1]\n",
    "            bornTimes[i - movement - 1] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[i - movement, :,k]\n",
    "            randomModelsWithTimes[i - movement, :,k] = randomModelsWithTimes[i - movement - 1, :,k]\n",
    "            randomModelsWithTimes[i - movement - 1, :,k] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[:,i - movement, k]\n",
    "            randomModelsWithTimes[:,i - movement, k] = randomModelsWithTimes[:,i - movement - 1, k]\n",
    "            randomModelsWithTimes[:,i - movement - 1, k] = shadowCache\n",
    "        \n",
    "            movement += 1\n",
    "        end\n",
    "    end\n",
    "    randomModelsWithTimes[:,280,k] = bornTimes[:]\n",
    "    \n",
    "    #re-label the random models for eirene calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] == 1\n",
    "                randomModelsWithTimes[i,j,k] = maximum([bornTimes[i], bornTimes[j]])\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 0\n",
    "                randomModelsWithTimes[i,j,k] = 3000\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "           \n",
    "    \n",
    "end\n",
    "println(\"output: randomModelsWithTimes, numTrials\")\n",
    "\n",
    "#cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "begin\n",
    "#perform computations\n",
    "randomCurves = pmap(i -> computeBarcodes(randomModelsWithTimes[:,1:279,i], randomModelsWithTimes[:,280,i]), 1:numTrials)\n",
    "\n",
    "        \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "RandomizedBettiCurves = zeros(279,3,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    RandomizedBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "\n",
    "\n",
    "edgeDensities = zeros(279, numTrials)\n",
    "for k in 1:numTrials\n",
    "    numberEdges = 0\n",
    "    \n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                numberEdges += 1\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        edgeDensities[i,k] += (numberEdges/binomial(i,2))\n",
    "    end\n",
    "end \n",
    "      \n",
    " \n",
    "    \n",
    "println(\"output: RandomizedBettiCurves, edgeDensities\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(279,3)\n",
    "bettiStd = zeros(279,3)\n",
    "averageTimes = zeros(279)\n",
    "edgeDensityMeans = zeros(279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    bettiMeans[i,1] = mean(RandomizedBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(RandomizedBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(RandomizedBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(RandomizedBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(RandomizedBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(RandomizedBettiCurves[i,3,:])\n",
    "    \n",
    "    averageTimes[i] = mean(randomModelsWithTimes[i,280,:])\n",
    "    \n",
    "    edgeDensityMeans[i,1] = mean(edgeDensities[i,:])\n",
    "    edgeDensityMeans[i,2] = std(edgeDensities[i,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd, averageTimes, edgeDensityMeans\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(averageTimes[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Additional Connections within 15% Average Wiring Cost\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(averageTimes[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , title = \"Additional Connections within 15% Average Wiring Cost\", label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "p3 = plot(averageTimes[:,1], edgeDensityMeans[:,1], color = [:gray], ribbon = edgeDensityMeans[:,2], line=(:dot,2),  ylabel = \"edge density\", xlims=(0,2400))\n",
    "\n",
    "plot(p1,p2,p3, layout = (3,1), size = (800,800))\n",
    "\n",
    "#savefig(\"within15.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find birth rate for each dimension also\n",
    "begin\n",
    "#perform computations\n",
    "randomRates = pmap(i -> computeRates(randomModelsWithTimes[:,1:279,i], randomModelsWithTimes[:,280,i]), 1:1000)\n",
    "\n",
    "        \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "RandomizedBettiRates = zeros(279,3,2,1000)\n",
    "\n",
    "for k in 1:1000\n",
    "    RandomizedBettiRates[:,:,:,k] = randomRates[k]\n",
    "end\n",
    "\n",
    "        \n",
    "    \n",
    "println(\"output: RandomizedBettiCurves\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(279,3,2)\n",
    "bettiStd = zeros(279,3,2)\n",
    "\n",
    "\n",
    "for i in 1:279\n",
    "    bettiMeans[i,1,1] = mean(RandomizedBettiRates[i,1,1,:])\n",
    "    bettiMeans[i,2,1] = mean(RandomizedBettiRates[i,2,1,:])\n",
    "    bettiMeans[i,3,1] = mean(RandomizedBettiRates[i,3,1,:])\n",
    "    bettiMeans[i,1,2] = mean(RandomizedBettiRates[i,1,2,:])\n",
    "    bettiMeans[i,2,2] = mean(RandomizedBettiRates[i,2,2,:])\n",
    "    bettiMeans[i,3,2] = mean(RandomizedBettiRates[i,3,2,:])\n",
    "    \n",
    "    bettiStd[i,1,1] = std(RandomizedBettiRates[i,1,1,:])\n",
    "    bettiStd[i,2,1] = std(RandomizedBettiRates[i,2,1,:])\n",
    "    bettiStd[i,3,1] = std(RandomizedBettiRates[i,3,1,:])\n",
    "    bettiStd[i,1,2] = std(RandomizedBettiRates[i,1,2,:])\n",
    "    bettiStd[i,2,2] = std(RandomizedBettiRates[i,2,2,:])\n",
    "    bettiStd[i,3,2] = std(RandomizedBettiRates[i,3,2,:])\n",
    "\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(averageTimes[:], bettiMeans[:,1,1], xlims = (0,2400), ribbon = bettiStd[:,1,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 0\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,1,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "p2 = plot(averageTimes[:], bettiMeans[:,2,1], xlims = (0,2400), ribbon = bettiStd[:,2,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 1\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,2,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "p3 = plot(averageTimes[:], bettiMeans[:,3,1], xlims = (0,2400), ribbon = bettiStd[:,3,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 2\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,3,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "\n",
    "p4 = plot(averageTimes[:], bettiMeans[:,1,2], xlims = (0,2400), ribbon = bettiStd[:,1,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 0\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,1,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "p5 = plot(averageTimes[:], bettiMeans[:,2,2], xlims = (0,2400), ribbon = bettiStd[:,2,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 1\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,2,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "p6 = plot(averageTimes[:], bettiMeans[:,3,2], xlims = (0,2400), ribbon = bettiStd[:,3,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 2\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,3,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "\n",
    "plot(p1,p4,p2,p5,p3,p6, layout = (3,2), size = (1400,800))\n",
    "\n",
    "#savefig(\"within15Rates.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#post plot cleanup\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "bettiMeans = nothing\n",
    "bettiStd = nothing\n",
    "edgeDensities = nothing\n",
    "edgeDensityMeans = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check to see if betti deltas correlate with biological categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make a 279x2 array, where [i,1] is the first neuron with the same birth time as i, and [i,2] is the last such\n",
    "firstAndLast = Array{Int64, 2}(undef,279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    foundFirst = false\n",
    "    foundLast = false\n",
    "    for j in 1:279\n",
    "        if labelsTimesPositionsType[j,2] == labelsTimesPositionsType[i,2]\n",
    "            if !foundFirst\n",
    "                foundFirst = true\n",
    "                firstAndLast[i,1] = j\n",
    "            end\n",
    "        else\n",
    "            if foundFirst && !foundLast\n",
    "                firstAndLast[i,2] = j-1\n",
    "                foundLast = true\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        if j == 279 && !foundLast\n",
    "            firstAndLast[i,2] = j\n",
    "        end\n",
    "    end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function randomLinearOrdering()\n",
    "    presentTimes = labelsTimesPositionsType[:,2]\n",
    "    \n",
    "    currentNeuron = 1\n",
    "    \n",
    "    while currentNeuron <= 279\n",
    "        n =  firstAndLast[currentNeuron, 2] - firstAndLast[currentNeuron, 1] + 1\n",
    "        \n",
    "        if n != 1\n",
    "            vector = Array{Int, 1}(undef, n)\n",
    "            for i in 1:n\n",
    "                vector[i] = i\n",
    "            end\n",
    "            p = nthperm(vector, rand(1:factorial(n)))\n",
    "        \n",
    "            for i in 1:n\n",
    "                presentTimes[currentNeuron] += p[i]/(n*100 + 1)\n",
    "                currentNeuron += 1\n",
    "            end\n",
    "            \n",
    "        else\n",
    "            currentNeuron += 1\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    return presentTimes\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make 1000 models by varying birth times\n",
    "numTrials = 1000\n",
    "randomModelsWithTimes = zeros(279,280,numTrials)\n",
    "\n",
    "\n",
    "\n",
    "#load all the models with the base\n",
    "for k in 1:numTrials\n",
    "    randomModelsWithTimes[:,1:279,k] = birthSortedAdjacencyMatrix[:,:]\n",
    "\n",
    "\n",
    "    #reconstruct each as a binary matrix to ease calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                randomModelsWithTimes[i,j,k] = 1\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 3000\n",
    "                randomModelsWithTimes[i,j,k] = 0\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "\n",
    "\n",
    "    #alter the birth times of each neuron\n",
    "    bornTimes = randomLinearOrdering()\n",
    "    \n",
    "    #reorder the current matrix by birth time\n",
    "    for i in 1:279\n",
    "        movement = 0\n",
    "    \n",
    "        while i - movement > 1 && bornTimes[i - movement] < bornTimes[i-movement - 1]\n",
    "            shadowCache = bornTimes[i - movement]\n",
    "            bornTimes[i - movement] = bornTimes[i - movement - 1]\n",
    "            bornTimes[i - movement - 1] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[i - movement, :,k]\n",
    "            randomModelsWithTimes[i - movement, :,k] = randomModelsWithTimes[i - movement - 1, :,k]\n",
    "            randomModelsWithTimes[i - movement - 1, :,k] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[:,i - movement, k]\n",
    "            randomModelsWithTimes[:,i - movement, k] = randomModelsWithTimes[:,i - movement - 1, k]\n",
    "            randomModelsWithTimes[:,i - movement - 1, k] = shadowCache\n",
    "        \n",
    "            movement += 1\n",
    "        end\n",
    "    end\n",
    "    randomModelsWithTimes[:,280,k] = bornTimes[:]\n",
    "    \n",
    "    #re-label the random models for eirene calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] == 1\n",
    "                randomModelsWithTimes[i,j,k] = maximum([bornTimes[i], bornTimes[j]])\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 0\n",
    "                randomModelsWithTimes[i,j,k] = 3000\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "println(\"output: randomModelsWithTimes, numTrials\")\n",
    "\n",
    "#cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#array to hold average + and - to each dimension of betti curve per node\n",
    "bettiDeltas = zeros(279,3,2,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "print(k,\" \")\n",
    "#computes the betti curves in dimensions 0-2\n",
    "eireneMatrix = eirene(randomModelsWithTimes[:,1:279,k], model = \"vr\", minrad = 0.5, maxrad = 3001, maxdim = 2)\n",
    " \n",
    "\n",
    "#determine number of points on x-axis\n",
    "numValues = 279\n",
    "\n",
    "\n",
    "\n",
    "#make arrays to store betti curves\n",
    "barcodes = barcode(eireneMatrix, dim=0)\n",
    "amount = size(barcodes,1)\n",
    "\n",
    "\n",
    "\n",
    "#calculate and store dimension 0 betti deltas\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if abs(barcodes[j,1] - barcodes[j,2]) >= 0.01\n",
    "            if barcodes[j,1] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,1,1,k] += 1\n",
    "            end\n",
    "            if barcodes[j,2] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,1,2,k] += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#calculate and store dimension 1 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=1)\n",
    "amount = size(barcodes,1)\n",
    "\n",
    "\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if abs(barcodes[j,1] - barcodes[j,2]) >= 0.01\n",
    "            if barcodes[j,1] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,2,1,k] += 1\n",
    "            end\n",
    "            if barcodes[j,2] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,2,2,k] += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "#calculate and store dimension 2 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=2)\n",
    "amount = size(barcodes,1)\n",
    "\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if abs(barcodes[j,1] - barcodes[j,2]) >= 0.01\n",
    "            if barcodes[j,1] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,3,1,k] += 1\n",
    "            end\n",
    "            if barcodes[j,2] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,3,2,k] += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "end\n",
    "println(\"output: bettiDeltas\")\n",
    "\n",
    "#cleanup\n",
    "barcodes = nothing\n",
    "amount = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets test +/- deltas against biomarkers\n",
    "\n",
    "#variables for connections within a type\n",
    "sensoryDeltas = zeros(85*numTrials,3,2) #85 sensory neurons\n",
    "motorDeltas = zeros(115*numTrials,3,2) #115 motor neurons\n",
    "interDeltas = zeros(79*numTrials,3,2) # 79 inter neurons\n",
    "headDeltas = zeros(160*numTrials,3,2) #160 head neurons\n",
    "bodyDeltas = zeros(92*numTrials,3,2) #92 body neurons\n",
    "tailDeltas = zeros(27*numTrials,3,2) #27 tail neurons\n",
    "sensoryCounter = 1\n",
    "motorCounter = 1\n",
    "interCounter = 1\n",
    "headCounter = 1\n",
    "bodyCounter = 1\n",
    "tailCounter = 1\n",
    "\n",
    "\n",
    "for k in 1:numTrials\n",
    "#loads all the various arrays\n",
    "for i in 1:279\n",
    "   \n",
    "    #see if the neuron is sensory\n",
    "    if labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "        \n",
    "        #load dim 0-2 + and -\n",
    "        sensoryDeltas[sensoryCounter,:,:] = bettiDeltas[birthOrder[i],:,:,k]\n",
    "        sensoryCounter += 1\n",
    "    end\n",
    "        \n",
    "     #see if the neuron is motor\n",
    "    if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "        \n",
    "        #load dim 0-2 + and -\n",
    "        motorDeltas[motorCounter,:,:] = bettiDeltas[birthOrder[i],:,:,k]\n",
    "        motorCounter += 1\n",
    "    end\n",
    "    \n",
    "    #see if the neuron is inter\n",
    "    if labelsTimesPositionsType[i,5] == \"inter\"\n",
    "        \n",
    "        #load dim 0-2 + and -\n",
    "        interDeltas[interCounter,:,:] = bettiDeltas[i,:,:,k]\n",
    "        interCounter += 1\n",
    "    end\n",
    "    \n",
    "    #see if the neuron is in the head\n",
    "    if labelsTimesPositionsType[i,3] < 0\n",
    "        headDeltas[headCounter,:,:] = bettiDeltas[i,:,:,k]\n",
    "        headCounter += 1\n",
    "    end\n",
    "    \n",
    "    #see if the neuron is in the body\n",
    "    if labelsTimesPositionsType[i,3] < 0.935 && labelsTimesPositionsType[i,3] > 0\n",
    "        bodyDeltas[bodyCounter,:,:] = bettiDeltas[i,:,:,k]\n",
    "        bodyCounter += 1\n",
    "    end\n",
    "    \n",
    "    #see if the neuron is in the tail\n",
    "    if labelsTimesPositionsType[i,3] > 0.935\n",
    "        tailDeltas[tailCounter,:,:] = bettiDeltas[i,:,:,k]\n",
    "        tailCounter += 1\n",
    "    end\n",
    "   \n",
    "            \n",
    "end\n",
    "end\n",
    "println(\"Output: sensoryDeltas and motorDeltas and interDeltas and headDeltas and bodyDeltas and tailDeltas\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "checkAgainst = zeros(92*numTrials+27*numTrials)\n",
    "checkAgainst[1:92*numTrials] = bodyDeltas[:,3,1]\n",
    "checkAgainst[92*numTrials + 1 : end] = tailDeltas[:,3,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(headDeltas[:,3,1], checkAgainst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensoryDeltas = zeros(85*numTrials,3,2) #85 sensory neurons\n",
    "motorDeltas = zeros(115*numTrials,3,2) #115 motor neurons\n",
    "interDeltas = zeros(79*numTrials,3,2) # 79 inter neurons\n",
    "headDeltas = zeros(160*numTrials,3,2) #160 head neurons\n",
    "bodyDeltas = zeros(92*numTrials,3,2) #92 body neurons\n",
    "tailDeltas = zeros(27*numTrials,3,2) #27 tail neurons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot positive deltas by type\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 12)\n",
    "plotArray[1] = sensoryDeltas[:,1,2]\n",
    "plotArray[2] = motorDeltas[:,1,2]\n",
    "plotArray[3] = interDeltas[:,1,2]\n",
    "plotArray[4] = sensoryDeltas[:,1,1]\n",
    "plotArray[5] = motorDeltas[:,1,1]\n",
    "plotArray[6] = interDeltas[:,1,1]\n",
    "plotArray[7] = headDeltas[:,1,2]\n",
    "plotArray[8] = bodyDeltas[:,1,2]\n",
    "plotArray[9] = tailDeltas[:,1,2]\n",
    "plotArray[10] = headDeltas[:,1,1]\n",
    "plotArray[11] = bodyDeltas[:,1,1]\n",
    "plotArray[12] = tailDeltas[:,1,1]\n",
    "\n",
    "#violin([ \"Sensory (-,0)\" \"Motor (-,0)\" \"Inter (-,0)\" \"Sensory (+,0)\" \"Motor (+,0)\" \"Inter (+,0)\" \"Head (-,0)\" \"Body (-,0)\" \"Tail (-,0)\" \"Head (+,0)\" \"Body (+,0)\" \"Tail (+,0)\"],plotArray[:],leg=false, #=color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)]=# title = \"Dimension Zero Betti Variation\", size = (1200, 400))\n",
    "boxplot([\"Sensory (-,0)\" \"Motor (-,0)\" \"Inter (-,0)\" \"Sensory (+,0)\" \"Motor (+,0)\" \"Inter (+,0)\" \"Head (-,0)\" \"Body (-,0)\" \"Tail (-,0)\" \"Head (+,0)\" \"Body (+,0)\" \"Tail (+,0)\"],plotArray[:],leg=false,  size = (1200, 400))\n",
    "#savefig(\"dimensionZeroBettiVariation.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot positive deltas by type\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 12)\n",
    "plotArray[1] = sensoryDeltas[:,2,2]\n",
    "plotArray[2] = motorDeltas[:,2,2]\n",
    "plotArray[3] = interDeltas[:,2,2]\n",
    "plotArray[4] = sensoryDeltas[:,2,1]\n",
    "plotArray[5] = motorDeltas[:,2,1]\n",
    "plotArray[6] = interDeltas[:,2,1]\n",
    "plotArray[7] = headDeltas[:,2,2]\n",
    "plotArray[8] = bodyDeltas[:,2,2]\n",
    "plotArray[9] = tailDeltas[:,2,2]\n",
    "plotArray[10] = headDeltas[:,2,1]\n",
    "plotArray[11] = bodyDeltas[:,2,1]\n",
    "plotArray[12] = tailDeltas[:,2,1]\n",
    "\n",
    "#violin([ \"Sensory (-,1)\" \"Motor (-,1)\" \"Inter (-,1)\" \"Sensory (+,1)\" \"Motor (+,1)\" \"Inter (+,1)\" \"Head (-,1)\" \"Body (-,1)\" \"Tail (-,1)\" \"Head (+,1)\" \"Body (+,1)\" \"Tail (+,1)\"],plotArray[:],leg=false, #=color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)]=# title = \"Dimension One Betti Variation\", size = (1200, 400))\n",
    "boxplot([\"Sensory (-,1)\" \"Motor (-,1)\" \"Inter (-,1)\" \"Sensory (+,1)\" \"Motor (+,1)\" \"Inter (+,1)\" \"Head (-,1)\" \"Body (-,1)\" \"Tail (-,1)\" \"Head (+,1)\" \"Body (+,1)\" \"Tail (+,1)\"],plotArray[:],leg=false#= color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)]=#, size = (1200, 400))\n",
    "#savefig(\"dimensionOneBettiVariation.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot positive deltas by type\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 12)\n",
    "plotArray[1] = sensoryDeltas[:,3,2]\n",
    "plotArray[2] = motorDeltas[:,3,2]\n",
    "plotArray[3] = interDeltas[:,3,2]\n",
    "plotArray[4] = sensoryDeltas[:,3,1]\n",
    "plotArray[5] = motorDeltas[:,3,1]\n",
    "plotArray[6] = interDeltas[:,3,1]\n",
    "plotArray[7] = headDeltas[:,3,2]\n",
    "plotArray[8] = bodyDeltas[:,3,2]\n",
    "plotArray[9] = tailDeltas[:,3,2]\n",
    "plotArray[10] = headDeltas[:,3,1]\n",
    "plotArray[11] = bodyDeltas[:,3,1]\n",
    "plotArray[12] = tailDeltas[:,3,1]\n",
    "\n",
    "#violin([ \"Sensory (-,2)\" \"Motor (-,2)\" \"Inter (-,2)\" \"Sensory (+,2)\" \"Motor (+,2)\" \"Inter (+,2)\" \"Head (-,2)\" \"Body (-,2)\" \"Tail (-,2)\" \"Head (+,2)\" \"Body (+,2)\" \"Tail (+,2)\"],plotArray[:],leg=false, #=color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)]=# title = \"Dimension Zero Betti Variation\", size = (1200, 400))\n",
    "boxplot([ \"Sensory (-,2)\" \"Motor (-,2)\" \"Inter (-,2)\" \"Sensory (+,2)\" \"Motor (+,2)\" \"Inter (+,2)\" \"Head (-,2)\" \"Body (-,2)\" \"Tail (-,2)\" \"Head (+,2)\" \"Body (+,2)\" \"Tail (+,2)\"],plotArray[:],leg=false, #=color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)],=# size = (1200, 400))\n",
    "#savefig(\"dimensionTwoBettiVariation.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine the variances of each neuron\n",
    "stdDev = zeros(279,3,2)\n",
    "\n",
    "for k in 1:279\n",
    "    for i in 1:3\n",
    "        for j in 1:2\n",
    "            stdDev[k,i,j] = std(bettiDeltas[k,i,j,:])\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "posZeroStd = stdDev[:,1,1]\n",
    "posOneStd = stdDev[:,2,1]\n",
    "posTwoStd = stdDev[:,3,1]\n",
    "\n",
    "negZeroStd = variances[:,1,2]\n",
    "negOneStd = variances[:,2,2]\n",
    "negTwoStd = variances[:,3,2]\n",
    "\n",
    "\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = posZeroStd[:]\n",
    "plotArray[2] = negZeroStd[:]\n",
    "plotArray[3] = posOneStd[:]\n",
    "plotArray[4] = negOneStd[:]\n",
    "plotArray[5] = posTwoStd[:]\n",
    "plotArray[6] = negTwoStd[:]\n",
    "\n",
    "\n",
    "\n",
    "boxplot([ \"Dim Zero +\" \"Dim Zero -\" \"Dim One +\" \"Dim One -\" \"Dim Two +\" \"Dim Two -\"],plotArray[:],leg=false, size = (1200, 400))\n",
    "#savefig(\"bettiVariationStds.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = nothing\n",
    "tailDeltas = nothing\n",
    "bodyDeltas = nothing\n",
    "headDeltas = nothing\n",
    "interDeltas = nothing\n",
    "motorDeltas = nothing\n",
    "sensoryDeltas = nothing\n",
    "tailCounter = nothing\n",
    "bodyCounter = nothing\n",
    "headCounter = nothing\n",
    "interCounter = nothing\n",
    "motorCounter = nothing\n",
    "sensoryCounter = nothing\n",
    "bettiDeltas = nothing\n",
    "randomModelsWithTimes = nothing\n",
    "numTrials = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####end of betti curve analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#beginning of topological similarity analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes zeroth dimensional topological similarity\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(birthSortedAdjacencyMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 0)\n",
    "\n",
    "zeroDimBarcodes = barcode(eireneGraph, dim = 0)\n",
    "originalBarcodes = zeros(size(zeroDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(zeroDimBarcodes)[1] > 0\n",
    "    for i in 1:size(zeroDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 0\n",
    "        originalBarcodes[i,2] = zeroDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = zeroDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "zeroDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(birthSortedAdjacencyMatrix[:,:], x[1], x[2], labelsTimesPositionsType[:,2], 0)), dimension = 0)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        zeroDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "zeroDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            zeroDimemsionalTopologicalSimilarity[i,j] = 1 - (zeroDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        end\n",
    "        if labelsTimesPositionsType[i,2] == labelsTimesPositionsType[j,2]\n",
    "            zeroDimemsionalTopologicalSimilarity[i,j] = 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: zeroDimensionalBND, zeroDimemsionalTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "zeroDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes first dimensional topological similarity\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(birthSortedAdjacencyMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 1)\n",
    "\n",
    "oneDimBarcodes = barcode(eireneGraph, dim = 1)\n",
    "originalBarcodes = zeros(size(oneDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(oneDimBarcodes)[1] > 0\n",
    "    for i in 1:size(oneDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 1\n",
    "        originalBarcodes[i,2] = oneDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = oneDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "oneDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(birthSortedAdjacencyMatrix[:,:], x[1], x[2], labelsTimesPositionsType[:,2],1)), dimension = 1)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        oneDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "oneDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            oneDimemsionalTopologicalSimilarity[i,j] = 1 - (oneDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        end\n",
    "        if labelsTimesPositionsType[i,2] == labelsTimesPositionsType[j,2]\n",
    "            oneDimemsionalTopologicalSimilarity[i,j] = 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: oneDimensionalBND, oneDimemsionalTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "oneDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes second dimensional topological similarity\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(birthSortedAdjacencyMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 2)\n",
    "\n",
    "twoDimBarcodes = barcode(eireneGraph, dim = 2)\n",
    "originalBarcodes = zeros(size(twoDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(twoDimBarcodes)[1] > 0\n",
    "    for i in 1:size(twoDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 2\n",
    "        originalBarcodes[i,2] = twoDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = twoDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "twoDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(birthSortedAdjacencyMatrix[:,:], x[1], x[2], labelsTimesPositionsType[:,2],2)), dimension = 2)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        twoDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "twoDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            twoDimemsionalTopologicalSimilarity[i,j] = 1 - (twoDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        end\n",
    "        if labelsTimesPositionsType[i,2] == labelsTimesPositionsType[j,2]\n",
    "            twoDimemsionalTopologicalSimilarity[i,j] = 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: twoDimensionalBND, twoDimemsionalTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "twoDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute average similarity across dimensions 1-3\n",
    "averageTopologicalSimilarity = zeroDimemsionalTopologicalSimilarity[:,:]/3 + oneDimemsionalTopologicalSimilarity[:,:]/3 + twoDimemsionalTopologicalSimilarity[:,:]/3\n",
    "averageBND = zeroDimensionalBND[:,:]/3 + oneDimensionalBND[:,:]/3 + twoDimensionalBND[:,:]/3\n",
    "\n",
    "println(\"output: averageTopologicalSimilarity, averageBND\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#finds locations of identical birth times\n",
    "sameTime = zeros(279,279)\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        sameTime[i,j] = 1\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(zeroDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(zeroDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionZeroSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(oneDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension One Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(oneDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionOneSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(twoDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(twoDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionTwoSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(averageTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Averaged Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(averageBND, yflip = true, aspect_ratio=:equal, title = \"Averaged Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"averageSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save these as csv's to save computation time\n",
    "\n",
    "CSV.write(\"dimensionZeroSimilarity.csv\", DataFrame(zeroDimemsionalTopologicalSimilarity), writeheader = false)\n",
    "CSV.write(\"dimensionOneSimilarity.csv\", DataFrame(oneDimemsionalTopologicalSimilarity), writeheader = false)\n",
    "CSV.write(\"dimensionTwoSimilarity.csv\", DataFrame(twoDimemsionalTopologicalSimilarity), writeheader = false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pull from csv as another option\n",
    "zeroDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "oneDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "twoDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "\n",
    "zeroDimensionalBND = zeros(279,279)\n",
    "oneDimensionalBND = zeros(279,279)\n",
    "twoDimensionalBND = zeros(279,279)\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        zeroDimemsionalTopologicalSimilarity[i,j] = zeroDimTopSim[i,j]\n",
    "        oneDimemsionalTopologicalSimilarity[i,j] = oneDimTopSim[i,j]\n",
    "        twoDimemsionalTopologicalSimilarity[i,j] = twoDimTopSim[i,j]\n",
    "        \n",
    "        timeDiff = abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2])\n",
    "        if timeDiff != 0\n",
    "            zeroDimensionalBND[i,j] =  timeDiff * (1 - zeroDimTopSim[i,j])\n",
    "            oneDimensionalBND[i,j] =  timeDiff * (1 - oneDimTopSim[i,j])\n",
    "            twoDimensionalBND[i,j] =  timeDiff * (1 - twoDimTopSim[i,j])\n",
    "        else\n",
    "            zeroDimensionalBND[i,j] = 0\n",
    "            oneDimensionalBND[i,j] = 0\n",
    "            twoDimensionalBND[i,j] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#as all of these computations are identical, simply pick which dimension you want to analyze here\n",
    "topologicalSimilarity = zeroDimemsionalTopologicalSimilarity[:,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test on reorderability between types\n",
    "\n",
    "#variables for connections within a type\n",
    "intraSensorySimilarities = zeros(3570) #85 sensory neurons\n",
    "intraMotorSimilarities = zeros(6555) #115 motor neurons\n",
    "intraInterSimilarities = zeros(3081) # 79 inter neurons\n",
    "intraSensoryCounter = 1\n",
    "intraMotorCounter = 1\n",
    "intraInterCounter = 1\n",
    "\n",
    "\n",
    "#variables for connections between different types and a fixed type\n",
    "interSensorySimilarities = zeros(16490) \n",
    "interMotorSimilarities = zeros(18860) \n",
    "interInterSimilarities = zeros(15800)\n",
    "interSensoryCounter = 1\n",
    "interMotorCounter = 1\n",
    "interInterCounter = 1\n",
    "\n",
    "#variables for general within types connections\n",
    "intraTypeSimilariteis = zeros(13206)\n",
    "intraCounter = 1\n",
    "\n",
    "#variabels for general between types connections\n",
    "interTypeSimilariteis = zeros(25575)\n",
    "interCounter = 1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#loads all the various arrays\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        #see if the connection is intra-type\n",
    "        if labelsTimesPositionsType[i,5] == labelsTimesPositionsType[j,5]\n",
    "            \n",
    "            #determine which type the connection is, add the similarity to the list, and up the ticker\n",
    "            if labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "                intraSensorySimilarities[intraSensoryCounter] = topologicalSimilarity[i,j]\n",
    "                intraSensoryCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "                intraMotorSimilarities[intraMotorCounter] = topologicalSimilarity[i,j]\n",
    "                intraMotorCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"inter\"\n",
    "                intraInterSimilarities[intraInterCounter] = topologicalSimilarity[i,j]\n",
    "                intraInterCounter += 1\n",
    "            end\n",
    "            \n",
    "            intraTypeSimilariteis[intraCounter]  = topologicalSimilarity[i,j]\n",
    "            intraCounter += 1\n",
    "            \n",
    "        end\n",
    "        \n",
    "        #otherwise the connection is a general InterType\n",
    "        if labelsTimesPositionsType[i,5] != labelsTimesPositionsType[j,5]\n",
    "            \n",
    "            #determine which types the connection is between, add the similarity to the lists, and up the tickers\n",
    "            if labelsTimesPositionsType[i,5] == \"sensory\" || labelsTimesPositionsType[j,5] == \"sensory\"\n",
    "                interSensorySimilarities[interSensoryCounter] = topologicalSimilarity[i,j]\n",
    "                interSensoryCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"motor\" || labelsTimesPositionsType[j,5] == \"motor\"\n",
    "                interMotorSimilarities[interMotorCounter] = topologicalSimilarity[i,j]\n",
    "                interMotorCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"inter\" || labelsTimesPositionsType[j,5] == \"inter\"\n",
    "                interInterSimilarities[interInterCounter] = topologicalSimilarity[i,j]\n",
    "                interInterCounter += 1\n",
    "            end\n",
    "                      \n",
    "            interTypeSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "            interCounter += 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"many outputs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter sensory pair\n",
    "MannWhitneyUTest(intraSensorySimilarities, interSensorySimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter motor pair\n",
    "MannWhitneyUTest(intraMotorSimilarities, interMotorSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter inter pair\n",
    "MannWhitneyUTest(intraInterSimilarities, interInterSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(intraTypeSimilariteis, interTypeSimilariteis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we would also like to determine summed similarity for each type\n",
    "\n",
    "summedSensorySimilarities = zeros(85) #85 sensory neurons\n",
    "summedMotorSimilarities = zeros(115) #115 motor neurons\n",
    "summedInterSimilarities = zeros(79) # 79 inter neurons\n",
    "summedSensoryCounter = 1\n",
    "summedMotorCounter = 1\n",
    "summedInterCounter = 1\n",
    "\n",
    "for i in 1:279\n",
    "    \n",
    "    if labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "        summedSensorySimilarities[summedSensoryCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedSensoryCounter += 1\n",
    "    end\n",
    "    \n",
    "   if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "        summedMotorSimilarities[summedMotorCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedMotorCounter += 1\n",
    "    end\n",
    "    \n",
    "    if labelsTimesPositionsType[i,5] == \"inter\"\n",
    "        summedInterSimilarities[summedInterCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedInterCounter += 1\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedSensorySimilarities, summedMotorSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedSensorySimilarities, summedInterSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedMotorSimilarities, summedInterSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 8)\n",
    "plotArray[1] = interMotorSimilarities[:]\n",
    "plotArray[2] = intraMotorSimilarities[:]\n",
    "plotArray[3] = interSensorySimilarities[:]\n",
    "plotArray[4] = intraSensorySimilarities[:]\n",
    "plotArray[5] = interInterSimilarities[:]\n",
    "plotArray[6] = intraInterSimilarities[:]\n",
    "plotArray[7] = interTypeSimilariteis[:]\n",
    "plotArray[8] = intraTypeSimilariteis[:]\n",
    "\n",
    "#violin([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"inter-Type\"],plotArray,leg=false, color = theme(:default))\n",
    "boxplot([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"intra-Type\"],title = \"Dimension Zero Similarity\", plotArray[1:6],leg=false)\n",
    "#savefig(\"dimensionZeroSimilarityPlots.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 3)\n",
    "plotArray[1] = summedSensorySimilarities[:]\n",
    "plotArray[2] = summedMotorSimilarities[:]\n",
    "plotArray[3] = summedInterSimilarities[:]\n",
    "\n",
    "\n",
    "boxplot([\"sensory\" \"motor\" \"inter\"],title = \"Dimension Zero Summed Similarity\", plotArray,leg=false)\n",
    "#savefig(\"dimensionZeroSummedSimilarityPlots.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#as all of these computations are identical, simply pick which dimension you want to analyze here\n",
    "topologicalSimilarity = zeroDimemsionalTopologicalSimilarity[:,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test on reorderability between region\n",
    "\n",
    "#variables for connections within a region\n",
    "intraHeadSimilarities = zeros(12720) #160 head neurons\n",
    "intraBodySimilarities = zeros(4186) #92 body neurons\n",
    "intraTailSimilarities = zeros(351) # 27 tail neurons\n",
    "intraHeadCounter = 1\n",
    "intraBodyCounter = 1\n",
    "intraTailCounter = 1\n",
    "\n",
    "\n",
    "#variables for connections between different types and a fixed type\n",
    "interHeadSimilarities = zeros(19040) \n",
    "interBodySimilarities = zeros(17204) \n",
    "interTailSimilarities = zeros(6804)\n",
    "interHeadCounter = 1\n",
    "interBodyCounter = 1\n",
    "interTailCounter = 1\n",
    "\n",
    "#variables for general within types connections\n",
    "intraRegionSimilariteis = zeros(17257)\n",
    "intraCounter = 1\n",
    "\n",
    "#variabels for general between types connections\n",
    "interRegionSimilariteis = zeros(21524)\n",
    "interCounter = 1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#loads all the various arrays\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        #see which region i belongs to\n",
    "        if labelsTimesPositionsType[i,3] < 0\n",
    "            #i is in the head\n",
    "            \n",
    "            if labelsTimesPositionsType[j,3] < 0\n",
    "                #j is in the head\n",
    "                intraHeadSimilarities[intraHeadCounter] = topologicalSimilarity[i,j]\n",
    "                intraRegionSimilariteis[intraCounter] = topologicalSimilarity[i,j]\n",
    "                intraHeadCounter += 1\n",
    "                intraCounter += 1\n",
    "                \n",
    "            elseif 0.935 > labelsTimesPositionsType[j,3] > 0\n",
    "               #j is in the body\n",
    "                interHeadSimilarities[interHeadCounter] = topologicalSimilarity[i,j]\n",
    "                interBodySimilarities[interBodyCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interHeadCounter += 1\n",
    "                interBodyCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            else\n",
    "                #j is in the tail\n",
    "                interHeadSimilarities[interHeadCounter] = topologicalSimilarity[i,j]\n",
    "                interTailSimilarities[interTailCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interHeadCounter += 1\n",
    "                interTailCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            end\n",
    "            \n",
    "               \n",
    "        elseif 0.935 > labelsTimesPositionsType[i,3] > 0\n",
    "           #i is in the body\n",
    "            \n",
    "            if labelsTimesPositionsType[j,3] < 0\n",
    "                #j is in the head\n",
    "                interHeadSimilarities[interHeadCounter] = topologicalSimilarity[i,j]\n",
    "                interBodySimilarities[interBodyCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interHeadCounter += 1\n",
    "                interBodyCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            elseif 0.935 > labelsTimesPositionsType[j,3] > 0\n",
    "               #j is in the body\n",
    "                intraBodySimilarities[intraBodyCounter] = topologicalSimilarity[i,j]\n",
    "                intraRegionSimilariteis[intraCounter] = topologicalSimilarity[i,j]\n",
    "                intraBodyCounter += 1\n",
    "                intraCounter += 1\n",
    "                \n",
    "            else\n",
    "                #j is in the tail\n",
    "                interBodySimilarities[interBodyCounter] = topologicalSimilarity[i,j]\n",
    "                interTailSimilarities[interTailCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interBodyCounter += 1\n",
    "                interTailCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            end\n",
    "            \n",
    "        else\n",
    "            #i is in the tail\n",
    "            \n",
    "            if labelsTimesPositionsType[j,3] < 0\n",
    "                #j is in the head\n",
    "                interHeadSimilarities[interHeadCounter] = topologicalSimilarity[i,j]\n",
    "                interTailSimilarities[interTailCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interHeadCounter += 1\n",
    "                interTailCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            elseif 0.935 > labelsTimesPositionsType[j,3] > 0\n",
    "               #j is in the body\n",
    "                interBodySimilarities[interBodyCounter] = topologicalSimilarity[i,j]\n",
    "                interTailSimilarities[interTailCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interBodyCounter += 1\n",
    "                interTailCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            else\n",
    "                #j is in the tail\n",
    "                intraTailSimilarities[intraTailCounter] = topologicalSimilarity[i,j]\n",
    "                intraRegionSimilariteis[intraCounter] = topologicalSimilarity[i,j]\n",
    "                intraTailCounter += 1\n",
    "                intraCounter += 1\n",
    "                \n",
    "            end\n",
    "            \n",
    "        end\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"many outputs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter sensory pair\n",
    "MannWhitneyUTest(intraHeadSimilarities, interHeadSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter motor pair\n",
    "MannWhitneyUTest(intraBodySimilarities, interBodySimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter inter pair\n",
    "MannWhitneyUTest(intraTailSimilarities, interTailSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(intraRegionSimilariteis, interRegionSimilariteis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we would also like to determine summed similarity for each type\n",
    "\n",
    "summedHeadSimilarities = zeros(160) #160 head neurons\n",
    "summedBodySimilarities = zeros(92) #92 body neurons\n",
    "summedTailSimilarities = zeros(27) #27 tail neurons\n",
    "summedHeadCounter = 1\n",
    "summedBodyCounter = 1\n",
    "summedTailCounter = 1\n",
    "\n",
    "for i in 1:279\n",
    "    \n",
    "    if labelsTimesPositionsType[i,3] < 0\n",
    "        summedHeadSimilarities[summedHeadCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedHeadCounter += 1\n",
    "    elseif labelsTimesPositionsType[i,3] < 0.935\n",
    "        summedBodySimilarities[summedBodyCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedBodyCounter += 1\n",
    "    else\n",
    "        summedTailSimilarities[summedTailCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedTailCounter += 1\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedHeadSimilarities, summedBodySimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedHeadSimilarities, summedTailSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedBodySimilarities, summedTailSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 8)\n",
    "plotArray[1] = interHeadSimilarities[:]\n",
    "plotArray[2] = intraHeadSimilarities[:]\n",
    "plotArray[3] = interBodySimilarities[:]\n",
    "plotArray[4] = intraBodySimilarities[:]\n",
    "plotArray[5] = interTailSimilarities[:]\n",
    "plotArray[6] = intraTailSimilarities[:]\n",
    "plotArray[7] = interRegionSimilariteis[:]\n",
    "plotArray[8] = intraRegionSimilariteis[:]\n",
    "\n",
    "#violin([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"inter-Type\"],plotArray,leg=false, color = theme(:default))\n",
    "boxplot([\"inter-Head\" \"intra-Head\" \"inter-Body\" \"intra-Body\" \"inter-Tail\" \"intra-Tail\" \"inter-Region\" \"intra-Region\"],title = \"Dimension Zero Similarity\", plotArray[1:6],leg=false)\n",
    "#savefig(\"dimensionZeroSimilarityPlotsRegion.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 3)\n",
    "plotArray[1] = summedHeadSimilarities[:]\n",
    "plotArray[2] = summedBodySimilarities[:]\n",
    "plotArray[3] = summedTailSimilarities[:]\n",
    "\n",
    "\n",
    "#violin([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"inter-Type\"],plotArray,leg=false, color = theme(:default))\n",
    "boxplot([\"Head\" \"Body\" \"Tail\"],title = \"Dimension Zero Similarity\", plotArray,leg=false)\n",
    "#savefig(\"dimensionZeroSummedSimilarityPlotsRegion.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixedPairValues = zeros(15678,3,2)\n",
    "unmixedPairValues = zeros(23103,3,2)\n",
    "mixedCounter = 0\n",
    "unmixedCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            unmixedCounter += 1\n",
    "            unmixedPairValues[unmixedCounter,1,1] = zeroDimemsionalTopologicalSimilarity[i,j]\n",
    "            unmixedPairValues[unmixedCounter,2,1] = oneDimemsionalTopologicalSimilarity[i,j]\n",
    "            unmixedPairValues[unmixedCounter,3,1] = twoDimemsionalTopologicalSimilarity[i,j]\n",
    "            \n",
    "            unmixedPairValues[unmixedCounter,1,2] = zeroDimensionalBND[i,j]\n",
    "            unmixedPairValues[unmixedCounter,2,2] = oneDimensionalBND[i,j]\n",
    "            unmixedPairValues[unmixedCounter,3,2] = twoDimensionalBND[i,j]\n",
    "        else\n",
    "            mixedCounter += 1\n",
    "            mixedPairValues[mixedCounter,1,1] = zeroDimemsionalTopologicalSimilarity[i,j]\n",
    "            mixedPairValues[mixedCounter,2,1] = oneDimemsionalTopologicalSimilarity[i,j]\n",
    "            mixedPairValues[mixedCounter,3,1] = twoDimemsionalTopologicalSimilarity[i,j]\n",
    "            \n",
    "            mixedPairValues[mixedCounter,1,2] = zeroDimensionalBND[i,j]\n",
    "            mixedPairValues[mixedCounter,2,2] = oneDimensionalBND[i,j]\n",
    "            mixedPairValues[mixedCounter,3,2] = twoDimensionalBND[i,j]\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Approximate Mann-Whitney U test\n",
       "-------------------------------\n",
       "Population details:\n",
       "    parameter of interest:   Location parameter (pseudomedian)\n",
       "    value under h_0:         0\n",
       "    point estimate:          0.5928271958104406\n",
       "\n",
       "Test summary:\n",
       "    outcome with 95% confidence: reject h_0\n",
       "    two-sided p-value:           <1e-99\n",
       "\n",
       "Details:\n",
       "    number of observations in each group: [23103, 15678]\n",
       "    Mann-Whitney-U statistic:             2.58931402e8\n",
       "    rank sums:                            [4.40801e8, 3.11202e8]\n",
       "    adjustment for ties:                  4.121762787606e12\n",
       "    normal approximation (μ, σ):          (7.7826985e7, 1.0430120403435383e6)\n"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(unmixedPairValues[:,3,1], mixedPairValues[:,3,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Approximate Mann-Whitney U test\n",
       "-------------------------------\n",
       "Population details:\n",
       "    parameter of interest:   Location parameter (pseudomedian)\n",
       "    value under h_0:         0\n",
       "    point estimate:          0.0\n",
       "\n",
       "Test summary:\n",
       "    outcome with 95% confidence: reject h_0\n",
       "    two-sided p-value:           <1e-8\n",
       "\n",
       "Details:\n",
       "    number of observations in each group: [23103, 15678]\n",
       "    Mann-Whitney-U statistic:             1.75782955e8\n",
       "    rank sums:                            [4.72577e8, 2.79425e8]\n",
       "    adjustment for ties:                  1.6575844277136e13\n",
       "    normal approximation (μ, σ):          (-5.321462e6, 915378.9501005334)\n"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(unmixedPairValues[:,1,2], mixedPairValues[:,1,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = unmixedPairValues[:,1,1]\n",
    "plotArray[2] = mixedPairValues[:,1,1]\n",
    "plotArray[3] = unmixedPairValues[:,2,1]\n",
    "plotArray[4] = mixedPairValues[:,2,1]\n",
    "plotArray[5] = unmixedPairValues[:,3,1]\n",
    "plotArray[6] = mixedPairValues[:,3,1]\n",
    "\n",
    "boxplot([\"zero un-mixed\" \"zero mixed\" \"one un-mixed\" \"one mixed\" \"two un-mixed\" \"two mixed\"], title = \"Topological Similarity\", plotArray[1:6],leg=false)\n",
    "savefig(\"mixedUnmixedTopSim.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = unmixedPairValues[:,1,2]\n",
    "plotArray[2] = mixedPairValues[:,1,2]\n",
    "plotArray[3] = unmixedPairValues[:,2,2]\n",
    "plotArray[4] = mixedPairValues[:,2,2]\n",
    "plotArray[5] = unmixedPairValues[:,3,2]\n",
    "plotArray[6] = mixedPairValues[:,3,2]\n",
    "\n",
    "boxplot([\"zero un-mixed\" \"zero mixed\" \"one un-mixed\" \"one mixed\" \"two un-mixed\" \"two mixed\"], title = \"Bottleneck Distance\", plotArray[1:6],leg=false)\n",
    "savefig(\"mixedUnmixedBND.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "#top sim with time gap deleted\n",
    "newTimesList = labelsTimesPositionsType[:,2]\n",
    "for i in 202:279\n",
    "    newTimesList[i] -= 765.55\n",
    "end\n",
    "\n",
    "\n",
    "gapDeletedMatrix = birthSortedAdjacencyMatrix[:,:]\n",
    "\n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        \n",
    "        if gapDeletedMatrix[i,j] < 3000 && ( i > 201  || j > 201)\n",
    "            \n",
    "            max = maximum([i,j])\n",
    "            gapDeletedMatrix[i,j] = newTimesList[max]\n",
    "        end\n",
    "    end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15678output: zeroDimensionalMixedPairBND, zeroDimemsionalMixedPairTopologicalSimilarity\n"
     ]
    }
   ],
   "source": [
    "#computes zeroth dimensional topological similarity for all mixed pairs, as the value *cannot* change for all other pairs\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(gapDeletedMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 0)\n",
    "\n",
    "zeroDimBarcodes = barcode(eireneGraph, dim = 0)\n",
    "originalBarcodes = zeros(size(zeroDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(zeroDimBarcodes)[1] > 0\n",
    "    for i in 1:size(zeroDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 0\n",
    "        originalBarcodes[i,2] = zeroDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = zeroDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 15678)\n",
    "counter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "zeroDimensionalMixedPairBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(gapDeletedMatrix[:,:], x[1], x[2], newTimesList, 0)), dimension = 0)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        storageCounter += 1\n",
    "        zeroDimensionalMixedPairBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "zeroDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            zeroDimemsionalMixedPairTopologicalSimilarity[i,j] = zeroDimemsionalTopologicalSimilarity[i,j]\n",
    "            zeroDimensionalMixedPairBND[i,j] = zeroDimensionalBND[i,j]\n",
    "        else\n",
    "        \n",
    "            if newTimesList[i] != newTimesList[j]\n",
    "                zeroDimemsionalMixedPairTopologicalSimilarity[i,j] = 1 - (zeroDimensionalMixedPairBND[i,j] / (newTimesList[i] - newTimesList[j]))\n",
    "            end\n",
    "            if newTimesList[i] == newTimesList[j]\n",
    "                zeroDimemsionalMixedPairTopologicalSimilarity[i,j] = 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: zeroDimensionalMixedPairBND, zeroDimemsionalMixedPairTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "zeroDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: oneDimensionalMixedPairBND, oneDimemsionalMixedPairTopologicalSimilarity\n"
     ]
    }
   ],
   "source": [
    "#computes first dimensional topological similarity for all mixed pairs, as the value *cannot* change for all other pairs\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(gapDeletedMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 1)\n",
    "\n",
    "oneDimBarcodes = barcode(eireneGraph, dim = 1)\n",
    "originalBarcodes = zeros(size(oneDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(oneDimBarcodes)[1] > 0\n",
    "    for i in 1:size(oneDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 1\n",
    "        originalBarcodes[i,2] = oneDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = oneDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 15678)\n",
    "counter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "oneDimensionalMixedPairBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(gapDeletedMatrix[:,:], x[1], x[2], newTimesList[:], 1)), dimension = 1)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        storageCounter += 1\n",
    "        oneDimensionalMixedPairBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "oneDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            oneDimemsionalMixedPairTopologicalSimilarity[i,j] = oneDimemsionalTopologicalSimilarity[i,j]\n",
    "            oneDimensionalMixedPairBND[i,j] = oneDimensionalBND[i,j]\n",
    "        else\n",
    "        \n",
    "            if newTimesList[i] != newTimesList[j]\n",
    "                oneDimemsionalMixedPairTopologicalSimilarity[i,j] = 1 - (oneDimensionalMixedPairBND[i,j] / (newTimesList[i] - newTimesList[j]))\n",
    "            end\n",
    "            if newTimesList[i] == newTimesList[j]\n",
    "                oneDimemsionalMixedPairTopologicalSimilarity[i,j] = 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: oneDimensionalMixedPairBND, oneDimemsionalMixedPairTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "oneDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: twoDimensionalMixedPairBND, twoDimemsionalMixedPairTopologicalSimilarity\n"
     ]
    }
   ],
   "source": [
    "#computes second dimensional topological similarity for all mixed pairs, as the value *cannot* change for all other pairs\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(gapDeletedMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 2)\n",
    "\n",
    "twoDimBarcodes = barcode(eireneGraph, dim = 2)\n",
    "originalBarcodes = zeros(size(twoDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(twoDimBarcodes)[1] > 0\n",
    "    for i in 1:size(twoDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 2\n",
    "        originalBarcodes[i,2] = twoDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = twoDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 15678)\n",
    "counter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "twoDimensionalMixedPairBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(gapDeletedMatrix[:,:], x[1], x[2], newTimesList[:], 2)), dimension = 2)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        storageCounter += 1\n",
    "        twoDimensionalMixedPairBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "twoDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            twoDimemsionalMixedPairTopologicalSimilarity[i,j] = twoDimemsionalTopologicalSimilarity[i,j]\n",
    "            twoDimensionalMixedPairBND[i,j] = twoDimensionalBND[i,j]\n",
    "        else\n",
    "        \n",
    "            if newTimesList[i] != newTimesList[j]\n",
    "                twoDimemsionalMixedPairTopologicalSimilarity[i,j] = 1 - (twoDimensionalMixedPairBND[i,j] / (newTimesList[i] - newTimesList[j]))\n",
    "            end\n",
    "            if newTimesList[i] == newTimesList[j]\n",
    "                twoDimemsionalMixedPairTopologicalSimilarity[i,j] = 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: twoDimensionalMixedPairBND, twoDimemsionalMixedPairTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "twoDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: averageMixedPairTopologicalSimilarity, averageMixedPairBND\n"
     ]
    }
   ],
   "source": [
    "#compute average similarity across dimensions 1-3\n",
    "averageMixedPairTopologicalSimilarity = zeroDimemsionalMixedPairTopologicalSimilarity[:,:]/3 + oneDimemsionalMixedPairTopologicalSimilarity[:,:]/3 + twoDimemsionalMixedPairTopologicalSimilarity[:,:]/3\n",
    "averageMixedPairBND = zeroDimensionalMixedPairBND[:,:]/3 + oneDimensionalMixedPairBND[:,:]/3 + twoDimensionalMixedPairBND[:,:]/3\n",
    "\n",
    "println(\"output: averageMixedPairTopologicalSimilarity, averageMixedPairBND\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(zeroDimemsionalMixedPairTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Gap Deleted Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(zeroDimensionalMixedPairBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Gap Deleted Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "savefig(\"dimensionZeroGapDeletedSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(oneDimemsionalMixedPairTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension One Gap Deleted Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(oneDimensionalMixedPairBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Gap Deleted Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "savefig(\"dimensionOneGapDeletedSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(twoDimemsionalMixedPairTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Gap Deleted Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(twoDimensionalMixedPairBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Gap Deleted Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "savefig(\"dimensionTwoGapDeletedSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(averageMixedPairTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Averaged Gap Deleted Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(averageMixedPairBND, yflip = true, aspect_ratio=:equal, title = \"Averaged Gap Deleted Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "savefig(\"averagedGapDeletedSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"dimensionTwoMixedPairSimilarity.csv\""
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#save these as csv's to save computation time\n",
    "\n",
    "CSV.write(\"dimensionZeroMixedPairSimilarity.csv\", DataFrame(zeroDimemsionalMixedPairTopologicalSimilarity), writeheader = false)\n",
    "CSV.write(\"dimensionOneMixedPairSimilarity.csv\", DataFrame(oneDimemsionalMixedPairTopologicalSimilarity), writeheader = false)\n",
    "CSV.write(\"dimensionTwoMixedPairSimilarity.csv\", DataFrame(twoDimemsionalMixedPairTopologicalSimilarity), writeheader = false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pull from csv as another option\n",
    "zeroDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "oneDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "twoDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "\n",
    "zeroDimensionalMixedPairBND = zeros(279,279)\n",
    "oneDimensionalMixedPairBND = zeros(279,279)\n",
    "twoDimensionalMixedPairBND = zeros(279,279)\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        zeroDimemsionalMixedPairTopologicalSimilarity[i,j] = zeroDimTopSimGD[i,j]\n",
    "        oneDimemsionalMixedPairTopologicalSimilarity[i,j] = oneDimTopSimGD[i,j]\n",
    "        twoDimemsionalMixedPairTopologicalSimilarity[i,j] = twoDimTopSimGD[i,j]\n",
    "        \n",
    "        timeDiff = abs(newTimesList[i] - newTimesList[j])\n",
    "        if timeDiff != 0\n",
    "            zeroDimensionalMixedPairBND[i,j] =  timeDiff * (1 - zeroDimTopSimGD[i,j])\n",
    "            oneDimensionalMixedPairBND[i,j] =  timeDiff * (1 - oneDimTopSimGD[i,j])\n",
    "            twoDimensionalMixedPairBND[i,j] =  timeDiff * (1 - twoDimTopSimGD[i,j])\n",
    "        else\n",
    "            zeroDimensionalMixedPairBND[i,j] = 0\n",
    "            oneDimensionalMixedPairBND[i,j] = 0\n",
    "            twoDimensionalMixedPairBND[i,j] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixedPairValues = zeros(15678,3,2)\n",
    "unmixedPairValues = zeros(23103,3,2)\n",
    "mixedCounter = 0\n",
    "unmixedCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            unmixedCounter += 1\n",
    "            unmixedPairValues[unmixedCounter,1,1] = zeroDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            unmixedPairValues[unmixedCounter,2,1] = oneDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            unmixedPairValues[unmixedCounter,3,1] = twoDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            \n",
    "            unmixedPairValues[unmixedCounter,1,2] = zeroDimensionalMixedPairBND[i,j]\n",
    "            unmixedPairValues[unmixedCounter,2,2] = oneDimensionalMixedPairBND[i,j]\n",
    "            unmixedPairValues[unmixedCounter,3,2] = twoDimensionalMixedPairBND[i,j]\n",
    "        else\n",
    "            mixedCounter += 1\n",
    "            mixedPairValues[mixedCounter,1,1] = zeroDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            mixedPairValues[mixedCounter,2,1] = oneDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            mixedPairValues[mixedCounter,3,1] = twoDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            \n",
    "            mixedPairValues[mixedCounter,1,2] = zeroDimensionalMixedPairBND[i,j]\n",
    "            mixedPairValues[mixedCounter,2,2] = oneDimensionalMixedPairBND[i,j]\n",
    "            mixedPairValues[mixedCounter,3,2] = twoDimensionalMixedPairBND[i,j]\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Approximate Mann-Whitney U test\n",
       "-------------------------------\n",
       "Population details:\n",
       "    parameter of interest:   Location parameter (pseudomedian)\n",
       "    value under h_0:         0\n",
       "    point estimate:          0.0\n",
       "\n",
       "Test summary:\n",
       "    outcome with 95% confidence: reject h_0\n",
       "    two-sided p-value:           <1e-33\n",
       "\n",
       "Details:\n",
       "    number of observations in each group: [23103, 15678]\n",
       "    Mann-Whitney-U statistic:             1.69877148e8\n",
       "    rank sums:                            [4.23368e8, 3.28634e8]\n",
       "    adjustment for ties:                  1.6575606421554e13\n",
       "    normal approximation (μ, σ):          (-1.1227269e7, 915381.5576514433)\n"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(unmixedPairValues[:,1,1], mixedPairValues[:,1,1])\n",
    "\n",
    "#0 - -33        0\n",
    "#1 - -99     - .0457\n",
    "#2 - -99     .347"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Approximate Mann-Whitney U test\n",
       "-------------------------------\n",
       "Population details:\n",
       "    parameter of interest:   Location parameter (pseudomedian)\n",
       "    value under h_0:         0\n",
       "    point estimate:          0.0\n",
       "\n",
       "Test summary:\n",
       "    outcome with 95% confidence: reject h_0\n",
       "    two-sided p-value:           <1e-8\n",
       "\n",
       "Details:\n",
       "    number of observations in each group: [23103, 15678]\n",
       "    Mann-Whitney-U statistic:             1.757953335e8\n",
       "    rank sums:                            [4.72562e8, 2.7944e8]\n",
       "    adjustment for ties:                  1.657690976307e13\n",
       "    normal approximation (μ, σ):          (-5.3090835e6, 915367.2693551129)\n"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(unmixedPairValues[:,1,2], mixedPairValues[:,1,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = unmixedPairValues[:,1,1]\n",
    "plotArray[2] = mixedPairValues[:,1,1]\n",
    "plotArray[3] = unmixedPairValues[:,2,1]\n",
    "plotArray[4] = mixedPairValues[:,2,1]\n",
    "plotArray[5] = unmixedPairValues[:,3,1]\n",
    "plotArray[6] = mixedPairValues[:,3,1]\n",
    "\n",
    "boxplot([\"zero un-mixed\" \"zero mixed\" \"one un-mixed\" \"one mixed\" \"two un-mixed\" \"two mixed\"], title = \"Topological Similarity\", plotArray[1:6],leg=false)\n",
    "savefig(\"mixedUnmixedTopSim.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = unmixedPairValues[:,1,2]\n",
    "plotArray[2] = mixedPairValues[:,1,2]\n",
    "plotArray[3] = unmixedPairValues[:,2,2]\n",
    "plotArray[4] = mixedPairValues[:,2,2]\n",
    "plotArray[5] = unmixedPairValues[:,3,2]\n",
    "plotArray[6] = mixedPairValues[:,3,2]\n",
    "\n",
    "boxplot([\"zero un-mixed\" \"zero mixed\" \"one un-mixed\" \"one mixed\" \"two un-mixed\" \"two mixed\"], title = \"Bottleneck Distance\", plotArray[1:6],leg=false)\n",
    "savefig(\"mixedUnmixedBND.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine top sim vs distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find set of actually existing time differences, to cut down computation costs on the next thing\n",
    "allTimeDifferences = zeros(38781)\n",
    "counter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        allTimeDifferences[counter] =  round(abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]); digits=2) \n",
    "    end\n",
    "end\n",
    "\n",
    "sort!(allTimeDifferences)\n",
    "\n",
    "actualTimeDifferences = zeros(10910)\n",
    "counter = 1\n",
    "actualTimeDifferences[counter] = allTimeDifferences[1]\n",
    "\n",
    "for i in 2:38781\n",
    "    if allTimeDifferences[i-1] != allTimeDifferences[i]\n",
    "        counter += 1\n",
    "        actualTimeDifferences[counter] = allTimeDifferences[i]\n",
    "    end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do similarity over increasing window calculations\n",
    "growingWindows = zeros(10910,3,2)  #window size x dimension x mean/std\n",
    "\n",
    "runningSumZero = zeros(38781)\n",
    "runningSumOne = zeros(38781)\n",
    "runningSumTwo = zeros(38781)\n",
    "counter = 0\n",
    "\n",
    "for k in actualTimeDifferences\n",
    "    counter += 1\n",
    "    runningCount = 0\n",
    "    \n",
    "    \n",
    "    #find averages given a time window of k\n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            \n",
    "            if abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]) <= k\n",
    "                runningCount += 1\n",
    "                runningSumZero[runningCount] = zeroDimemsionalTopologicalSimilarity[i,j] \n",
    "                runningSumOne[runningCount] = oneDimemsionalTopologicalSimilarity[i,j]   \n",
    "                runningSumTwo[runningCount] = twoDimemsionalTopologicalSimilarity[i,j]  \n",
    "            end\n",
    "            \n",
    "        end\n",
    "    end\n",
    "    \n",
    "    growingWindows[counter,1,1] = mean(runningSumZero[1:runningCount])\n",
    "    growingWindows[counter,2,1] = mean(runningSumZero[1:runningCount])\n",
    "    growingWindows[counter,3,1] = mean(runningSumZero[1:runningCount])\n",
    "    \n",
    "    growingWindows[counter,1,2] = std(runningSumZero[1:runningCount])\n",
    "    growingWindows[counter,2,2] = std(runningSumZero[1:runningCount])\n",
    "    growingWindows[counter,3,2] = std(runningSumZero[1:runningCount])\n",
    "end\n",
    "\n",
    "print(\"output: growingWindows, runningSumZero, runningSumOne, runningSumTwo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for each pair of neurons, find their distance and their top sim\n",
    "dimZeroDistances = zeros(38781,2)\n",
    "dimOneDistances = zeros(38781,2)\n",
    "dimTwoDistances = zeros(38781,2)\n",
    "birthTimeColors = zeros(38781,2)\n",
    "randomColors = zeros(38781,2)\n",
    "birthNumber = zeros(38781,2)\n",
    "byNeuronType = zeros(38781,2)\n",
    "byRegion = zeros(38781,2)\n",
    "byxcoor = zeros(38781,2)\n",
    "movieColors = zeros(38781,2)\n",
    "\n",
    "counter = 0\n",
    "colors = rand(279)\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        counter += 1\n",
    "        time = round(abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]); digits=2) \n",
    "        \n",
    "        dimZeroDistances[counter,1] = time \n",
    "        dimOneDistances[counter,1] = time \n",
    "        dimTwoDistances[counter,1] = time\n",
    "        \n",
    "        dimZeroDistances[counter,2] = zeroDimemsionalTopologicalSimilarity[i,j] \n",
    "        dimOneDistances[counter,2] = oneDimemsionalTopologicalSimilarity[i,j]   \n",
    "        dimTwoDistances[counter,2] = twoDimemsionalTopologicalSimilarity[i,j]\n",
    "        \n",
    "        randomColors[counter,:] = [colors[i],colors[j]]\n",
    "        birthTimeColors[counter,:] = [labelsTimesPositionsType[i,2] , labelsTimesPositionsType[j,2]]\n",
    "        birthNumber[counter,:] = [i,j]\n",
    "        \n",
    "        #do region sorting\n",
    "        if labelsTimesPositionsType[i,3] < 0\n",
    "            byRegion[counter,1] = 0\n",
    "        elseif labelsTimesPositionsType[i,3] < 0.935\n",
    "            byRegion[counter,1] = 0.5\n",
    "        else\n",
    "            byRegion[counter,1] = 1\n",
    "        end\n",
    "        \n",
    "        if labelsTimesPositionsType[j,3] < 0\n",
    "            byRegion[counter,2] = 0\n",
    "        elseif labelsTimesPositionsType[j,3] < 0.935\n",
    "            byRegion[counter,2] = 0.5\n",
    "        else\n",
    "            byRegion[counter,2] = 1\n",
    "        end\n",
    "        \n",
    "        \n",
    "        #type sorting\n",
    "        if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "            byNeuronType[counter,1] = 0\n",
    "        elseif labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "            byNeuronType[counter,1] = 0.5\n",
    "        else\n",
    "            byNeuronType[counter,1] = 1\n",
    "        end\n",
    "        \n",
    "        if labelsTimesPositionsType[j,5] == \"motor\"\n",
    "            byNeuronType[counter,2] = 0\n",
    "        elseif labelsTimesPositionsType[j,5] == \"sensory\"\n",
    "            byNeuronType[counter,2] = 0.5\n",
    "        else\n",
    "            byNeuronType[counter,2] = 1\n",
    "        end\n",
    "        \n",
    "        byxcoor[counter,:] = [labelsTimesPositionsType[i,3] , labelsTimesPositionsType[j,3]]\n",
    "        \n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1 = scatter(dimZeroDistances[:,1], dimZeroDistances[:,2], color = [:green theme(:default)], legend=false , title = \"Topological Similarity vs Temporal Distance\", label = [\"dim 0\"],  xlabel = \"Temporal Distance\", ylabel = \"Topological Similarity\")\n",
    "p2 = scatter(dimOneDistances[:,1], dimOneDistances[:,2], color = [:red theme(:default)], legend=false , title = \"Topological Similarity vs Temporal Distance\", label = [\"dim 1\"],  xlabel = \"Temporal Distance\", ylabel = \"Topological Similarity\")\n",
    "p3 = scatter(dimTwoDistances[:,1], dimTwoDistances[:,2], color = [:blue theme(:default)], legend=false , title = \"Topological Similarity vs Temporal Distance\", label = [\"dim 2\"],  xlabel = \"Temporal Distance\", ylabel = \"Topological Similarity\")\n",
    "\n",
    "plot(p1,p2,p3,layout = (3,1), size = (1000,1000))\n",
    "#savefig(\"topSimByDistance.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the gworing window curves\n",
    "\n",
    "\n",
    "p1 = plot(actualTimeDifferences[:], growingWindows[:,1,1], ribbon = growingWindows[:,1,2], color = [:green theme(:default)], legend=:bottomright , title = \"Topological Similarity over Growing Time Window\", label = [\"dim 0\"],  xlabel = \"window size\", ylabel = \"average topological similarity\")\n",
    "\n",
    "p2 = plot(actualTimeDifferences[:], growingWindows[:,2,1], ribbon = growingWindows[:,2,2], color = [:red theme(:default)], legend=:bottomright , title = \"Topological Similarity over Growing Time Window\", label = [\"dim 1\"],  xlabel = \"window size\", ylabel = \"average topological similarity\")\n",
    "\n",
    "p3 = plot(actualTimeDifferences[:], growingWindows[:,3,1], ribbon = growingWindows[:,3,2], color = [:blue theme(:default)], legend=:bottomright , title = \"Topological Similarity over Growing Time Window\", label = [\"dim 2\"],  xlabel = \"window size\", ylabel = \"average topological similarity\")\n",
    "\n",
    "\n",
    "plot(p1,p2,p3, layout = (3,1), size = (800,800))\n",
    "\n",
    "#savefig(\"growingTimeWindow.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p2 = plot(actualTimeDifferences[:], growingWindows[:,2,1], ribbon = growingWindows[:,2,2], col = rand(38781), legend=:bottomright , title = \"Topological Similarity over Growing Time Window\", label = [\"dim 1\"],  xlabel = \"window size\", ylabel = \"average topological similarity\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(1:38781, birthBoundsForColor[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "birthTimeColors\n",
    "randomColors\n",
    "birthNumber\n",
    "byNeuronType\n",
    "byRegion\n",
    "byxcoor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyplot(size=(800,800), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(dimOneDistances[:,1], dimOneDistances[:,2], markerstrokewidth=0,\n",
    "        m = :blues,\n",
    "        zcolor = randomColors[:,1] # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyplot(size=(800,800), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(dimTwoDistances[:,1], dimTwoDistances[:,2], markerstrokewidth=0,\n",
    "        m = :blues,\n",
    "        zcolor = randomColors[:,1] # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )\n",
    "scatter!(dimOne[4,:,1], dimOne[4,:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets visualize each node's chart one at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for each pair of neurons, find their distance and their top sim\n",
    "dimOne = zeros(279,278,2)\n",
    "dimTwo = zeros(279,278,2)\n",
    " = zeros(279,)\n",
    "\n",
    "\n",
    "\n",
    "for current in 1:279\n",
    "    counter = 0\n",
    "    colors[current] = current\n",
    "for i in 1:279\n",
    "    \n",
    "    if i != current\n",
    "        counter += 1\n",
    "        time = round(abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[current,2]); digits=2) \n",
    "        \n",
    "        dimOne[current,counter,1] = time \n",
    "        dimTwo[current,counter,1] = time\n",
    "        \n",
    "        dimOne[current,counter,2] = oneDimemsionalTopologicalSimilarity[i,current]   \n",
    "        dimTwo[current,counter,2] = twoDimemsionalTopologicalSimilarity[i,current]\n",
    "        \n",
    "    end\n",
    "end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scatter(dimOne[i,:,1], dimOne[i,:,2], markerstrokewidth=0,\n",
    "        m = :blues,\n",
    "        zcolor = current[i] # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "movieColors = zeros(279,38781)\n",
    "\n",
    "for k in 1:279\n",
    "    counter = 0\n",
    "    \n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            counter += 1\n",
    "            if i == k || j == k\n",
    "                movieColors[k,counter] = 1\n",
    "            else\n",
    "                movieColors[k,counter] = 0\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anim = @animate for i in 1:279\n",
    "    \n",
    "    pyplot(size=(800,800), legend=false)  \n",
    "\n",
    "    scatter(dimTwoDistances[:,1], dimTwoDistances[:,2], markerstrokewidth=0, m = :blues, zcolor = movieColors[i,:], title = string(\"Dimension Two: Neuron #\", i,))\n",
    "    scatter!(dimTwo[i,:,1], dimTwo[i,:,2], m = :blues, zcolor = ones(278))\n",
    "\n",
    "end\n",
    "\n",
    "gif(anim, \"dimTwoIndividualsHighlighted.gif\", fps = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 1\n",
    "p1 = scatter([0],[0], markerstrokewidth=0, m = :blues, zcolor = movieColors[i,:], title = string(\"Dimension One: Neuron #\", i,))\n",
    "scatter!(dimOne[2,:,1], dimOne[i,:,2], m = :blues, zcolor = ones(278))\n",
    "scatter!(dimOne[2,:,1], dimOne[2,:,2], m = :blues, zcolor = ones(278))\n",
    "for k in 1:200\n",
    "    scatter!(dimOne[k,:,1], dimOne[k,:,2], m = :blues, zcolor = ones(278))\n",
    "end\n",
    "plot(p1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anim = @animate for i in 1:279\n",
    "    \n",
    "    pyplot(size=(800,800), legend=false)  \n",
    "\n",
    "    p1 = scatter([0], [0], markerstrokewidth=0, m = :blues, zcolor = movieColors[i,:], title = string(\"Dimension One: Neuron #\", i,))\n",
    "    \n",
    "    for k in 1:i\n",
    "        color = zeros(278)\n",
    "        fill!(color, rand())\n",
    "        scatter!(dimOne[i,:,1], dimOne[i,:,2], m = :blues, zcolor = ones(278))\n",
    "    end\n",
    "    \n",
    "    plot(p1)\n",
    "\n",
    "end\n",
    "\n",
    "gif(anim, \"dimOneGrowing.gif\", fps = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anim = @animate for i in 1:279\n",
    "    scatter(dimOne[i,:,1], dimOne[i,:,2], xlims = (0,2400), ylims = (0,1), title = string(\"Neuron #\", i,))\n",
    "end\n",
    "\n",
    "gif(anim, \"demo.gif\", fps = 1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plots(size=(800,800), legend=false)  # set backend and set some session defaults\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyplot(size=(800,800), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(dimOne[now,:,1], dimOne[now,:,2], markerstrokewidth=0,\n",
    "        m = :blues,\n",
    "        zcolor = rand(278) # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyplot(size=(800,800), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(dimTwo[now,:,1], dimTwo[now,:,2], markerstrokewidth=0,\n",
    "        m = :blues, zcolor = rand(278) # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plotlyjs(size=(400,200), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(rand(30), alpha = 1, markerstrokewidth=0,\n",
    "        m= :reds,\n",
    "        zcolor = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30] # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scatter(fill(randn(10), 6), fill(randn(10), 6), framestyle=[:box :semi :origin :zerolines :grid :none], title=[\":box\" \":semi\" \":origin\" \":zerolines\" \":grid\" \":none\"], color=permutedims(1:6), layout=6, label=\"\", markerstrokewidth=0, ticks=-2:2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try to categorize unfriendly nodes by adjacent location on matrix deltas\n",
    "topologicalSimilarityDeltas  = zeros(279,276) #nnodes x ndeltas size\n",
    "reductThisIntoThat = zeros(279,278)\n",
    "\n",
    "#iterates over the nodes\n",
    "for i in 1:279\n",
    "    adjustment = 0\n",
    "    \n",
    "    #iterate over all pairs of nodes including i\n",
    "    for j in 1:278\n",
    "           \n",
    "        #to avoid having to 'switch directions,' take the max of the symmetric pair\n",
    "        first = maximum([topologicalSimilarity[i,j], topologicalSimilarity[j,i]])\n",
    "        second = maximum([topologicalSimilarity[i,j+1], topologicalSimilarity[j+1,i]])\n",
    "        delta = abs(first - second)\n",
    "        reductThisIntoThat[i,j] = delta\n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "#remove deltas in which (i,i) contributed\n",
    "for i in 2:278\n",
    "    counter = 0\n",
    "    \n",
    "    for j in 1:278\n",
    "        if j < i - 1\n",
    "            counter += 1\n",
    "            topologicalSimilarityDeltas[i,counter] = reductThisIntoThat[i,j]\n",
    "        end\n",
    "        \n",
    "        if j > i \n",
    "            counter += 1\n",
    "            topologicalSimilarityDeltas[i,counter] = reductThisIntoThat[i,j]\n",
    "        end\n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "#the above construction is problematic for nodes 1 and 279, so we store them in a separate matrix\n",
    "for k in 1:276\n",
    "    topologicalSimilarityDeltas[1,k] = reductThisIntoThat[1,k+1]\n",
    "    topologicalSimilarityDeltas[279,k] = reductThisIntoThat[279,k]\n",
    "end\n",
    "\n",
    "\n",
    "println(\"output: topologicalSimilarityDeltas and deltasForExtrema\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute variance and mean for each node's set of deltas\n",
    "nodeDeltaMeanAndVaraince = zeros(279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    nodeDeltaMeanAndVaraince[i,1] = mean(topologicalSimilarityDeltas[i,:])\n",
    "    nodeDeltaMeanAndVaraince[i,2] = var(topologicalSimilarityDeltas[i,:])\n",
    "end\n",
    "\n",
    "println(\"output: nodeDeltaMeanAndVaraince\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots delta mean and delta varaince\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 2)\n",
    "plotArray[1] = nodeDeltaMeanAndVaraince[:,1]\n",
    "plotArray[2] = nodeDeltaMeanAndVaraince[:,2]\n",
    "\n",
    "\n",
    "\n",
    "violin([\"Delta Mean\" \"Delta Variance\" ],plotArray,leg=false, color = theme(:dark))\n",
    "boxplot!([\"Delta Mean\" \"Delta Variance\"],plotArray,leg=false, color = theme(:dark))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot distance/similarity pairs\n",
    "scatter(nodeDeltaMeanAndVaraince[:,1], nodeDeltaMeanAndVaraince[:,2], legend = false, title = \"Delta Mean vs Delta Variance\", xlabel = \"delta mean\", ylabel = \"delta variance\", color = :grey)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute correlation between distance and topological similarity\n",
    "cor(nodeDeltaMeanAndVaraince[:,1], nodeDeltaMeanAndVaraince[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newTopSim = topologicalSimilarity[:,:]\n",
    "newDeltaMean = nodeDeltaMeanAndVaraince[:,2]\n",
    "\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && newDeltaMean[i - movement] < newDeltaMean[i-movement - 1]\n",
    "        shadowCache = newDeltaMean[i - movement]\n",
    "        newDeltaMean[i - movement] = newDeltaMean[i - movement - 1]\n",
    "        newDeltaMean[i - movement - 1] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[i - movement, :]\n",
    "        newTopSim[i - movement, :] = newTopSim[i - movement - 1, :]\n",
    "        newTopSim[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[:, i - movement]\n",
    "        newTopSim[:, i - movement] = newTopSim[:, i - movement - 1]\n",
    "        newTopSim[:, i - movement - 1] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if newTopSim[j,i] > 0\n",
    "            newTopSim[i,j] = newTopSim[j,i]\n",
    "            newTopSim[j,i] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "heatmap(newTopSim, yflip = false, aspect_ratio=:equal, title = \"One-Two Dimensional Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newDeltaMean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do summed similarity calculations\n",
    "summedSimilarity = zeros(279)\n",
    "\n",
    "for i in 1:279\n",
    "    summedSimilarity[i] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[:,i])\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots delta mean and delta varaince\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 1)\n",
    "plotArray[1] = summedSimilarity\n",
    "\n",
    "\n",
    "\n",
    "violin([\"Summed Similarity\" ],plotArray,leg=false, color = theme(:dark))\n",
    "boxplot!([\"Summed Similarity\"],plotArray,leg=false, color = theme(:dark))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newTopSim = topologicalSimilarity[:,:]\n",
    "newDeltaMean = summedSimilarity[:]\n",
    "\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && newDeltaMean[i - movement] < newDeltaMean[i-movement - 1]\n",
    "        shadowCache = newDeltaMean[i - movement]\n",
    "        newDeltaMean[i - movement] = newDeltaMean[i - movement - 1]\n",
    "        newDeltaMean[i - movement - 1] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[i - movement, :]\n",
    "        newTopSim[i - movement, :] = newTopSim[i - movement - 1, :]\n",
    "        newTopSim[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[:, i - movement]\n",
    "        newTopSim[:, i - movement] = newTopSim[:, i - movement - 1]\n",
    "        newTopSim[:, i - movement - 1] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if newTopSim[j,i] > 0\n",
    "            newTopSim[i,j] = newTopSim[j,i]\n",
    "            newTopSim[j,i] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "heatmap(newTopSim, yflip = false, aspect_ratio=:equal, title = \"One-Two Dimensional Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#end of reorderability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#beginning of null models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate 1000 erdos renyi \n",
    "nTrials = 1000\n",
    "prob = 2287 / 38781\n",
    "\n",
    "erdosRenyiTrails = zeros(279,279,nTrials)\n",
    "fill!(erdosRenyiTrails, 3000)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    \n",
    "    for i in 1:279\n",
    "       erdosRenyiTrails[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "    end\n",
    "    \n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            if rand() <= prob\n",
    "                erdosRenyiTrails[i,j,k] = labelsTimesPositionsType[i,2]\n",
    "                erdosRenyiTrails[j,i,k] = labelsTimesPositionsType[i,2]\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "print(\"output: erdosRenyiTrials, prob, nTrials\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine persistent homology of erdos renyi\n",
    "\n",
    "begin\n",
    "    randomCurves = pmap(i -> computeBarcodesNullModels(erdosRenyiTrails[:,:,i], times), 1:nTrials)\n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "erdosRenyiBettiCurves = zeros(numValues,3,nTrials)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    erdosRenyiBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "            \n",
    "    \n",
    "println(\"output: erdosRenyiBettiCurves, randomCurves\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of nTrials copies\n",
    "bettiMeans = zeros(numValues,3)\n",
    "bettiStd = zeros(numValues,3)\n",
    "\n",
    "for i in 1:numValues\n",
    "    bettiMeans[i,1] = mean(erdosRenyiBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(erdosRenyiBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(erdosRenyiBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(erdosRenyiBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(erdosRenyiBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(erdosRenyiBettiCurves[i,3,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Erdos Renyi\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0,15))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "\n",
    "plot(p1,p2, layout = (2,1), size = (800,800))\n",
    "\n",
    "#savefig(\"erdosRenyiTrials.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "computeRates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reorderability of erdos renyi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute reorderability for nTrials trials\n",
    "nTrials = 5\n",
    "\n",
    "\n",
    "begin\n",
    "    trials = pmap(i -> nullModelTopologicalSimilarityStorage(erdosRenyiTrails[:,:,i], i, \"Erdos Renyi\"), 1:nTrials)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute reorderability for nTrials trials\n",
    "nTrials = 5\n",
    "\n",
    "ERdimZeroTrials = zeros(279,279,nTrials)\n",
    "ERdimOneTrials = zeros(279,279,nTrials)\n",
    "ERdimTwoTrials = zeros(279,279,nTrials)\n",
    "\n",
    "begin\n",
    "    trials = pmap(i -> nullModelTopologicalSimilarityStorage(erdosRenyiTrails[:,:,i], i, \"Erdos Renyi\"), 1:nTrials)\n",
    "end\n",
    "\n",
    "\n",
    "#time to unpack this stuff\n",
    "for k in 1:nTrials\n",
    "    ERdimZeroTrials[:,:,k] = trials[k][1]\n",
    "    ERdimOneTrials[:,:,k] = trials[k][2]\n",
    "    ERdimTwoTrials[:,:,k] = trials[k][3]\n",
    "end\n",
    "\n",
    "aveDimZeroTopSim = zeros(279,279)\n",
    "aveDimOneTopSim = zeros(279,279)\n",
    "aveDimTwoTopSim = zeros(279,279)\n",
    "\n",
    "stdDimZeroTopSim = zeros(279,279)\n",
    "stdDimOneTopSim = zeros(279,279)\n",
    "stdDimTwoTopSim = zeros(279,279)\n",
    "\n",
    "aveDimZeroBND = zeros(279,279)\n",
    "aveDimOneBND = zeros(279,279)\n",
    "aveDimTwoBND = zeros(279,279)\n",
    "\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        aveDimZeroTopSim[i,j] = mean(ERdimZeroTrials[i,j,:])\n",
    "        aveDimOneTopSim[i,j] = mean(ERdimOneTrials[i,j,:])\n",
    "        aveDimTwoTopSim[i,j] = mean(ERdimTwoTrials[i,j,:])\n",
    "        \n",
    "        stdDimZeroTopSim[i,j] = std(ERdimZeroTrials[i,j,:])\n",
    "        stdDimOneTopSim[i,j] = std(ERdimOneTrials[i,j,:])\n",
    "        stdDimTwoTopSim[i,j] = std(ERdimTwoTrials[i,j,:])\n",
    "        \n",
    "        timeDiff = abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2])\n",
    "        \n",
    "        if timeDiff == 0\n",
    "            aveDimZeroBND[i,j] = 0\n",
    "            aveDimOneBND[i,j] = 0\n",
    "            aveDimTwoBND[i,j] = 0\n",
    "        else\n",
    "            aveDimZeroBND[i,j] = timeDiff * (1 - aveDimZeroTopSim[i,j])\n",
    "            aveDimOneBND[i,j] = timeDiff * (1 - aveDimOneTopSim[i,j])\n",
    "            aveDimTwoBND[i,j] = timeDiff * (1 - aveDimTwoTopSim[i,j])\n",
    "        end\n",
    "        \n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save these cause it takes a big while to compute these things\n",
    "\n",
    "CSV.write(\"OneErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,1]), writeheader = false)\n",
    "CSV.write(\"TwoErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,2]), writeheader = false)\n",
    "CSV.write(\"ThreeErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,3]), writeheader = false)\n",
    "CSV.write(\"FourErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,4]), writeheader = false)\n",
    "CSV.write(\"FiveErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,5]), writeheader = false)\n",
    "\n",
    "CSV.write(\"OneErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,1]), writeheader = false)\n",
    "CSV.write(\"TwoErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,2]), writeheader = false)\n",
    "CSV.write(\"ThreeErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,3]), writeheader = false)\n",
    "CSV.write(\"FourErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,4]), writeheader = false)\n",
    "CSV.write(\"FiveErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,5]), writeheader = false)\n",
    "\n",
    "CSV.write(\"OneErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,1]), writeheader = false)\n",
    "CSV.write(\"TwoErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,2]), writeheader = false)\n",
    "CSV.write(\"ThreeErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,3]), writeheader = false)\n",
    "CSV.write(\"FourErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,4]), writeheader = false)\n",
    "CSV.write(\"FiveErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,5]), writeheader = false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimZeroTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Erdos Renyi\", color = :tempo)\n",
    "p2 = heatmap(aveDimZeroBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Erdos Renyi\", color = :grays)\n",
    "p3 = heatmap(stdDimZeroTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Erdos Renyi\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionZeroSimilarityErdosRenyi.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimOneTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension One Erdos Renyi\", color = :tempo)\n",
    "p2 = heatmap(aveDimOneBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Erdos Renyi\", color = :grays)\n",
    "p3 = heatmap(stdDimOneTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension One Erdos Renyi\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionOneSimilarityErdosRenyi.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimTwoTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Erdos Renyi\", color = :tempo)\n",
    "p2 = heatmap(aveDimTwoBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Erdos Renyi\", color = :grays)\n",
    "p3 = heatmap(stdDimTwoTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Erdos Renyi\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionTwoSimilarityErdosRenyi.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimzero, dimone, dimtwo = nullModelTopologicalSimilarity(birthSortedAdjacencyMatrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap(dimzero, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap(dimone, yflip = true, aspect_ratio=:equal, title = \"Dimension One Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap(dimtwo, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cleanup\n",
    "nTrials = nothing\n",
    "prob = nothing\n",
    "erdosRenyiTrails = nothing\n",
    "randomCurves = nothing\n",
    "erdosRenyiBettiCurves = nothing\n",
    "bettiMeans = nothing\n",
    "bettiStd = nothing\n",
    "ERdimZeroTrials = nothing\n",
    "ERdimOneTrials = nothing\n",
    "ERdimTwoTrials = nothing\n",
    "trials = nothing\n",
    "aveDimZeroTopSim = nothing\n",
    "aveDimOneTopSim = nothing\n",
    "aveDimTwoTopSim = nothing\n",
    "stdDimZeroTopSim = nothing\n",
    "stdDimOneTopSim = nothing\n",
    "stdDimTwoTopSim = nothing\n",
    "aveDimZeroBND = nothing\n",
    "aveDimOneBND = nothing\n",
    "aveDimTwoBND = nothing\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "p3 = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: spatialGrowthTrials, threshold, nTrials"
     ]
    }
   ],
   "source": [
    "#generate 1000 spatial growth\n",
    "nTrials = 1000\n",
    "threshold = 2287\n",
    "spatialGrowthTrials = zeros(279,279,nTrials)\n",
    "fill!(spatialGrowthTrials, 3000)\n",
    "\n",
    "\n",
    "for k in 1:nTrials\n",
    "    \n",
    "    for i in 1:279\n",
    "       spatialGrowthTrials[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "    end\n",
    "    \n",
    "    #spawn 279 random points\n",
    "    points = zeros(279,2)\n",
    "    for i in 1:279\n",
    "        points[i,:] = [rand(0:1000), rand(0:40)]\n",
    "    end\n",
    "    \n",
    "    #find distances between all points\n",
    "    fullDistances = zeros(38781, 3)\n",
    "    counter = 0\n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            counter += 1\n",
    "            \n",
    "            a = points[i,1]\n",
    "            b = points[i,2]\n",
    "            c = points[j,1]\n",
    "            d = points[j,2]\n",
    "            \n",
    "            fullDistances[counter,:] = [evaluate(Euclidean(),[a,b], [c,d]), i, j]\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #sort distances\n",
    "    fullDistances = sortslices(fullDistances, dims = 1)\n",
    "    \n",
    "    #add connections between closest 2287 connections\n",
    "    for l in 1:2287\n",
    "        i = trunc(Int, fullDistances[l,2])\n",
    "        j = trunc(Int, fullDistances[l,3])\n",
    "\n",
    "        spatialGrowthTrials[i,j,k] = labelsTimesPositionsType[i,2]\n",
    "        spatialGrowthTrials[j,i,k] = labelsTimesPositionsType[i,2]      \n",
    "    end\n",
    "            \n",
    "end\n",
    "print(\"output: spatialGrowthTrials, threshold, nTrials\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine persistent homology of spatial growth\n",
    "\n",
    "begin\n",
    "randomCurves = pmap(i -> computeBarcodesNullModels(spatialGrowthTrials[:,:,i], times), 1:nTrials)\n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "spatialGrowthBettiCurves = zeros(numValues,3,nTrials)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    spatialGrowthBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "            \n",
    "    \n",
    "println(\"output: spatialGrowthBettiCurves\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of nTrials copies\n",
    "bettiMeans = zeros(numValues,3)\n",
    "bettiStd = zeros(numValues,3)\n",
    "\n",
    "for i in 1:numValues\n",
    "    bettiMeans[i,1] = mean(spatialGrowthBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(spatialGrowthBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(spatialGrowthBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(spatialGrowthBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(spatialGrowthBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(spatialGrowthBettiCurves[i,3,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Spatial Growth\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0, 20))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "\n",
    "plot(p1,p2, layout = (2,1), size = (800,800))\n",
    "\n",
    "#savefig(\"spatialGrowthTrials.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute reorderability for nTrials trials\n",
    "nTrials = 5\n",
    "\n",
    "\n",
    "begin\n",
    "    trials = pmap(i -> nullModelTopologicalSimilarityStorage(spatialGrowthTrials[:,:,i], i, \"Spatial Growth\"), 1:nTrials)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimZeroTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Spatial Growth\", color = :tempo)\n",
    "p2 = heatmap(aveDimZeroBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Spatial Growth\", color = :grays)\n",
    "p3 = heatmap(stdDimZeroTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Spatial Growth\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionZeroSimilaritySpatialGrowth.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimOneTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension One Spatial Growth\", color = :tempo)\n",
    "p2 = heatmap(aveDimOneBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Spatial Growth\", color = :grays)\n",
    "p3 = heatmap(stdDimOneTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension One Spatial Growth\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionOneSimilaritySpatialGrowth.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimTwoTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Spatial Growth\", color = :tempo)\n",
    "p2 = heatmap(aveDimTwoBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Spatial Growth\", color = :grays)\n",
    "p3 = heatmap(stdDimTwoTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Spatial Growth\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionTwoSimilaritySpatialGrowth.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cleanup\n",
    "nTrials = nothing\n",
    "prob = nothing\n",
    "spatialGrowthTrials = nothing\n",
    "randomCurves = nothing\n",
    "spatialGrowthBettiCurves = nothing\n",
    "bettiMeans = nothing\n",
    "bettiStd = nothing\n",
    "SGdimZeroTrials = nothing\n",
    "SGdimOneTrials = nothing\n",
    "SGdimTwoTrials = nothing\n",
    "trials = nothing\n",
    "aveDimZeroTopSim = nothing\n",
    "aveDimOneTopSim = nothing\n",
    "aveDimTwoTopSim = nothing\n",
    "stdDimZeroTopSim = nothing\n",
    "stdDimOneTopSim = nothing\n",
    "stdDimTwoTopSim = nothing\n",
    "aveDimZeroBND = nothing\n",
    "aveDimOneBND = nothing\n",
    "aveDimTwoBND = nothing\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "p3 = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293 294 295 296 297 298 299 300 301 302 303 304 305 306 307 308 309 310 311 312 313 314 315 316 317 318 319 320 321 322 323 324 325 326 327 328 329 330 331 332 333 334 335 336 337 338 339 340 341 342 343 344 345 346 347 348 349 350 351 352 353 354 355 356 357 358 359 360 361 362 363 364 365 366 367 368 369 370 371 372 373 374 375 376 377 378 379 380 381 382 383 384 385 386 387 388 389 390 391 392 393 394 395 396 397 398 399 400 401 402 403 404 405 406 407 408 409 410 411 412 413 414 415 416 417 418 419 420 421 422 423 424 425 426 427 428 429 430 431 432 433 434 435 436 437 438 439 440 441 442 443 444 445 446 447 448 449 450 451 452 453 454 455 456 457 458 459 460 461 462 463 464 465 466 467 468 469 470 471 472 473 474 475 476 477 478 479 480 481 482 483 484 485 486 487 488 489 490 491 492 493 494 495 496 497 498 499 500 501 502 503 504 505 506 507 508 509 510 511 512 513 514 515 516 517 518 519 520 521 522 523 524 525 526 527 528 529 530 531 532 533 534 535 536 537 538 539 540 541 542 543 544 545 546 547 548 549 550 551 552 553 554 555 556 557 558 559 560 561 562 563 564 565 566 567 568 569 570 571 572 573 574 575 576 577 578 579 580 581 582 583 584 585 586 587 588 589 590 591 592 593 594 595 596 597 598 599 600 601 602 603 604 605 606 607 608 609 610 611 612 613 614 615 616 617 618 619 620 621 622 623 624 625 626 627 628 629 630 631 632 633 634 635 636 637 638 639 640 641 642 643 644 645 646 647 648 649 650 651 652 653 654 655 656 657 658 659 660 661 662 663 664 665 666 667 668 669 670 671 672 673 674 675 676 677 678 679 680 681 682 683 684 685 686 687 688 689 690 691 692 693 694 695 696 697 698 699 700 701 702 703 704 705 706 707 708 709 710 711 712 713 714 715 716 717 718 719 720 721 722 723 724 725 726 727 728 729 730 731 732 733 734 735 736 737 738 739 740 741 742 743 744 745 746 747 748 749 750 751 752 753 754 755 756 757 758 759 760 761 762 763 764 765 766 767 768 769 770 771 772 773 774 775 776 777 778 779 780 781 782 783 784 785 786 787 788 789 790 791 792 793 794 795 796 797 798 799 800 801 802 803 804 805 806 807 808 809 810 811 812 813 814 815 816 817 818 819 820 821 822 823 824 825 826 827 828 829 830 831 832 833 834 835 836 837 838 839 840 841 842 843 844 845 846 847 848 849 850 851 852 853 854 855 856 857 858 859 860 861 862 863 864 865 866 867 868 869 870 871 872 873 874 875 876 877 878 879 880 881 882 883 884 885 886 887 888 889 890 891 892 893 894 895 896 897 898 899 900 901 902 903 904 905 906 907 908 909 910 911 912 913 914 915 916 917 918 919 920 921 922 923 924 925 926 927 928 929 930 931 932 933 934 935 936 937 938 939 940 941 942 943 944 945 946 947 948 949 950 951 952 953 954 955 956 957 958 959 960 961 962 963 964 965 966 967 968 969 970 971 972 973 974 975 976 977 978 979 980 981 982 983 984 985 986 987 988 989 990 991 992 993 994 995 996 997 998 999 1000 output: rewiredTrials, threshold, nTrials"
     ]
    }
   ],
   "source": [
    "#generate 1000 rewiring trials\n",
    "nTrials = 1000\n",
    "threshold = 2287\n",
    "rewiredTrials = zeros(279,279,nTrials)\n",
    "fill!(rewiredTrials, 3000)\n",
    "\n",
    "#find set of all edges in the graph\n",
    "edges = Array{Int64, 2}(undef, 2287, 2)\n",
    "edgeCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "            edgeCounter += 1\n",
    "            edges[edgeCounter,:] = [i,j]\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "for k in 1:nTrials\n",
    "    print(k, \" \")\n",
    "    for i in 1:279\n",
    "       rewiredTrials[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "    end\n",
    "    \n",
    "    #perform 22870 random rewirings\n",
    "    for i in 1:22870\n",
    "        \n",
    "        redundant = true\n",
    "        \n",
    "        while redundant\n",
    "            redundant = false\n",
    "            \n",
    "            initial = rand(1:2287)\n",
    "            terminal = rand(1:2287)\n",
    "            parity = rand(1:2)\n",
    "        \n",
    "           while (initial == terminal)\n",
    "                terminal = rand(1:2287)\n",
    "            end\n",
    "           \n",
    "            if parity == 0\n",
    "                a = edges[initial, 1]\n",
    "                d = edges[initial, 2]\n",
    "                c = edges[terminal, 1]\n",
    "                b = edges[terminal, 2]\n",
    "            else\n",
    "                a = edges[initial, 1]\n",
    "                c = edges[initial, 2]\n",
    "                b = edges[terminal, 1]\n",
    "                d = edges[terminal, 2]\n",
    "            end\n",
    "            \n",
    "            for i in 1:2287\n",
    "                if edges[i,:] == [initial, terminal] || edges[i,:] == [terminal, initial]\n",
    "                    redundant = true\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            if !redundant\n",
    "               edges[initial, 1] = a\n",
    "               edges[initial, 2] = b\n",
    "               edges[terminal, 1] = c\n",
    "               edges[terminal, 2] = d\n",
    "            end\n",
    "        end\n",
    "            \n",
    "    end\n",
    "    \n",
    "    #add edges\n",
    "    for i in 1:2287\n",
    "        max = maximum(edges[i,:])\n",
    "        \n",
    "        rewiredTrials[edges[i,1],edges[i,2],k] = labelsTimesPositionsType[max,2]\n",
    "        rewiredTrials[edges[i,2],edges[i,1],k] = labelsTimesPositionsType[max,2] \n",
    "    end\n",
    "end\n",
    "print(\"output: rewiredTrials, threshold, nTrials\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: rewiredBettiCurves\n"
     ]
    }
   ],
   "source": [
    "#determine persistent homology of spatial growth\n",
    "\n",
    "begin\n",
    "randomCurves = pmap(i -> computeBarcodesNullModels(rewiredTrials[:,:,i], times), 1:nTrials)\n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "rewiredBettiCurves = zeros(numValues,3,nTrials)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    rewiredBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "            \n",
    "    \n",
    "println(\"output: rewiredBettiCurves\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: bettiMeans, bettiStd"
     ]
    }
   ],
   "source": [
    "#computes average and standard deviation of nTrials copies\n",
    "bettiMeans = zeros(numValues,3)\n",
    "bettiStd = zeros(numValues,3)\n",
    "\n",
    "for i in 1:numValues\n",
    "    bettiMeans[i,1] = mean(rewiredBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(rewiredBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(rewiredBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(rewiredBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(rewiredBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(rewiredBettiCurves[i,3,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"800\" height=\"800\" viewBox=\"0 0 3200 3200\">\n",
       "<defs>\n",
       "  <clipPath id=\"clip6800\">\n",
       "    <rect x=\"0\" y=\"0\" width=\"3200\" height=\"3200\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip6800)\" d=\"\n",
       "M0 3200 L3200 3200 L3200 0 L0 0  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip6801\">\n",
       "    <rect x=\"640\" y=\"320\" width=\"2241\" height=\"2241\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip6800)\" d=\"\n",
       "M229.135 1462.83 L3152.76 1462.83 L3152.76 121.675 L229.135 121.675  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip6802\">\n",
       "    <rect x=\"229\" y=\"121\" width=\"2925\" height=\"1342\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  640.626,1462.83 640.626,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1306.32,1462.83 1306.32,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1972.02,1462.83 1972.02,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  2637.71,1462.83 2637.71,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,1424.15 3152.76,1424.15 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,1227.02 3152.76,1227.02 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,1029.9 3152.76,1029.9 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,832.777 3152.76,832.777 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,635.654 3152.76,635.654 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,438.53 3152.76,438.53 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,241.406 3152.76,241.406 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1462.83 3152.76,1462.83 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1462.83 229.135,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  640.626,1462.83 640.626,1442.72 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1306.32,1462.83 1306.32,1442.72 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1972.02,1462.83 1972.02,1442.72 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2637.71,1462.83 2637.71,1442.72 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1424.15 272.99,1424.15 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1227.02 272.99,1227.02 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1029.9 272.99,1029.9 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,832.777 272.99,832.777 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,635.654 272.99,635.654 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,438.53 272.99,438.53 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,241.406 272.99,241.406 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 640.626, 1520.83)\" x=\"640.626\" y=\"1520.83\">500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1306.32, 1520.83)\" x=\"1306.32\" y=\"1520.83\">1000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1972.02, 1520.83)\" x=\"1972.02\" y=\"1520.83\">1500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 2637.71, 1520.83)\" x=\"2637.71\" y=\"1520.83\">2000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 1441.65)\" x=\"197.135\" y=\"1441.65\">0</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 1244.52)\" x=\"197.135\" y=\"1244.52\">100</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 1047.4)\" x=\"197.135\" y=\"1047.4\">200</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 850.277)\" x=\"197.135\" y=\"850.277\">300</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 653.154)\" x=\"197.135\" y=\"653.154\">400</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 456.03)\" x=\"197.135\" y=\"456.03\">500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 258.906)\" x=\"197.135\" y=\"258.906\">600</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:82px; text-anchor:middle;\" transform=\"rotate(0, 1690.95, 72)\" x=\"1690.95\" y=\"72\">Identical Degree Distribution</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(0, 1690.95, 1604.7)\" x=\"1690.95\" y=\"1604.7\">time</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(-90, 73.2861, 792.255)\" x=\"73.2861\" y=\"792.255\">betti number</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip6802)\" d=\"\n",
       "M311.879 1424.15 L345.071 1424.15 L362.379 1424.15 L365.654 1424.21 L365.894 1424.21 L366.812 1424.23 L367.691 1424.28 L368.237 1424.3 L368.49 1424.34 L369.355 1424.53 \n",
       "  L369.422 1424.58 L370.58 1424.59 L371.765 1424.62 L372.937 1424.73 L373.176 1424.74 L374.108 1424.7 L374.348 1424.49 L375.293 1423.36 L375.6 1422.48 L376.691 1419.44 \n",
       "  L376.851 1417.15 L376.864 1416.08 L377.637 1413.79 L379.101 1413.15 L379.341 1406.88 L379.367 1404.69 L381.005 1401.4 L381.378 1398.96 L381.844 1396.54 L382.123 1395.2 \n",
       "  L382.336 1393.56 L382.549 1387.69 L382.563 1377.22 L382.856 1374.03 L383.082 1369.6 L383.721 1352.72 L384.333 1350.8 L384.893 1349.64 L385.652 1346.74 L386.836 1327.45 \n",
       "  L387.862 1310.45 L388.075 1302.8 L389.326 1298.33 L389.379 1298.73 L389.579 1298.36 L390.671 1297.69 L391.962 1290.42 L393.067 1293 L393.187 1283.95 L394.119 1280.22 \n",
       "  L394.266 1275.99 L394.319 1267.04 L395.251 1261.35 L395.371 1257.67 L395.437 1248.01 L395.57 1240.81 L396.609 1231.01 L396.622 1226.03 L397.594 1220.59 L398.952 1224.52 \n",
       "  L399.312 1214.41 L400.563 1211.71 L400.976 1206.89 L401.815 1201.73 L402.454 1196.48 L403.226 1187.74 L403.625 1183.7 L404.797 1178.84 L405.556 1173.36 L406.807 1168.04 \n",
       "  L408.045 1165.13 L408.259 1169.17 L408.511 1162.14 L409.337 1154.7 L410.509 1149.33 L410.655 1138.64 L412.625 1138.53 L414.17 1133.2 L416.793 1130.11 L417.552 1127.71 \n",
       "  L420.028 1122.34 L420.534 1116.99 L421.2 1113.88 L421.785 1108.63 L421.852 1104.31 L423.596 1100.59 L424.275 1092.26 L428.123 1086.59 L430.466 1076.47 L437.509 1071.55 \n",
       "  L441.743 1067.43 L444.552 1061.72 L452.075 1055.59 L452.221 1049.39 L453.952 1039.89 L454.737 1033.89 L455.989 1030.37 L498.7 1024.56 L499.033 1019.41 L502.574 1007.88 \n",
       "  L506.542 1002.17 L510.309 996.012 L512.653 990.79 L514.996 984.759 L517.353 978.626 L519.696 973.789 L520.042 967.713 L523.45 959.578 L523.517 953.997 L526.619 947.723 \n",
       "  L526.752 941.238 L527.205 935.561 L527.91 929.27 L530.507 922.963 L531.439 915.941 L532.943 910.389 L533.649 905.895 L534.82 900.913 L535.113 887.978 L536.618 881.457 \n",
       "  L538.335 874.801 L539.653 868.183 L542.849 861.085 L543.022 854.192 L544.193 848.089 L545.352 841.989 L546.603 836.822 L547.841 830.398 L559.145 812.251 L559.331 805.656 \n",
       "  L561.022 801.29 L564.99 789.796 L572.299 799.226 L572.765 786.87 L574.056 802.332 L575.308 795.294 L576.28 788.387 L578.623 781.5 L634.288 775.69 L636.166 768.878 \n",
       "  L644.154 761.236 L1669.66 754.758 L1679.11 749.188 L1829.42 743.192 L1838.88 730.08 L1848.33 722.866 L1856.45 710.916 L1919.16 698.066 L2083.19 691.138 L2088.25 669.405 \n",
       "  L2092.64 662.392 L2105.15 648.296 L2109.81 635.228 L2122.06 629.992 L2138.04 627.401 L2138.97 606.866 L2139.64 599.683 L2155.88 542.895 L2172.79 537.97 L2189.7 516.351 \n",
       "  L2206.61 492.118 L2223.51 478.314 L2240.42 445.858 L2257.33 389.038 L2274.24 366.184 L2291.15 350.836 L2308.06 323.404 L2460.24 317.861 L2494.05 309.924 L2644.77 292.959 \n",
       "  L2771.38 271.491 L3070.01 254.459 L3070.01 159.632 L2771.38 176.865 L2644.77 198.74 L2494.05 216.536 L2460.24 225.114 L2308.06 231.461 L2291.15 259.508 L2274.24 275.711 \n",
       "  L2257.33 298.744 L2240.42 356.398 L2223.51 389.26 L2206.61 403.61 L2189.7 428.693 L2172.79 450.303 L2155.88 455.377 L2139.64 515.869 L2138.97 523.636 L2138.04 545.006 \n",
       "  L2122.06 547.563 L2109.81 552.724 L2105.15 566.141 L2092.64 581.188 L2088.25 588.34 L2083.19 610.439 L1919.16 617.728 L1856.45 631.001 L1848.33 643.285 L1838.88 650.516 \n",
       "  L1829.42 664.521 L1679.11 670.951 L1669.66 676.499 L644.154 683.394 L636.166 691.163 L634.288 698.244 L578.623 704.431 L576.28 711.505 L575.308 718.908 L574.056 726.02 \n",
       "  L572.765 709.386 L572.299 722.432 L564.99 712.8 L561.022 725.19 L559.331 729.954 L559.145 737.249 L547.841 756.654 L546.603 763.354 L545.352 768.978 L544.193 775.328 \n",
       "  L543.022 781.525 L542.849 788.569 L539.653 795.987 L538.335 802.703 L536.618 809.664 L535.113 816.342 L534.82 829.905 L533.649 835.394 L532.943 840.661 L531.439 846.586 \n",
       "  L530.507 853.516 L527.91 860.314 L527.205 867.105 L526.752 873.559 L526.619 880.458 L523.517 886.954 L523.45 892.865 L520.042 902.054 L519.696 908.498 L517.353 913.624 \n",
       "  L514.996 919.871 L512.653 926.207 L510.309 931.976 L506.542 938.46 L502.574 945.097 L499.033 957.19 L498.7 962.627 L455.989 968.933 L454.737 972.605 L453.952 978.586 \n",
       "  L452.221 988.425 L452.075 995.072 L444.552 1001.4 L441.743 1007.41 L437.509 1011.82 L430.466 1017.29 L428.123 1027.7 L424.275 1033.56 L423.596 1042.11 L421.852 1045.9 \n",
       "  L421.785 1050.85 L421.2 1056.67 L420.534 1059.98 L420.028 1065.63 L417.552 1071.36 L416.793 1074.23 L414.17 1078.12 L412.625 1084.13 L410.655 1083.98 L410.509 1095.52 \n",
       "  L409.337 1101.12 L408.511 1109.6 L408.259 1117.23 L408.045 1113.11 L406.807 1116.58 L405.556 1122.66 L404.797 1128.89 L403.625 1134.1 L403.226 1138.53 L402.454 1147.59 \n",
       "  L401.815 1153.64 L400.976 1159.64 L400.563 1165.1 L399.312 1168 L398.952 1178.62 L397.594 1176.84 L396.622 1183.16 L396.609 1188.69 L395.57 1199.21 L395.437 1206.98 \n",
       "  L395.371 1217.41 L395.251 1221.11 L394.319 1227.42 L394.266 1237.12 L394.119 1241.79 L393.187 1245.89 L393.067 1256.21 L391.962 1254.98 L390.671 1263.16 L389.579 1267.06 \n",
       "  L389.379 1267.49 L389.326 1269.82 L388.075 1274.95 L387.862 1283.19 L386.836 1301.69 L385.652 1323.26 L384.893 1326.23 L384.333 1327.5 L383.721 1329.86 L383.082 1348.32 \n",
       "  L382.856 1353.33 L382.563 1357.21 L382.549 1369.39 L382.336 1376.42 L382.123 1378.37 L381.844 1380.05 L381.378 1383.01 L381.005 1386.18 L379.367 1390.59 L379.341 1393.52 \n",
       "  L379.101 1401.54 L377.637 1402.59 L376.864 1405.52 L376.851 1406.97 L376.691 1410.47 L375.6 1415.34 L375.293 1417.01 L374.348 1419.88 L374.108 1421.22 L373.176 1422.22 \n",
       "  L372.937 1422.39 L371.765 1423.28 L370.58 1423.41 L369.422 1423.46 L369.355 1423.55 L368.49 1423.91 L368.237 1423.97 L367.691 1424 L366.812 1424.06 L365.894 1424.08 \n",
       "  L365.654 1424.08 L362.379 1424.15 L345.071 1424.15 L311.879 1424.15  Z\n",
       "  \" fill=\"#ff0000\" fill-rule=\"evenodd\" fill-opacity=\"0.5\"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#ff0000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  311.879,1424.15 345.071,1424.15 362.379,1424.15 365.654,1424.15 365.894,1424.15 366.812,1424.14 367.691,1424.14 368.237,1424.13 368.49,1424.12 369.355,1424.04 \n",
       "  369.422,1424.02 370.58,1424 371.765,1423.95 372.937,1423.56 373.176,1423.48 374.108,1422.96 374.348,1422.18 375.293,1420.19 375.6,1418.91 376.691,1414.95 \n",
       "  376.851,1412.06 376.864,1410.8 377.637,1408.19 379.101,1407.34 379.341,1400.2 379.367,1397.64 381.005,1393.79 381.378,1390.99 381.844,1388.29 382.123,1386.78 \n",
       "  382.336,1384.99 382.549,1378.54 382.563,1367.21 382.856,1363.68 383.082,1358.96 383.721,1341.29 384.333,1339.15 384.893,1337.93 385.652,1335 386.836,1314.57 \n",
       "  387.862,1296.82 388.075,1288.87 389.326,1284.07 389.379,1283.11 389.579,1282.71 390.671,1280.42 391.962,1272.7 393.067,1274.6 393.187,1264.92 394.119,1261 \n",
       "  394.266,1256.55 394.319,1247.23 395.251,1241.23 395.371,1237.54 395.437,1227.5 395.57,1220.01 396.609,1209.85 396.622,1204.6 397.594,1198.72 398.952,1201.57 \n",
       "  399.312,1191.2 400.563,1188.4 400.976,1183.26 401.815,1177.69 402.454,1172.03 403.226,1163.14 403.625,1158.9 404.797,1153.87 405.556,1148.01 406.807,1142.31 \n",
       "  408.045,1139.12 408.259,1143.2 408.511,1135.87 409.337,1127.91 410.509,1122.42 410.655,1111.31 412.625,1111.33 414.17,1105.66 416.793,1102.17 417.552,1099.53 \n",
       "  420.028,1093.99 420.534,1088.49 421.2,1085.28 421.785,1079.74 421.852,1075.11 423.596,1071.35 424.275,1062.91 428.123,1057.15 430.466,1046.88 437.509,1041.68 \n",
       "  441.743,1037.42 444.552,1031.56 452.075,1025.33 452.221,1018.91 453.952,1009.24 454.737,1003.25 455.989,999.652 498.7,993.594 499.033,988.3 502.574,976.488 \n",
       "  506.542,970.316 510.309,963.994 512.653,958.499 514.996,952.315 517.353,946.125 519.696,941.144 520.042,934.883 523.45,926.222 523.517,920.475 526.619,914.091 \n",
       "  526.752,907.398 527.205,901.333 527.91,894.792 530.507,888.24 531.439,881.264 532.943,875.525 533.649,870.644 534.82,865.409 535.113,852.16 536.618,845.561 \n",
       "  538.335,838.752 539.653,832.085 542.849,824.827 543.022,817.859 544.193,811.709 545.352,805.483 546.603,800.088 547.841,793.526 559.145,774.75 559.331,767.805 \n",
       "  561.022,763.24 564.99,751.298 572.299,760.829 572.765,748.128 574.056,764.176 575.308,757.101 576.28,749.946 578.623,742.966 634.288,736.967 636.166,730.021 \n",
       "  644.154,722.315 1669.66,715.629 1679.11,710.07 1829.42,703.856 1838.88,690.298 1848.33,683.076 1856.45,670.958 1919.16,657.897 2083.19,650.789 2088.25,628.872 \n",
       "  2092.64,621.79 2105.15,607.218 2109.81,593.976 2122.06,588.778 2138.04,586.203 2138.97,565.251 2139.64,557.776 2155.88,499.136 2172.79,494.137 2189.7,472.522 \n",
       "  2206.61,447.864 2223.51,433.787 2240.42,401.128 2257.33,343.891 2274.24,320.948 2291.15,305.172 2308.06,277.433 2460.24,271.488 2494.05,263.23 2644.77,245.85 \n",
       "  2771.38,224.178 3070.01,207.046 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip6802)\" d=\"\n",
       "M311.879 1424.15 L345.071 1424.15 L362.379 1424.15 L365.654 1424.15 L365.894 1424.15 L366.812 1424.15 L367.691 1424.15 L368.237 1424.15 L368.49 1424.15 L369.355 1424.15 \n",
       "  L369.422 1424.15 L370.58 1424.15 L371.765 1424.15 L372.937 1424.15 L373.176 1424.15 L374.108 1424.15 L374.348 1424.15 L375.293 1424.15 L375.6 1424.15 L376.691 1424.15 \n",
       "  L376.851 1424.15 L376.864 1424.15 L377.637 1424.15 L379.101 1424.15 L379.341 1424.21 L379.367 1424.21 L381.005 1424.21 L381.378 1424.21 L381.844 1424.21 L382.123 1424.21 \n",
       "  L382.336 1424.21 L382.549 1424.21 L382.563 1424.21 L382.856 1424.21 L383.082 1424.21 L383.721 1424.23 L384.333 1424.23 L384.893 1424.23 L385.652 1424.23 L386.836 1424.25 \n",
       "  L387.862 1424.25 L388.075 1424.26 L389.326 1424.26 L389.379 1424.49 L389.579 1424.49 L390.671 1424.82 L391.962 1424.82 L393.067 1424.87 L393.187 1424.88 L394.119 1424.87 \n",
       "  L394.266 1424.86 L394.319 1424.87 L395.251 1424.88 L395.371 1424.87 L395.437 1424.86 L395.57 1424.84 L396.609 1424.83 L396.622 1424.81 L397.594 1424.81 L398.952 1423.76 \n",
       "  L399.312 1423.64 L400.563 1423.64 L400.976 1423.4 L401.815 1423.36 L402.454 1423.28 L403.226 1423.23 L403.625 1422.8 L404.797 1422.59 L405.556 1422.51 L406.807 1422.3 \n",
       "  L408.045 1422.29 L408.259 1420.57 L408.511 1420.54 L409.337 1419.61 L410.509 1419.49 L410.655 1419.35 L412.625 1415.89 L414.17 1415.85 L416.793 1414.98 L417.552 1413.42 \n",
       "  L420.028 1413.34 L420.534 1413.27 L421.2 1413.26 L421.785 1413.17 L421.852 1413.16 L423.596 1413.16 L424.275 1413.16 L428.123 1413.09 L430.466 1412.59 L437.509 1411.94 \n",
       "  L441.743 1410.97 L444.552 1410.71 L452.075 1410.49 L452.221 1410.33 L453.952 1410.22 L454.737 1410.12 L455.989 1410.12 L498.7 1410.01 L499.033 1409.99 L502.574 1408.97 \n",
       "  L506.542 1408.47 L510.309 1408.37 L512.653 1408.31 L514.996 1408.24 L517.353 1408.15 L519.696 1408.13 L520.042 1408.03 L523.45 1405.69 L523.517 1405.65 L526.619 1405.36 \n",
       "  L526.752 1404.94 L527.205 1404.3 L527.91 1404.06 L530.507 1403.98 L531.439 1403.86 L532.943 1403.85 L533.649 1402.73 L534.82 1401.62 L535.113 1401.15 L536.618 1400.63 \n",
       "  L538.335 1400.34 L539.653 1399.93 L542.849 1399.72 L543.022 1399.48 L544.193 1399.41 L545.352 1399.4 L546.603 1398.13 L547.841 1397.55 L559.145 1395.15 L559.331 1391.18 \n",
       "  L561.022 1389.55 L564.99 1387.68 L572.299 1376.19 L572.765 1375.98 L574.056 1363.83 L575.308 1363.67 L576.28 1363.57 L578.623 1363.15 L634.288 1363.13 L636.166 1362.69 \n",
       "  L644.154 1362.69 L1669.66 1362.63 L1679.11 1361.05 L1829.42 1361.03 L1838.88 1359.65 L1848.33 1359.49 L1856.45 1359.39 L1919.16 1358.4 L2083.19 1358.31 L2088.25 1356.59 \n",
       "  L2092.64 1356.54 L2105.15 1355.96 L2109.81 1355.86 L2122.06 1353.75 L2138.04 1353.75 L2138.97 1353.42 L2139.64 1353.33 L2155.88 1349.46 L2172.79 1349.46 L2189.7 1349.34 \n",
       "  L2206.61 1345.45 L2223.51 1343.65 L2240.42 1343.6 L2257.33 1342.24 L2274.24 1341.95 L2291.15 1341.81 L2308.06 1340.82 L2460.24 1338.77 L2494.05 1338.64 L2644.77 1338.23 \n",
       "  L2771.38 1331.08 L3070.01 1330.61 L3070.01 1280.92 L2771.38 1281.62 L2644.77 1290.67 L2494.05 1291.17 L2460.24 1291.32 L2308.06 1293.74 L2291.15 1294.92 L2274.24 1295.14 \n",
       "  L2257.33 1295.45 L2240.42 1297.28 L2223.51 1297.33 L2206.61 1299.84 L2189.7 1304.61 L2172.79 1304.79 L2155.88 1304.79 L2139.64 1309.67 L2138.97 1309.8 L2138.04 1310.19 \n",
       "  L2122.06 1310.19 L2109.81 1312.61 L2105.15 1312.77 L2092.64 1313.54 L2088.25 1313.63 L2083.19 1315.47 L1919.16 1315.61 L1856.45 1316.96 L1848.33 1317.14 L1838.88 1317.32 \n",
       "  L1829.42 1319.39 L1679.11 1319.41 L1669.66 1321.41 L644.154 1321.5 L636.166 1321.5 L634.288 1321.93 L578.623 1321.97 L576.28 1322.48 L575.308 1322.69 L574.056 1322.83 \n",
       "  L572.765 1338.79 L572.299 1339.14 L564.99 1354.14 L561.022 1356.95 L559.331 1359.19 L559.145 1364.95 L547.841 1368.45 L546.603 1369.44 L545.352 1371.26 L544.193 1371.31 \n",
       "  L543.022 1371.4 L542.849 1371.73 L539.653 1372.07 L538.335 1372.77 L536.618 1373.17 L535.113 1374.02 L534.82 1374.75 L533.649 1376.52 L532.943 1378.31 L531.439 1378.39 \n",
       "  L530.507 1378.56 L527.91 1378.67 L527.205 1378.96 L526.752 1380.02 L526.619 1380.53 L523.517 1380.95 L523.45 1381.02 L520.042 1384.63 L519.696 1384.86 L517.353 1384.89 \n",
       "  L514.996 1385.1 L512.653 1385.18 L510.309 1385.2 L506.542 1385.35 L502.574 1386.02 L499.033 1387.6 L498.7 1387.67 L455.989 1387.82 L454.737 1387.82 L453.952 1387.99 \n",
       "  L452.221 1388.16 L452.075 1388.36 L444.552 1388.68 L441.743 1389.17 L437.509 1390.96 L430.466 1392.16 L428.123 1393.01 L424.275 1393.14 L423.596 1393.19 L421.852 1393.2 \n",
       "  L421.785 1393.23 L421.2 1393.41 L420.534 1393.41 L420.028 1393.59 L417.552 1393.76 L416.793 1396.27 L414.17 1398.49 L412.625 1398.62 L410.655 1405.47 L410.509 1405.78 \n",
       "  L409.337 1405.99 L408.511 1408.27 L408.259 1408.3 L408.045 1412.55 L406.807 1412.56 L405.556 1413.1 L404.797 1413.38 L403.625 1414.15 L403.226 1415.26 L402.454 1415.34 \n",
       "  L401.815 1415.59 L400.976 1415.69 L400.563 1416.31 L399.312 1416.32 L398.952 1416.67 L397.594 1419.84 L396.622 1420.01 L396.609 1420.06 L395.57 1420.17 L395.437 1420.7 \n",
       "  L395.371 1420.79 L395.251 1420.79 L394.319 1420.81 L394.266 1420.93 L394.119 1420.94 L393.187 1420.96 L393.067 1421.1 L391.962 1422.59 L390.671 1422.63 L389.579 1423.66 \n",
       "  L389.379 1423.66 L389.326 1424.02 L388.075 1424.02 L387.862 1424.03 L386.836 1424.03 L385.652 1424.06 L384.893 1424.06 L384.333 1424.06 L383.721 1424.06 L383.082 1424.08 \n",
       "  L382.856 1424.08 L382.563 1424.08 L382.549 1424.08 L382.336 1424.08 L382.123 1424.08 L381.844 1424.08 L381.378 1424.08 L381.005 1424.08 L379.367 1424.08 L379.341 1424.08 \n",
       "  L379.101 1424.15 L377.637 1424.15 L376.864 1424.15 L376.851 1424.15 L376.691 1424.15 L375.6 1424.15 L375.293 1424.15 L374.348 1424.15 L374.108 1424.15 L373.176 1424.15 \n",
       "  L372.937 1424.15 L371.765 1424.15 L370.58 1424.15 L369.422 1424.15 L369.355 1424.15 L368.49 1424.15 L368.237 1424.15 L367.691 1424.15 L366.812 1424.15 L365.894 1424.15 \n",
       "  L365.654 1424.15 L362.379 1424.15 L345.071 1424.15 L311.879 1424.15  Z\n",
       "  \" fill=\"#0000ff\" fill-rule=\"evenodd\" fill-opacity=\"0.5\"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#0000ff; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  311.879,1424.15 345.071,1424.15 362.379,1424.15 365.654,1424.15 365.894,1424.15 366.812,1424.15 367.691,1424.15 368.237,1424.15 368.49,1424.15 369.355,1424.15 \n",
       "  369.422,1424.15 370.58,1424.15 371.765,1424.15 372.937,1424.15 373.176,1424.15 374.108,1424.15 374.348,1424.15 375.293,1424.15 375.6,1424.15 376.691,1424.15 \n",
       "  376.851,1424.15 376.864,1424.15 377.637,1424.15 379.101,1424.15 379.341,1424.15 379.367,1424.15 381.005,1424.15 381.378,1424.15 381.844,1424.15 382.123,1424.15 \n",
       "  382.336,1424.15 382.549,1424.15 382.563,1424.15 382.856,1424.15 383.082,1424.15 383.721,1424.14 384.333,1424.14 384.893,1424.14 385.652,1424.14 386.836,1424.14 \n",
       "  387.862,1424.14 388.075,1424.14 389.326,1424.14 389.379,1424.07 389.579,1424.07 390.671,1423.72 391.962,1423.71 393.067,1422.98 393.187,1422.92 394.119,1422.91 \n",
       "  394.266,1422.9 394.319,1422.84 395.251,1422.83 395.371,1422.83 395.437,1422.78 395.57,1422.51 396.609,1422.44 396.622,1422.41 397.594,1422.32 398.952,1420.21 \n",
       "  399.312,1419.98 400.563,1419.97 400.976,1419.54 401.815,1419.47 402.454,1419.31 403.226,1419.25 403.625,1418.47 404.797,1417.98 405.556,1417.81 406.807,1417.43 \n",
       "  408.045,1417.42 408.259,1414.43 408.511,1414.4 409.337,1412.8 410.509,1412.64 410.655,1412.41 412.625,1407.26 414.17,1407.17 416.793,1405.63 417.552,1403.59 \n",
       "  420.028,1403.46 420.534,1403.34 421.2,1403.34 421.785,1403.2 421.852,1403.18 423.596,1403.17 424.275,1403.15 428.123,1403.05 430.466,1402.38 437.509,1401.45 \n",
       "  441.743,1400.07 444.552,1399.69 452.075,1399.42 452.221,1399.24 453.952,1399.11 454.737,1398.97 455.989,1398.97 498.7,1398.84 499.033,1398.79 502.574,1397.49 \n",
       "  506.542,1396.91 510.309,1396.78 512.653,1396.75 514.996,1396.67 517.353,1396.52 519.696,1396.49 520.042,1396.33 523.45,1393.35 523.517,1393.3 526.619,1392.94 \n",
       "  526.752,1392.48 527.205,1391.63 527.91,1391.37 530.507,1391.27 531.439,1391.13 532.943,1391.08 533.649,1389.63 534.82,1388.19 535.113,1387.59 536.618,1386.9 \n",
       "  538.335,1386.55 539.653,1386 542.849,1385.72 543.022,1385.44 544.193,1385.36 545.352,1385.33 546.603,1383.79 547.841,1383 559.145,1380.05 559.331,1375.18 \n",
       "  561.022,1373.25 564.99,1370.91 572.299,1357.67 572.765,1357.39 574.056,1343.33 575.308,1343.18 576.28,1343.03 578.623,1342.56 634.288,1342.53 636.166,1342.1 \n",
       "  644.154,1342.09 1669.66,1342.02 1679.11,1340.23 1829.42,1340.21 1838.88,1338.49 1848.33,1338.32 1856.45,1338.17 1919.16,1337 2083.19,1336.89 2088.25,1335.11 \n",
       "  2092.64,1335.04 2105.15,1334.36 2109.81,1334.24 2122.06,1331.97 2138.04,1331.97 2138.97,1331.61 2139.64,1331.5 2155.88,1327.12 2172.79,1327.12 2189.7,1326.97 \n",
       "  2206.61,1322.64 2223.51,1320.49 2240.42,1320.44 2257.33,1318.84 2274.24,1318.55 2291.15,1318.36 2308.06,1317.28 2460.24,1315.05 2494.05,1314.91 2644.77,1314.45 \n",
       "  2771.38,1306.35 3070.01,1305.76 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#ff0000; stroke-width:8; stroke-opacity:1; fill:none\" stroke-dasharray=\"4, 8\" points=\"\n",
       "  311.879,1424.15 345.071,1424.15 362.379,1424.15 365.654,1424.15 365.894,1424.15 366.812,1424.15 367.691,1424.15 368.237,1424.15 368.49,1424.15 369.355,1424.15 \n",
       "  369.422,1424.15 370.58,1424.15 371.765,1424.15 372.937,1424.15 373.176,1424.15 374.108,1424.15 374.348,1424.15 375.293,1422.18 375.6,1422.18 376.691,1422.18 \n",
       "  376.851,1422.18 376.864,1420.21 377.637,1420.21 379.101,1420.21 379.341,1412.32 379.367,1406.41 381.005,1404.44 381.378,1402.46 381.844,1402.46 382.123,1402.46 \n",
       "  382.336,1402.46 382.549,1394.58 382.563,1388.67 382.856,1388.67 383.082,1386.69 383.721,1372.9 384.333,1372.9 384.893,1372.9 385.652,1368.95 386.836,1357.13 \n",
       "  387.862,1347.27 388.075,1343.33 389.326,1337.41 389.379,1335.44 389.579,1335.44 390.671,1333.47 391.962,1327.56 393.067,1331.5 393.187,1321.64 394.119,1321.64 \n",
       "  394.266,1319.67 394.319,1321.64 395.251,1315.73 395.371,1311.79 395.437,1309.82 395.57,1313.76 396.609,1311.79 396.622,1307.84 397.594,1301.93 398.952,1297.99 \n",
       "  399.312,1292.07 400.563,1292.07 400.976,1284.19 401.815,1282.22 402.454,1278.28 403.226,1278.28 403.625,1282.22 404.797,1292.07 405.556,1288.13 406.807,1290.1 \n",
       "  408.045,1288.13 408.259,1299.96 408.511,1294.05 409.337,1301.93 410.509,1299.96 410.655,1299.96 412.625,1301.93 414.17,1296.02 416.793,1297.99 417.552,1303.9 \n",
       "  420.028,1303.9 420.534,1303.9 421.2,1303.9 421.785,1303.9 421.852,1301.93 423.596,1301.93 424.275,1299.96 428.123,1297.99 430.466,1297.99 437.509,1301.93 \n",
       "  441.743,1301.93 444.552,1294.05 452.075,1299.96 452.221,1303.9 453.952,1301.93 454.737,1301.93 455.989,1299.96 498.7,1299.96 499.033,1301.93 502.574,1299.96 \n",
       "  506.542,1296.02 510.309,1297.99 512.653,1297.99 514.996,1296.02 517.353,1297.99 519.696,1297.99 520.042,1290.1 523.45,1292.07 523.517,1292.07 526.619,1290.1 \n",
       "  526.752,1288.13 527.205,1286.16 527.91,1286.16 530.507,1286.16 531.439,1290.1 532.943,1290.1 533.649,1282.22 534.82,1284.19 535.113,1282.22 536.618,1278.28 \n",
       "  538.335,1278.28 539.653,1280.25 542.849,1280.25 543.022,1282.22 544.193,1282.22 545.352,1282.22 546.603,1280.25 547.841,1278.28 559.145,1282.22 559.331,1282.22 \n",
       "  561.022,1284.19 564.99,1290.1 572.299,1292.07 572.765,1290.1 574.056,1290.1 575.308,1290.1 576.28,1282.22 578.623,1276.3 634.288,1272.36 636.166,1272.36 \n",
       "  644.154,1262.51 1669.66,1262.51 1679.11,1262.51 1829.42,1260.54 1838.88,1264.48 1848.33,1264.48 1856.45,1264.48 1919.16,1258.56 2083.19,1248.71 2088.25,1250.68 \n",
       "  2092.64,1246.74 2105.15,1248.71 2109.81,1248.71 2122.06,1250.68 2138.04,1250.68 2138.97,1250.68 2139.64,1246.74 2155.88,1248.71 2172.79,1248.71 2189.7,1240.82 \n",
       "  2206.61,1240.82 2223.51,1238.85 2240.42,1236.88 2257.33,1227.02 2274.24,1225.05 2291.15,1215.2 2308.06,1211.25 2460.24,1213.23 2494.05,1217.17 2644.77,1213.23 \n",
       "  2771.38,1211.25 3070.01,1209.28 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6802)\" style=\"stroke:#0000ff; stroke-width:8; stroke-opacity:1; fill:none\" stroke-dasharray=\"4, 8\" points=\"\n",
       "  311.879,1424.15 345.071,1424.15 362.379,1424.15 365.654,1424.15 365.894,1424.15 366.812,1424.15 367.691,1424.15 368.237,1424.15 368.49,1424.15 369.355,1424.15 \n",
       "  369.422,1424.15 370.58,1424.15 371.765,1424.15 372.937,1424.15 373.176,1424.15 374.108,1424.15 374.348,1424.15 375.293,1424.15 375.6,1424.15 376.691,1424.15 \n",
       "  376.851,1424.15 376.864,1424.15 377.637,1424.15 379.101,1424.15 379.341,1424.15 379.367,1424.15 381.005,1424.15 381.378,1424.15 381.844,1424.15 382.123,1424.15 \n",
       "  382.336,1424.15 382.549,1424.15 382.563,1424.15 382.856,1424.15 383.082,1424.15 383.721,1424.15 384.333,1424.15 384.893,1424.15 385.652,1424.15 386.836,1424.15 \n",
       "  387.862,1424.15 388.075,1424.15 389.326,1424.15 389.379,1424.15 389.579,1424.15 390.671,1424.15 391.962,1424.15 393.067,1424.15 393.187,1424.15 394.119,1424.15 \n",
       "  394.266,1424.15 394.319,1424.15 395.251,1424.15 395.371,1424.15 395.437,1424.15 395.57,1424.15 396.609,1424.15 396.622,1424.15 397.594,1424.15 398.952,1424.15 \n",
       "  399.312,1424.15 400.563,1424.15 400.976,1422.18 401.815,1422.18 402.454,1422.18 403.226,1422.18 403.625,1418.23 404.797,1414.29 405.556,1414.29 406.807,1412.32 \n",
       "  408.045,1412.32 408.259,1410.35 408.511,1410.35 409.337,1408.38 410.509,1408.38 410.655,1408.38 412.625,1408.38 414.17,1408.38 416.793,1402.46 417.552,1396.55 \n",
       "  420.028,1396.55 420.534,1396.55 421.2,1396.55 421.785,1394.58 421.852,1394.58 423.596,1394.58 424.275,1394.58 428.123,1394.58 430.466,1392.61 437.509,1388.67 \n",
       "  441.743,1384.72 444.552,1384.72 452.075,1382.75 452.221,1382.75 453.952,1380.78 454.737,1376.84 455.989,1376.84 498.7,1376.84 499.033,1376.84 502.574,1365.01 \n",
       "  506.542,1365.01 510.309,1361.07 512.653,1361.07 514.996,1361.07 517.353,1361.07 519.696,1361.07 520.042,1361.07 523.45,1353.18 523.517,1353.18 526.619,1353.18 \n",
       "  526.752,1353.18 527.205,1347.27 527.91,1345.3 530.507,1345.3 531.439,1345.3 532.943,1345.3 533.649,1341.36 534.82,1341.36 535.113,1333.47 536.618,1333.47 \n",
       "  538.335,1333.47 539.653,1329.53 542.849,1323.61 543.022,1321.64 544.193,1321.64 545.352,1319.67 546.603,1319.67 547.841,1313.76 559.145,1288.13 559.331,1270.39 \n",
       "  561.022,1264.48 564.99,1250.68 572.299,1252.65 572.765,1252.65 574.056,1252.65 575.308,1252.65 576.28,1252.65 578.623,1252.65 634.288,1252.65 636.166,1252.65 \n",
       "  644.154,1252.65 1669.66,1252.65 1679.11,1240.82 1829.42,1240.82 1838.88,1236.88 1848.33,1236.88 1856.45,1236.88 1919.16,1236.88 2083.19,1236.88 2088.25,1236.88 \n",
       "  2092.64,1236.88 2105.15,1230.97 2109.81,1229 2122.06,1227.02 2138.04,1227.02 2138.97,1223.08 2139.64,1223.08 2155.88,1209.28 2172.79,1209.28 2189.7,1209.28 \n",
       "  2206.61,1203.37 2223.51,1201.4 2240.42,1201.4 2257.33,1193.51 2274.24,1193.51 2291.15,1193.51 2308.06,1193.51 2460.24,1193.51 2494.05,1193.51 2644.77,1193.51 \n",
       "  2771.38,1183.66 3070.01,1179.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip6800)\" d=\"\n",
       "M2379.51 1270.83 L3056.76 1270.83 L3056.76 968.434 L2379.51 968.434  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2379.51,1270.83 3056.76,1270.83 3056.76,968.434 2379.51,968.434 2379.51,1270.83 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#ff0000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2411.51,1028.91 2603.51,1028.91 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 2635.51, 1046.41)\" x=\"2635.51\" y=\"1046.41\">[&quot;dim 1&quot;, &quot;dim 2&quot;]</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#0000ff; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2411.51,1089.39 2603.51,1089.39 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 2635.51, 1106.89)\" x=\"2635.51\" y=\"1106.89\">[&quot;dim 1&quot;, &quot;dim 2&quot;]</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#ff0000; stroke-width:8; stroke-opacity:1; fill:none\" stroke-dasharray=\"4, 8\" points=\"\n",
       "  2411.51,1149.87 2603.51,1149.87 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 2635.51, 1167.37)\" x=\"2635.51\" y=\"1167.37\">original dim 1</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#0000ff; stroke-width:8; stroke-opacity:1; fill:none\" stroke-dasharray=\"4, 8\" points=\"\n",
       "  2411.51,1210.35 2603.51,1210.35 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 2635.51, 1227.85)\" x=\"2635.51\" y=\"1227.85\">original dim 2</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip6800)\" d=\"\n",
       "M229.135 3025.62 L3152.76 3025.62 L3152.76 1684.46 L229.135 1684.46  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip6803\">\n",
       "    <rect x=\"229\" y=\"1684\" width=\"2925\" height=\"1342\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  640.626,3025.62 640.626,1684.46 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1306.32,3025.62 1306.32,1684.46 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1972.02,3025.62 1972.02,1684.46 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  2637.71,3025.62 2637.71,1684.46 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,3025.62 3152.76,3025.62 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,2690.33 3152.76,2690.33 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,2355.04 3152.76,2355.04 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,2019.75 3152.76,2019.75 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,1684.46 3152.76,1684.46 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,3025.62 3152.76,3025.62 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,3025.62 229.135,1684.46 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  640.626,3025.62 640.626,3005.5 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1306.32,3025.62 1306.32,3005.5 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1972.02,3025.62 1972.02,3005.5 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2637.71,3025.62 2637.71,3005.5 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,3025.62 272.99,3025.62 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,2690.33 272.99,2690.33 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,2355.04 272.99,2355.04 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,2019.75 272.99,2019.75 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1684.46 272.99,1684.46 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 640.626, 3083.62)\" x=\"640.626\" y=\"3083.62\">500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1306.32, 3083.62)\" x=\"1306.32\" y=\"3083.62\">1000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1972.02, 3083.62)\" x=\"1972.02\" y=\"3083.62\">1500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 2637.71, 3083.62)\" x=\"2637.71\" y=\"3083.62\">2000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 3043.12)\" x=\"197.135\" y=\"3043.12\">0</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 2707.83)\" x=\"197.135\" y=\"2707.83\">5</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 2372.54)\" x=\"197.135\" y=\"2372.54\">10</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 2037.25)\" x=\"197.135\" y=\"2037.25\">15</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 197.135, 1701.96)\" x=\"197.135\" y=\"1701.96\">20</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(0, 1690.95, 3167.48)\" x=\"1690.95\" y=\"3167.48\">time</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(-90, 100.048, 2355.04)\" x=\"100.048\" y=\"2355.04\">betti number</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip6803)\" d=\"\n",
       "M311.879 2958.56 L345.071 2922.78 L362.379 2867.07 L365.654 2785.4 L365.894 2734.19 L366.812 2649.61 L367.691 2653.02 L368.237 2618.19 L368.49 2601.22 L369.355 2660.21 \n",
       "  L369.422 2627.87 L370.58 2602.57 L371.765 2587.69 L372.937 2653.35 L373.176 2632 L374.108 2671.88 L374.348 2690.83 L375.293 2758.6 L375.6 2784.76 L376.691 2830.41 \n",
       "  L376.851 2835.06 L376.864 2840.16 L377.637 2856.52 L379.101 2851.16 L379.341 2879.98 L379.367 2886.98 L381.005 2891.73 L381.378 2887.37 L381.844 2894.28 L382.123 2893.05 \n",
       "  L382.336 2893.47 L382.549 2891.73 L382.563 2913.08 L382.856 2916.97 L383.082 2916.6 L383.721 2933.45 L384.333 2932.54 L384.893 2925.56 L385.652 2926.55 L386.836 2933.55 \n",
       "  L387.862 2943.28 L388.075 2914.92 L389.326 2917.26 L389.379 2931.11 L389.579 2913.97 L390.671 2929.97 L391.962 2932.5 L393.067 2939.51 L393.187 2939.24 L394.119 2940.3 \n",
       "  L394.266 2938.7 L394.319 2935.27 L395.251 2936.21 L395.371 2937.29 L395.437 2941.42 L395.57 2944.25 L396.609 2945.02 L396.622 2947 L397.594 2948.23 L398.952 2953.99 \n",
       "  L399.312 2956.05 L400.563 2955.51 L400.976 2956.86 L401.815 2957.93 L402.454 2958.04 L403.226 2960.22 L403.625 2961.16 L404.797 2963.02 L405.556 2964.18 L406.807 2964.79 \n",
       "  L408.045 2964.53 L408.259 2966.99 L408.511 2967.38 L409.337 2968.73 L410.509 2969.05 L410.655 2969.27 L412.625 2970.32 L414.17 2970.26 L416.793 2971.04 L417.552 2971.36 \n",
       "  L420.028 2971.76 L420.534 2971.88 L421.2 2972.1 L421.785 2972.43 L421.852 2972.69 L423.596 2972.53 L424.275 2972.82 L428.123 2972.85 L430.466 2973.11 L437.509 2973.41 \n",
       "  L441.743 2973.51 L444.552 2973.78 L452.075 2973.73 L452.221 2973.82 L453.952 2973.8 L454.737 2973.5 L455.989 2973.89 L498.7 2973.76 L499.033 2973.79 L502.574 2973.73 \n",
       "  L506.542 2973.62 L510.309 2973.71 L512.653 2973.82 L514.996 2973.86 L517.353 2973.98 L519.696 2974.01 L520.042 2974.1 L523.45 2974.12 L523.517 2974.2 L526.619 2974.26 \n",
       "  L526.752 2974.31 L527.205 2974.27 L527.91 2974.16 L530.507 2974.07 L531.439 2974.11 L532.943 2973.81 L533.649 2973.87 L534.82 2973.9 L535.113 2973.58 L536.618 2973.39 \n",
       "  L538.335 2973.19 L539.653 2972.97 L542.849 2972.97 L543.022 2972.95 L544.193 2972.94 L545.352 2972.91 L546.603 2972.77 L547.841 2972.73 L559.145 2972.47 L559.331 2971.98 \n",
       "  L561.022 2971.85 L564.99 2971.67 L572.299 2971.28 L572.765 2971.32 L574.056 2970.58 L575.308 2970.58 L576.28 2970.58 L578.623 2970.33 L634.288 2970.2 L636.166 2970.07 \n",
       "  L644.154 2970.13 L1669.66 2969.92 L1679.11 2969.69 L1829.42 2969.69 L1838.88 2969.53 L1848.33 2969.36 L1856.45 2969.19 L1919.16 2968.44 L2083.19 2968.44 L2088.25 2968.22 \n",
       "  L2092.64 2968.22 L2105.15 2967.62 L2109.81 2967.62 L2122.06 2967.36 L2138.04 2970.52 L2138.97 2970.33 L2139.64 2970.27 L2155.88 2967.99 L2172.79 2967.99 L2189.7 2969.03 \n",
       "  L2206.61 2967.87 L2223.51 2967.62 L2240.42 2966.27 L2257.33 2964.57 L2274.24 2964 L2291.15 2963.68 L2308.06 2963.34 L2460.24 2962.96 L2494.05 2962.96 L2644.77 2962.53 \n",
       "  L2771.38 2958.56 L3070.01 2958.56 L3070.01 2958.56 L2771.38 2958.56 L2644.77 2954.06 L2494.05 2953.49 L2460.24 2953.49 L2308.06 2952.98 L2291.15 2952.5 L2274.24 2952.05 \n",
       "  L2257.33 2951.21 L2240.42 2948.43 L2223.51 2945.88 L2206.61 2945.36 L2189.7 2942.73 L2172.79 2945.1 L2155.88 2945.1 L2139.64 2939.88 L2138.97 2939.68 L2138.04 2939.09 \n",
       "  L2122.06 2946.41 L2109.81 2945.88 L2105.15 2945.88 L2092.64 2944.61 L2088.25 2944.61 L2083.19 2944.12 L1919.16 2944.12 L1856.45 2942.84 L1848.33 2942.39 L1838.88 2941.96 \n",
       "  L1829.42 2941.53 L1679.11 2941.53 L1669.66 2940.9 L644.154 2940.28 L636.166 2940.48 L634.288 2940.08 L578.623 2939.68 L576.28 2938.9 L575.308 2938.9 L574.056 2938.9 \n",
       "  L572.765 2936.95 L572.299 2937.13 L564.99 2935.53 L561.022 2934.68 L559.331 2934.01 L559.145 2932.32 L547.841 2930.31 L546.603 2930.01 L545.352 2928.39 L544.193 2927.81 \n",
       "  L543.022 2927.67 L542.849 2927.39 L539.653 2927.25 L538.335 2926.5 L536.618 2925.89 L535.113 2924.49 L534.82 2922.7 L533.649 2922.06 L532.943 2921.19 L531.439 2920.48 \n",
       "  L530.507 2919.99 L527.91 2919.22 L527.205 2918.58 L526.752 2917.6 L526.619 2917.25 L523.517 2916.9 L523.45 2916.44 L520.042 2914.45 L519.696 2914.01 L517.353 2913.9 \n",
       "  L514.996 2913.36 L512.653 2912.46 L510.309 2912.03 L506.542 2911.72 L502.574 2909.59 L499.033 2906.58 L498.7 2906.48 L455.989 2905.41 L454.737 2905.81 L453.952 2904.29 \n",
       "  L452.221 2903.07 L452.075 2902.48 L444.552 2902.17 L441.743 2901.23 L437.509 2899.58 L430.466 2897.34 L428.123 2895.32 L424.275 2894.68 L423.596 2893.36 L421.852 2893.33 \n",
       "  L421.785 2892.65 L421.2 2891.37 L420.534 2891.33 L420.028 2890.64 L417.552 2889.84 L416.793 2886.27 L414.17 2883.16 L412.625 2882.16 L410.655 2876.23 L410.509 2872.83 \n",
       "  L409.337 2871.54 L408.511 2866.33 L408.259 2864.7 L408.045 2859.38 L406.807 2858.72 L405.556 2855.84 L404.797 2854.05 L403.625 2849.07 L403.226 2843.17 L402.454 2841.73 \n",
       "  L401.815 2839.03 L400.976 2836.87 L400.563 2832.33 L399.312 2832.99 L398.952 2827.41 L397.594 2813.58 L396.622 2810.39 L396.609 2806.33 L395.57 2805.9 L395.437 2799.6 \n",
       "  L395.371 2793.54 L395.251 2790.87 L394.319 2790.2 L394.266 2802.32 L394.119 2810.25 L393.187 2807.56 L393.067 2808.36 L391.962 2791.63 L390.671 2787.59 L389.579 2756.51 \n",
       "  L389.379 2784.98 L389.326 2753.63 L388.075 2748.32 L387.862 2790.91 L386.836 2769.39 L385.652 2754.93 L384.893 2750.69 L384.333 2761.95 L383.721 2765.34 L383.082 2738.33 \n",
       "  L382.856 2732.99 L382.563 2721.32 L382.549 2684.87 L382.336 2686.21 L382.123 2685.29 L381.844 2687.01 L381.378 2674.07 L381.005 2677.09 L379.367 2672.45 L379.341 2659.07 \n",
       "  L379.101 2615.6 L377.637 2627.14 L376.864 2601.26 L376.851 2593.88 L376.691 2581.63 L375.6 2515.83 L375.293 2480.3 L374.348 2387.12 L374.108 2370.54 L373.176 2340.81 \n",
       "  L372.937 2371.36 L371.765 2313.1 L370.58 2340.06 L369.422 2376.46 L369.355 2417.75 L368.49 2387.01 L368.237 2425.7 L367.691 2477.11 L366.812 2503.72 L365.894 2619.24 \n",
       "  L365.654 2681.5 L362.379 2809.71 L345.071 2877.66 L311.879 2958.56  Z\n",
       "  \" fill=\"#008000\" fill-rule=\"evenodd\" fill-opacity=\"0.5\"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#008000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  311.879,2958.56 345.071,2900.22 362.379,2838.39 365.654,2733.45 365.894,2676.72 366.812,2576.67 367.691,2565.06 368.237,2521.95 368.49,2494.12 369.355,2538.98 \n",
       "  369.422,2502.16 370.58,2471.32 371.765,2450.4 372.937,2512.36 373.176,2486.41 374.108,2521.21 374.348,2538.98 375.293,2619.45 375.6,2650.3 376.691,2706.02 \n",
       "  376.851,2714.47 376.864,2720.71 377.637,2741.83 379.101,2733.38 379.341,2769.52 379.367,2779.72 381.005,2784.41 381.378,2780.72 381.844,2790.65 382.123,2789.17 \n",
       "  382.336,2789.84 382.549,2788.3 382.563,2817.2 382.856,2824.98 383.082,2827.46 383.721,2849.39 384.333,2847.24 384.893,2838.12 385.652,2840.74 386.836,2851.47 \n",
       "  387.862,2867.09 388.075,2831.62 389.326,2835.44 389.379,2858.04 389.579,2835.24 390.671,2858.78 391.962,2862.06 393.067,2873.93 393.187,2873.4 394.119,2875.27 \n",
       "  394.266,2870.51 394.319,2862.73 395.251,2863.54 395.371,2865.42 395.437,2870.51 395.57,2875.07 396.609,2875.68 396.622,2878.69 397.594,2880.91 398.952,2890.7 \n",
       "  399.312,2894.52 400.563,2893.92 400.976,2896.87 401.815,2898.48 402.454,2899.88 403.226,2901.7 403.625,2905.12 404.797,2908.54 405.556,2910.01 406.807,2911.75 \n",
       "  408.045,2911.96 408.259,2915.84 408.511,2916.85 409.337,2920.14 410.509,2920.94 410.655,2922.75 412.625,2926.24 414.17,2926.71 416.793,2928.65 417.552,2930.6 \n",
       "  420.028,2931.2 420.534,2931.6 421.2,2931.74 421.785,2932.54 421.852,2933.01 423.596,2932.94 424.275,2933.75 428.123,2934.08 430.466,2935.22 437.509,2936.5 \n",
       "  441.743,2937.37 444.552,2937.97 452.075,2938.11 452.221,2938.44 453.952,2939.05 454.737,2939.65 455.989,2939.65 498.7,2940.12 499.033,2940.19 502.574,2941.66 \n",
       "  506.542,2942.67 510.309,2942.87 512.653,2943.14 514.996,2943.61 517.353,2943.94 519.696,2944.01 520.042,2944.28 523.45,2945.28 523.517,2945.55 526.619,2945.75 \n",
       "  526.752,2945.95 527.205,2946.42 527.91,2946.69 530.507,2947.03 531.439,2947.29 532.943,2947.5 533.649,2947.97 534.82,2948.3 535.113,2949.04 536.618,2949.64 \n",
       "  538.335,2949.84 539.653,2950.11 542.849,2950.18 543.022,2950.31 544.193,2950.38 545.352,2950.65 546.603,2951.39 547.841,2951.52 559.145,2952.39 559.331,2952.99 \n",
       "  561.022,2953.26 564.99,2953.6 572.299,2954.2 572.765,2954.13 574.056,2954.74 575.308,2954.74 576.28,2954.74 578.623,2955.01 634.288,2955.14 636.166,2955.27 \n",
       "  644.154,2955.21 1669.66,2955.41 1679.11,2955.61 1829.42,2955.61 1838.88,2955.74 1848.33,2955.88 1856.45,2956.01 1919.16,2956.28 2083.19,2956.28 2088.25,2956.41 \n",
       "  2092.64,2956.41 2105.15,2956.75 2109.81,2956.75 2122.06,2956.88 2138.04,2954.81 2138.97,2955.01 2139.64,2955.07 2155.88,2956.55 2172.79,2956.55 2189.7,2955.88 \n",
       "  2206.61,2956.62 2223.51,2956.75 2240.42,2957.35 2257.33,2957.89 2274.24,2958.02 2291.15,2958.09 2308.06,2958.16 2460.24,2958.23 2494.05,2958.23 2644.77,2958.29 \n",
       "  2771.38,2958.56 3070.01,2958.56 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6803)\" style=\"stroke:#008000; stroke-width:8; stroke-opacity:1; fill:none\" stroke-dasharray=\"4, 8\" points=\"\n",
       "  311.879,2958.56 345.071,2891.5 362.379,2824.44 365.654,2757.39 365.894,2757.39 366.812,2757.39 367.691,2690.33 368.237,2623.27 368.49,2556.21 369.355,2556.21 \n",
       "  369.422,2489.16 370.58,2489.16 371.765,2489.16 372.937,2489.16 373.176,2489.16 374.108,2556.21 374.348,2556.21 375.293,2556.21 375.6,2489.16 376.691,2690.33 \n",
       "  376.851,2757.39 376.864,2824.44 377.637,2824.44 379.101,2824.44 379.341,2891.5 379.367,2891.5 381.005,2891.5 381.378,2891.5 381.844,2891.5 382.123,2824.44 \n",
       "  382.336,2824.44 382.549,2824.44 382.563,2824.44 382.856,2958.56 383.082,2958.56 383.721,2891.5 384.333,2891.5 384.893,2891.5 385.652,2891.5 386.836,2891.5 \n",
       "  387.862,2891.5 388.075,2824.44 389.326,2824.44 389.379,2958.56 389.579,2958.56 390.671,2958.56 391.962,2958.56 393.067,2958.56 393.187,2958.56 394.119,2958.56 \n",
       "  394.266,2958.56 394.319,2891.5 395.251,2891.5 395.371,2891.5 395.437,2891.5 395.57,2958.56 396.609,2958.56 396.622,2958.56 397.594,2958.56 398.952,2958.56 \n",
       "  399.312,2958.56 400.563,2958.56 400.976,2958.56 401.815,2958.56 402.454,2958.56 403.226,2958.56 403.625,2958.56 404.797,2958.56 405.556,2958.56 406.807,2958.56 \n",
       "  408.045,2958.56 408.259,2958.56 408.511,2958.56 409.337,2958.56 410.509,2958.56 410.655,2958.56 412.625,2958.56 414.17,2958.56 416.793,2958.56 417.552,2958.56 \n",
       "  420.028,2958.56 420.534,2958.56 421.2,2891.5 421.785,2891.5 421.852,2891.5 423.596,2891.5 424.275,2891.5 428.123,2891.5 430.466,2891.5 437.509,2891.5 \n",
       "  441.743,2891.5 444.552,2891.5 452.075,2891.5 452.221,2891.5 453.952,2891.5 454.737,2891.5 455.989,2891.5 498.7,2891.5 499.033,2891.5 502.574,2891.5 \n",
       "  506.542,2891.5 510.309,2891.5 512.653,2891.5 514.996,2891.5 517.353,2891.5 519.696,2891.5 520.042,2891.5 523.45,2891.5 523.517,2891.5 526.619,2891.5 \n",
       "  526.752,2891.5 527.205,2891.5 527.91,2891.5 530.507,2891.5 531.439,2891.5 532.943,2891.5 533.649,2891.5 534.82,2891.5 535.113,2891.5 536.618,2891.5 \n",
       "  538.335,2891.5 539.653,2891.5 542.849,2891.5 543.022,2891.5 544.193,2891.5 545.352,2891.5 546.603,2891.5 547.841,2891.5 559.145,2891.5 559.331,2891.5 \n",
       "  561.022,2891.5 564.99,2891.5 572.299,2891.5 572.765,2891.5 574.056,2891.5 575.308,2891.5 576.28,2891.5 578.623,2891.5 634.288,2891.5 636.166,2891.5 \n",
       "  644.154,2891.5 1669.66,2891.5 1679.11,2891.5 1829.42,2891.5 1838.88,2891.5 1848.33,2891.5 1856.45,2891.5 1919.16,2891.5 2083.19,2891.5 2088.25,2891.5 \n",
       "  2092.64,2891.5 2105.15,2891.5 2109.81,2891.5 2122.06,2891.5 2138.04,2891.5 2138.97,2891.5 2139.64,2891.5 2155.88,2891.5 2172.79,2891.5 2189.7,2958.56 \n",
       "  2206.61,2958.56 2223.51,2958.56 2240.42,2958.56 2257.33,2958.56 2274.24,2958.56 2291.15,2958.56 2308.06,2958.56 2460.24,2958.56 2494.05,2958.56 2644.77,2958.56 \n",
       "  2771.38,2958.56 3070.01,2958.56 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip6800)\" d=\"\n",
       "M2450.55 2833.62 L3056.76 2833.62 L3056.76 2652.18 L2450.55 2652.18  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2450.55,2833.62 3056.76,2833.62 3056.76,2652.18 2450.55,2652.18 2450.55,2833.62 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#008000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2482.55,2712.66 2674.55,2712.66 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 2706.55, 2730.16)\" x=\"2706.55\" y=\"2730.16\">[&quot;dim 0&quot;]</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip6800)\" style=\"stroke:#008000; stroke-width:8; stroke-opacity:1; fill:none\" stroke-dasharray=\"4, 8\" points=\"\n",
       "  2482.55,2773.14 2674.55,2773.14 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6800)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 2706.55, 2790.64)\" x=\"2706.55\" y=\"2790.64\">original dim 0</text>\n",
       "</g>\n",
       "</svg>\n"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Identical Degree Distribution\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0, 20))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "\n",
    "plot(p1,p2, layout = (2,1), size = (800,800))\n",
    "\n",
    "#savefig(\"rewiredTrials.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate 1000 rewiring trials\n",
    "nTrials = 1000\n",
    "configurationTrials = zeros(279,279,nTrials)\n",
    "fill!(configurationTrials, 3000)\n",
    "\n",
    "#find set of all edges in the graph\n",
    "degreeDistribution = Array{Int64, 1}(undef, 279)\n",
    "for i in 1:279\n",
    "    degree = 0\n",
    "    \n",
    "    for j in 1:279\n",
    "        if j != i && birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "            degree += 1\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    degreeDistribution[i] = degree\n",
    "end\n",
    "\n",
    "\n",
    "for k in 1:nTrials\n",
    "    print(k, \"  \")\n",
    "    for i in 1:279\n",
    "       configurationTrials[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "    end\n",
    "    \n",
    "    ddHolder = degreeDistribution[:]\n",
    "    \n",
    "    errorCounter = 0\n",
    "    \n",
    "   while sum(ddHolder) > 0\n",
    "        \n",
    "        initial = rand(1:279)\n",
    "        terminal = rand(1:279)\n",
    "        \n",
    "        if ddHolder[initial] != 0 && ddHolder[terminal] != 0\n",
    "            if configurationTrials[initial, terminal,k] == 3000\n",
    "                \n",
    "                max = maximum([initial, terminal])\n",
    "                ddHolder[initial] -= 1\n",
    "                ddHolder[terminal] -= 1\n",
    "                configurationTrials[initial, terminal,k] = labelsTimesPositionsType[max, 2]\n",
    "                configurationTrials[terminal, initial,k] = labelsTimesPositionsType[max, 2]\n",
    "            end\n",
    "            \n",
    "        end\n",
    "        \n",
    "        errorCounter += 1\n",
    "        \n",
    "        if errorCounter == 10000\n",
    "            errorCounter = 0\n",
    "            fill!(configurationTrials[:,:,k], 3000)\n",
    "            ddHolder = degreeDistribution[:]\n",
    "            \n",
    "            for i in 1:279\n",
    "               configurationTrials[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "print(\"output: configurationTrials, nTrials\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine persistent homology of spatial growth\n",
    "\n",
    "begin\n",
    "randomCurves = pmap(i -> computeBarcodesNullModels(configurationTrials[:,:,i], times), 1:nTrials)\n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "rewiredBettiCurves = zeros(numValues,3,nTrials)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    configurationBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "            \n",
    "    \n",
    "println(\"output: configurationBettiCurves\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of nTrials copies\n",
    "bettiMeans = zeros(numValues,3)\n",
    "bettiStd = zeros(numValues,3)\n",
    "\n",
    "for i in 1:numValues\n",
    "    bettiMeans[i,1] = mean(configurationBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(configurationBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(configurationBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(configurationBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(configurationBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(configurationBettiCurves[i,3,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Configurationj\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0, 20))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "\n",
    "plot(p1,p2, layout = (2,1), size = (800,800))\n",
    "\n",
    "savefig(\"configurationTrials.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": "5d83d9ce-a431-48f1-a173-dd2eed910831",
   "lastKernelId": "28baed09-383e-4905-acba-f8382fcbba90"
  },
  "kernelspec": {
   "display_name": "Julia 1.0.5",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
