{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load neccesary packages\n",
    "using CSV\n",
    "using DataFrames\n",
    "using Plots\n",
    "using Statistics\n",
    "using StatsPlots\n",
    "using RDatasets\n",
    "using Combinatorics\n",
    "using HypothesisTests\n",
    "using StatsBase\n",
    "using VegaLite\n",
    "using Distances\n",
    "using Distributed\n",
    "addprocs(7)\n",
    "using Eirene   \n",
    "@everywhere using SharedArrays\n",
    "\n",
    "\n",
    "@everywhere push!(LOAD_PATH, $\"E:functions/\") \n",
    "@everywhere using cElegansFunctions\n",
    "@everywhere using RCall\n",
    "\n",
    "\n",
    "@everywhere R\"library(TDA)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load neccesary data files\n",
    "positions = CSV.read(\"From Gene Expression File//279positions.csv\", copycols = true, header = false)\n",
    "positionsLabels = CSV.read(\"From Gene Expression File//279positionsLabels.csv\", copycols = true, header = false)\n",
    "birthTimes = CSV.read(\"From Gene Expression File//279birthTimes.csv\", copycols = true, header = false)\n",
    "birthTimesLabels = CSV.read(\"From Gene Expression File//279birthTimesLabels.csv\", copycols = true, header = false)\n",
    "connectionList = CSV.read(\"From Gene Expression File//NeuronConnect_corrected.csv\", copycols = true, header = false)\n",
    "labeledTypeList = CSV.read(\"From Gene Expression File//279labeledTypes.csv\", copycols = true, header = false)\n",
    "zeroDimTopSim = CSV.read(\"From Gene Expression File//dimensionZeroSimilarity.csv\", copycols = true, header = false)\n",
    "oneDimTopSim = CSV.read(\"From Gene Expression File//dimensionOneSimilarity.csv\", copycols = true, header = false)\n",
    "twoDimTopSim = CSV.read(\"From Gene Expression File//dimensionTwoSimilarity.csv\", copycols = true, header = false)\n",
    "zeroDimTopSimGD = CSV.read(\"From Gene Expression File//dimensionZeroMixedPairSimilarity.csv\", copycols = true, header = false)\n",
    "oneDimTopSimGD = CSV.read(\"From Gene Expression File//dimensionOneMixedPairSimilarity.csv\", copycols = true, header = false)\n",
    "twoDimTopSimGD = CSV.read(\"From Gene Expression File//dimensionTwoMixedPairSimilarity.csv\", copycols = true, header = false)\n",
    "\n",
    "\n",
    "#done\n",
    "println(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#process all the data into a single array sorted by birth time\n",
    "\n",
    "#creates an array for positions with their labels\n",
    "positionsMatrix = Array{Any, 2}(UndefInitializer(),279,3)\n",
    "\n",
    "for i in 1:279\n",
    "    positionsMatrix[i,1] = positionsLabels[1][i]\n",
    "    positionsMatrix[i,2] = positions[i,1]\n",
    "    positionsMatrix[i,3] = positions[i,2]\n",
    "end\n",
    "\n",
    "#creates an array for positions with their labels\n",
    "birthTimesMatrix = Array{Any, 2}(UndefInitializer(),279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    birthTimesMatrix[i,1] = birthTimesLabels[1][i]\n",
    "    birthTimesMatrix[i,2] = birthTimes[1][i]\n",
    "end\n",
    "\n",
    "#creates an array for types with their labels\n",
    "typeMatrix = Array{Any, 2}(UndefInitializer(), 279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    typeMatrix[i,1] = labeledTypeList[1][i]\n",
    "    typeMatrix[i,2] = labeledTypeList[2][i]\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#sorts all the above matricies by neuron label to make them compatible\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && positionsMatrix[i - movement, 1] < positionsMatrix[i-movement - 1, 1]\n",
    "        shadowCache = positionsMatrix[i - movement, :]\n",
    "        positionsMatrix[i - movement, :] = positionsMatrix[i - movement - 1, :]\n",
    "        positionsMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "    \n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && birthTimesMatrix[i - movement, 1] < birthTimesMatrix[i-movement - 1, 1]\n",
    "        shadowCache = birthTimesMatrix[i - movement, :]\n",
    "        birthTimesMatrix[i - movement, :] = birthTimesMatrix[i - movement - 1, :]\n",
    "        birthTimesMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "    \n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && typeMatrix[i - movement, 1] < typeMatrix[i-movement - 1, 1]\n",
    "        shadowCache = typeMatrix[i - movement, :]\n",
    "        typeMatrix[i - movement, :] = typeMatrix[i - movement - 1, :]\n",
    "        typeMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#stores all these data points into a single concise matrix\n",
    "labelsTimesPositionsType = Array{Any, 2}(UndefInitializer(),279,5)\n",
    "\n",
    "labelsTimesPositionsType[:,1:2] = birthTimesMatrix[:,:]\n",
    "labelsTimesPositionsType[:,3:4] = positionsMatrix[:,2:3]\n",
    "labelsTimesPositionsType[:,5] = typeMatrix[:,2]\n",
    "\n",
    "#alter the positions isometrically to look nicer \n",
    "labelsTimesPositionsType[:,3] *= -1\n",
    "labelsTimesPositionsType[:,4] *= -1\n",
    "\n",
    "\n",
    "#sorts new matrix by birth time\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && labelsTimesPositionsType[i - movement, 2] < labelsTimesPositionsType[i-movement - 1, 2]\n",
    "        shadowCache = labelsTimesPositionsType[i - movement, :]\n",
    "        labelsTimesPositionsType[i - movement, :] = labelsTimesPositionsType[i - movement - 1, :]\n",
    "        labelsTimesPositionsType[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#processes connection list and stores that in its own matrix\n",
    "\n",
    "connectionBase = Array{Any, 2}(UndefInitializer(),6417,4)\n",
    "\n",
    "for i in 1:6417\n",
    "    for j in 1:4\n",
    "        connectionBase[i,j] = connectionList[i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#create adjacency matrix sorted by birth time weighted for a node filtered persistent homology calculation\n",
    "birthSortedAdjacencyMatrix = zeros(279,279)\n",
    "       \n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        birthSortedAdjacencyMatrix[i,j] = 3000\n",
    "    end\n",
    "end\n",
    "\n",
    "for k in 1:6417\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if connectionBase[k,1] == labelsTimesPositionsType[i,1] && connectionBase[k,2] == labelsTimesPositionsType[j,1]\n",
    "                max = maximum([labelsTimesPositionsType[i,2],labelsTimesPositionsType[j,2]])\n",
    "                birthSortedAdjacencyMatrix[i,j] = max\n",
    "            end\n",
    "            if connectionBase[k,1] == labelsTimesPositionsType[j,1] && connectionBase[k,2] == labelsTimesPositionsType[i,1]\n",
    "                max = maximum([labelsTimesPositionsType[i,2],labelsTimesPositionsType[j,2]])\n",
    "                birthSortedAdjacencyMatrix[i,j] = max\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "#add all the mostly trivial self loops, which are neccesary for dimension 0 calculations\n",
    "for i in 1:279\n",
    "    birthSortedAdjacencyMatrix[i,i] = labelsTimesPositionsType[i,2]\n",
    "end       \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#construct an array that is a list of times at which neurons are born, and how many born at that time. also constructs a list providing the number on this list for each neuron\n",
    "birthOrder = Array{Int64, 1}(undef, 279)\n",
    "times = Array{Float64, 2}(undef, 2,2)\n",
    "times[1,:] = [labelsTimesPositionsType[1,2],1]\n",
    "birthOrder[1] = 1\n",
    "times[2,:] = [labelsTimesPositionsType[2,2],1]\n",
    "birthOrder[2] = 2\n",
    "birthOrderCounter = 2\n",
    "\n",
    "for i in 3:279\n",
    "    if labelsTimesPositionsType[i,2] != times[end,1]\n",
    "        times = cat(dims = 1, times[:,:], transpose([labelsTimesPositionsType[i,2],1]))\n",
    "        birthOrderCounter += 1\n",
    "    else\n",
    "        times[end,2] += 1\n",
    "    end\n",
    "    birthOrder[i] = birthOrderCounter\n",
    "end\n",
    "\n",
    "\n",
    "#determine number of points on x-axis\n",
    "numValues = size(times,1)\n",
    "\n",
    "#end step\n",
    "println(\"output: labelsTimesPositionsType, birthSortedAdjacencyMatrix, birthOrder, numValues, and times\")\n",
    "\n",
    "#cleanup\n",
    "positionsMatrix = nothing\n",
    "positions = nothing\n",
    "positionsLabels = nothing\n",
    "birthTimes = nothing\n",
    "birthTimesLabels = nothing\n",
    "connectionList = nothing\n",
    "labeledTypeList = nothing\n",
    "birthTimesMatrix = nothing\n",
    "typeMatrix = nothing\n",
    "connectionBase = nothing\n",
    "birthOrderCounter = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes the betti curves in dimensions 0-3, and birth/death rates for each also\n",
    "eireneMatrix = eirene(birthSortedAdjacencyMatrix, model = \"vr\", minrad = 0.5, maxrad = 3001, maxdim = 3)\n",
    " \n",
    "\n",
    "\n",
    "\n",
    "#make arrays to store betti curves\n",
    "BettiCurves = zeros(numValues,4)\n",
    "BettiRates = zeros(numValues,3,2)\n",
    "barcodes = barcode(eireneMatrix, dim=0)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "\n",
    "#calculate and store dimension 0 betti values\n",
    "for i in 1:numValues\n",
    "    \n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "            BettiRates[i,1,1] += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "            BettiRates[i,1,2] += 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,1] = bettiNumber\n",
    "end\n",
    "\n",
    "\n",
    "#calculate and store dimension 1 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=1)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "            BettiRates[i,2,1] += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "            BettiRates[i,2,2] += 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,2] = bettiNumber\n",
    "end\n",
    "\n",
    "#calculate and store dimension 2 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=2)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "            BettiRates[i,3,1] += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "            BettiRates[i,3,2] += 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,3] = bettiNumber\n",
    "end\n",
    "\n",
    "\n",
    "#calculate and store dimension 3 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=3)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "for i in 1:numValues\n",
    "        \n",
    "\n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,4] = bettiNumber\n",
    "end\n",
    "\n",
    "#end step\n",
    "println(\"output: numValues, BettiCurves\")\n",
    "\n",
    "#cleanup\n",
    "barcodes = nothing\n",
    "bettiNumber = nothing\n",
    "amount = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot of betti curves\n",
    "\n",
    "\n",
    "#determine percent of neurons born with time\n",
    "percentBorn = zeros(numValues)\n",
    "percentCounter = 1\n",
    "percentBorn[percentCounter] = 1/2.79\n",
    "for i in 2:279\n",
    "    if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[i-1,2]\n",
    "        percentCounter += 1\n",
    "        percentBorn[percentCounter] += percentBorn[percentCounter-1]\n",
    "    end\n",
    "    percentBorn[percentCounter] += 1/2.79  \n",
    "end\n",
    "\n",
    "#determine edge density with time\n",
    "edgeDensity = zeros(numValues,2)\n",
    "edgeCounter = 0\n",
    "\n",
    "for t in 1:numValues\n",
    "    edgeCounter += trunc(Int, times[t,2])\n",
    "    \n",
    "    numberEdges = 0\n",
    "\n",
    "    for i in 1:edgeCounter\n",
    "        for j in 1:i-1\n",
    "            if birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "                numberEdges += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    edgeDensity[t,2] = numberEdges/binomial(edgeCounter,2)\n",
    "    edgeDensity[t,1] = times[t,1]\n",
    "end\n",
    "\n",
    "\n",
    "#do the actual plotting\n",
    "p1 = plot(times[:,1], [BettiCurves[:,2], BettiCurves[:,3]], color = [:red :blue theme(:default)], title = \"C Elegans Betti Curves\", label = [\"dim 1\" \"dim 2\" \"dim 3\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"betti number\", xlims=(0,2400))\n",
    "sticks!(1:1300, location[:], lc = :black, label = \"period of no neuron birth\")\n",
    "p2 = plot(edgeDensity[:,1], edgeDensity[:,2], lc = [:grey], line=(:dot,2), legendfont = font(4), xlabel = \"minutes\", ylabel = \"edge density\", legend = false, label = \"edge density\",  xlims=(0,2400))\n",
    "p3 = plot(times[:,1], percentBorn[:], lc = [:black], line=(:dot,2),  ylabel = \"percent of neurons born\", xlims=(0,2400), legend = false)\n",
    "p4 = plot(times[:,1], [BettiCurves[:,1], BettiCurves[:,4]], color = [:green :purple theme(:default)], title = \"C Elegans Betti Curves\", label = [\"dim 0\" \"dim 3\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"betti number\", ylims = (0,10), xlims=(0,2400))\n",
    "\n",
    "\n",
    "plot(p1,p4,p2,p3, layout = (4,1), size = (800,800))\n",
    "#savefig(\"oneThroughThreeBettiCurves.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotbarcode_pjs(eireneMatrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do the actual plotting\n",
    "p1 = plot(times[:,1], BettiRates[:,1,1], color = [ :blue theme(:default)], title = \"Dimension Zero\", label = [\"birth\" ], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "p2 = plot(times[:,1], BettiRates[:,2,1], color = [ :blue theme(:default)], title = \"Dimension One\", label = [\"birth\" ], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "p3 = plot(times[:,1], BettiRates[:,3,1], color = [ :blue theme(:default)], title = \"Dimension Two\", label = [\"birth\" ], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "\n",
    "p4 = plot(times[:,1], BettiRates[:,1,2], color = [:red  theme(:default)], title = \"Dimension Zero\", label = [\"death\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "p5 = plot(times[:,1], BettiRates[:,2,2], color = [:red  theme(:default)], title = \"Dimension One\", label = [\"death\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "p6 = plot(times[:,1], BettiRates[:,3,2], color = [:red  theme(:default)], title = \"Dimension Two\", label = [\"death\"], legend=:bottomright , legendfont = font(4),  xlabel = \"minutes\", ylabel = \"number of cavities\", xlims=(0,2400))\n",
    "\n",
    "\n",
    "plot(p1,p4,p2,p5,p3,p6, layout = (3,2), size = (1400,800))\n",
    "#savefig(\"oneThroughTwoBettiRates.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#post-plotting cleanup\n",
    "location = nothing\n",
    "percentBorn = nothing\n",
    "edgeDensity = nothing\n",
    "numberEdges = nothing\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "p3 = nothing\n",
    "p4 = nothing\n",
    "eireneMatrix = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make 1000 models by varying birth times and adding random edges\n",
    "numTrials = 1000\n",
    "randomModelsWithTimes = zeros(279,280,numTrials)\n",
    "\n",
    "\n",
    "\n",
    "#load all the models with the base\n",
    "for k in 1:numTrials\n",
    "    randomModelsWithTimes[:,1:279,k] = birthSortedAdjacencyMatrix[:,:]\n",
    "\n",
    "\n",
    "    #reconstruct each as a binary matrix to ease calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                randomModelsWithTimes[i,j,k] = 1\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 3000\n",
    "                randomModelsWithTimes[i,j,k] = 0\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "\n",
    "    #for each model, add 810 random connections\n",
    "    for counter in 1:810\n",
    "        initial = rand(1:279)\n",
    "        terminal = rand(1:279)\n",
    "        while terminal == initial\n",
    "            terminal = rand(1:279)\n",
    "        end\n",
    "        \n",
    "        randomModelsWithTimes[initial,terminal,k] = 1\n",
    "        randomModelsWithTimes[terminal,initial,k] = 1\n",
    "    end\n",
    "\n",
    "\n",
    "    #alter the birth times of each neuron\n",
    "    bornTimes = labelsTimesPositionsType[:,2]\n",
    "\n",
    "    for i in 1:201\n",
    "        sign = rand()\n",
    "    \n",
    "        if sign < 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.1\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] += error\n",
    "        end\n",
    "    \n",
    "        if sign >= 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.1\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] -= error\n",
    "        end\n",
    "    end\n",
    "\n",
    "    for i in 202:279\n",
    "        sign = rand()\n",
    "    \n",
    "        if sign < 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.02\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] += error\n",
    "        end\n",
    "    \n",
    "        if sign >= 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.02\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] -= error\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #reorder the current matrix by birth time\n",
    "    for i in 1:279\n",
    "        movement = 0\n",
    "    \n",
    "        while i - movement > 1 && bornTimes[i - movement] < bornTimes[i-movement - 1]\n",
    "            shadowCache = bornTimes[i - movement]\n",
    "            bornTimes[i - movement] = bornTimes[i - movement - 1]\n",
    "            bornTimes[i - movement - 1] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[i - movement, :,k]\n",
    "            randomModelsWithTimes[i - movement, :,k] = randomModelsWithTimes[i - movement - 1, :,k]\n",
    "            randomModelsWithTimes[i - movement - 1, :,k] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[:,i - movement, k]\n",
    "            randomModelsWithTimes[:,i - movement, k] = randomModelsWithTimes[:,i - movement - 1, k]\n",
    "            randomModelsWithTimes[:,i - movement - 1, k] = shadowCache\n",
    "        \n",
    "            movement += 1\n",
    "        end\n",
    "    end\n",
    "    randomModelsWithTimes[:,280,k] = bornTimes[:]\n",
    "    \n",
    "    #re-label the random models for eirene calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] == 1\n",
    "                randomModelsWithTimes[i,j,k] = maximum([bornTimes[i], bornTimes[j]])\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 0\n",
    "                randomModelsWithTimes[i,j,k] = 3000\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "println(\"output: randomModelsWithTimes, numTrials\")\n",
    "\n",
    "#cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "begin\n",
    "#perform computations\n",
    "randomCurves = pmap(i -> computeBarcodes(randomModelsWithTimes[:,1:279,i], randomModelsWithTimes[:,280,i]), 1:numTrials)   \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "RandomizedBettiCurves = zeros(279,3,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    RandomizedBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "\n",
    "\n",
    "edgeDensities = zeros(279, numTrials)\n",
    "for k in 1:numTrials\n",
    "    numberEdges = 0\n",
    "    \n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                numberEdges += 1\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        edgeDensities[i,k] += (numberEdges/binomial(i,2))\n",
    "    end\n",
    "end \n",
    "            \n",
    "    \n",
    "println(\"output: RandomizedBettiCurves, edgeDensities\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(279,3)\n",
    "bettiStd = zeros(279,3)\n",
    "averageTimes = zeros(279)\n",
    "edgeDensityMeans = zeros(279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    bettiMeans[i,1] = mean(RandomizedBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(RandomizedBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(RandomizedBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(RandomizedBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(RandomizedBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(RandomizedBettiCurves[i,3,:])\n",
    "    \n",
    "    averageTimes[i] = mean(randomModelsWithTimes[i,280,:])\n",
    "    \n",
    "    edgeDensityMeans[i,1] = mean(edgeDensities[i,:])\n",
    "    edgeDensityMeans[i,2] = std(edgeDensities[i,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd, averageTimes, edgeDensityMeans\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(averageTimes[:], bettiMeans[:,2:3], xlims = (0,2400), ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:topleft , title = \"Additional Connections and Altered Birth Times\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(averageTimes[:], bettiMeans[:,1], xlims = (0,2400), ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:topright , title = \"Additional Connections and Altered Birth Times\", label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), color = [:green theme(:default)], label = [ \"original dim 0\"])\n",
    "p3 = plot(averageTimes[:,1], edgeDensityMeans[:,1], color = [:gray], ribbon = edgeDensityMeans[:,2], line=(:dot,2),  ylabel = \"edge density\", xlims=(0,2400))\n",
    "\n",
    "plot(p1,p2,p3, layout = (3,1), size = (800,800))\n",
    "\n",
    "#savefig(\"alteredOne.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find birth rate for each dimension also\n",
    "begin\n",
    "#perform computations\n",
    "randomRates = pmap(i -> computeRates(randomModelsWithTimes[:,1:279,i], randomModelsWithTimes[:,280,i]), 1:numTrials)\n",
    "\n",
    "        \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "RandomizedBettiRates = zeros(279,3,2,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    RandomizedBettiRates[:,:,:,k] = randomRates[k]\n",
    "end\n",
    "\n",
    "        \n",
    "    \n",
    "println(\"output: RandomizedBettiCurves\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(279,3,2)\n",
    "bettiStd = zeros(279,3,2)\n",
    "\n",
    "\n",
    "for i in 1:279\n",
    "    bettiMeans[i,1,1] = mean(RandomizedBettiRates[i,1,1,:])\n",
    "    bettiMeans[i,2,1] = mean(RandomizedBettiRates[i,2,1,:])\n",
    "    bettiMeans[i,3,1] = mean(RandomizedBettiRates[i,3,1,:])\n",
    "    bettiMeans[i,1,2] = mean(RandomizedBettiRates[i,1,2,:])\n",
    "    bettiMeans[i,2,2] = mean(RandomizedBettiRates[i,2,2,:])\n",
    "    bettiMeans[i,3,2] = mean(RandomizedBettiRates[i,3,2,:])\n",
    "    \n",
    "    bettiStd[i,1,1] = std(RandomizedBettiRates[i,1,1,:])\n",
    "    bettiStd[i,2,1] = std(RandomizedBettiRates[i,2,1,:])\n",
    "    bettiStd[i,3,1] = std(RandomizedBettiRates[i,3,1,:])\n",
    "    bettiStd[i,1,2] = std(RandomizedBettiRates[i,1,2,:])\n",
    "    bettiStd[i,2,2] = std(RandomizedBettiRates[i,2,2,:])\n",
    "    bettiStd[i,3,2] = std(RandomizedBettiRates[i,3,2,:])\n",
    "\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(averageTimes[:], bettiMeans[:,1,1], xlims = (0,2400), ribbon = bettiStd[:,1,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 0\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,1,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "p2 = plot(averageTimes[:], bettiMeans[:,2,1], xlims = (0,2400), ribbon = bettiStd[:,2,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 1\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,2,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "p3 = plot(averageTimes[:], bettiMeans[:,3,1], xlims = (0,2400), ribbon = bettiStd[:,3,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 2\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,3,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "\n",
    "p4 = plot(averageTimes[:], bettiMeans[:,1,2], xlims = (0,2400), ribbon = bettiStd[:,1,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 0\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,1,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "p5 = plot(averageTimes[:], bettiMeans[:,2,2], xlims = (0,2400), ribbon = bettiStd[:,2,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 1\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,2,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "p6 = plot(averageTimes[:], bettiMeans[:,3,2], xlims = (0,2400), ribbon = bettiStd[:,3,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 2\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,3,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "\n",
    "plot(p1,p4,p2,p5,p3,p6, layout = (3,2), size = (1400,800))\n",
    "\n",
    "#savefig(\"alteredOneRates.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#post plot cleanup\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "bettiMeans = nothing\n",
    "bettiStd = nothing\n",
    "averageTimes = nothing\n",
    "edgeDensities = nothing\n",
    "edgeDensityMeans = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make 1000 models by varying birth times and adding random edges (make average wiring cost is within +/- 15%)\n",
    "numTrials = 1000\n",
    "bound = 0.15\n",
    "randomModelsWithTimes = zeros(279,280,numTrials)\n",
    "\n",
    "\n",
    "\n",
    "#find the latent average wiring cost\n",
    "latentWiringCostInfo = zeros(2)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "            a = labelsTimesPositionsType[i,3]\n",
    "            b = labelsTimesPositionsType[i,4]\n",
    "            c = labelsTimesPositionsType[j,3]\n",
    "            d = labelsTimesPositionsType[j,4]\n",
    "            latentWiringCostInfo[1] += evaluate(Euclidean(),[a,b], [c,d])\n",
    "            latentWiringCostInfo[2] += 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "            \n",
    "#find 100 sets of 810 that are within bound% of wiring cost\n",
    "begin\n",
    "\n",
    "newConnections = pmap(i -> findWiringWithinBounds(labelsTimesPositionsType[:,3:4], bound, latentWiringCostInfo), 1:numTrials)\n",
    "        \n",
    "end\n",
    "connections = zeros(810,2,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    connections[:,:,k] = newConnections[k]\n",
    "end\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#load all the models with the base\n",
    "for k in 1:numTrials\n",
    "    \n",
    "    randomModelsWithTimes[:,1:279,k] = birthSortedAdjacencyMatrix[:,:]\n",
    "    \n",
    "    \n",
    "    #reconstruct each as a binary matrix to ease calculations \n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                randomModelsWithTimes[i,j,k] = 1\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 3000\n",
    "                randomModelsWithTimes[i,j,k] = 0\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    \n",
    "    for i in 1:180\n",
    "        randomModelsWithTimes[trunc(Int,connections[i,1,k]),trunc(Int,connections[i,2,k]),k] = 1\n",
    "        randomModelsWithTimes[trunc(Int,connections[i,2,k]),trunc(Int,connections[i,1,k]),k] = 1\n",
    "    end\n",
    "\n",
    "    #alter the birth times of each neuron\n",
    "    bornTimes = labelsTimesPositionsType[:,2]\n",
    "\n",
    "    for i in 1:201\n",
    "        sign = rand()\n",
    "    \n",
    "        if sign < 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.1\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] += error\n",
    "            \n",
    "        end\n",
    "    \n",
    "        if sign >= 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.1\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] -= error\n",
    "            \n",
    "        end\n",
    "    end\n",
    "\n",
    "    for i in 202:279\n",
    "        sign = rand()\n",
    "    \n",
    "        if sign < 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.02\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] += error\n",
    "           \n",
    "        end\n",
    "    \n",
    "        if sign >= 0.5\n",
    "            error = rand()\n",
    "            percentError = error*0.02\n",
    "            error = bornTimes[i] * percentError\n",
    "            bornTimes[i] -= error\n",
    "            \n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #reorder the current matrix by birth time\n",
    "    for i in 1:279\n",
    "        movement = 0\n",
    "    \n",
    "        while i - movement > 1 && bornTimes[i - movement] < bornTimes[i-movement - 1]\n",
    "            shadowCache = bornTimes[i - movement]\n",
    "            bornTimes[i - movement] = bornTimes[i - movement - 1]\n",
    "            bornTimes[i - movement - 1] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[i - movement, :,k]\n",
    "            randomModelsWithTimes[i - movement, :,k] = randomModelsWithTimes[i - movement - 1, :,k]\n",
    "            randomModelsWithTimes[i - movement - 1, :,k] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[:,i - movement, k]\n",
    "            randomModelsWithTimes[:,i - movement, k] = randomModelsWithTimes[:,i - movement - 1, k]\n",
    "            randomModelsWithTimes[:,i - movement - 1, k] = shadowCache\n",
    "        \n",
    "            movement += 1\n",
    "        end\n",
    "    end\n",
    "    randomModelsWithTimes[:,280,k] = bornTimes[:]\n",
    "    \n",
    "    #re-label the random models for eirene calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] == 1\n",
    "                randomModelsWithTimes[i,j,k] = maximum([bornTimes[i], bornTimes[j]])\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 0\n",
    "                randomModelsWithTimes[i,j,k] = 3000\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "           \n",
    "    \n",
    "end\n",
    "println(\"output: randomModelsWithTimes, numTrials\")\n",
    "\n",
    "#cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "begin\n",
    "#perform computations\n",
    "randomCurves = pmap(i -> computeBarcodes(randomModelsWithTimes[:,1:279,i], randomModelsWithTimes[:,280,i]), 1:numTrials)\n",
    "\n",
    "        \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "RandomizedBettiCurves = zeros(279,3,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    RandomizedBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "\n",
    "\n",
    "edgeDensities = zeros(279, numTrials)\n",
    "for k in 1:numTrials\n",
    "    numberEdges = 0\n",
    "    \n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                numberEdges += 1\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        edgeDensities[i,k] += (numberEdges/binomial(i,2))\n",
    "    end\n",
    "end \n",
    "      \n",
    " \n",
    "    \n",
    "println(\"output: RandomizedBettiCurves, edgeDensities\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(279,3)\n",
    "bettiStd = zeros(279,3)\n",
    "averageTimes = zeros(279)\n",
    "edgeDensityMeans = zeros(279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    bettiMeans[i,1] = mean(RandomizedBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(RandomizedBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(RandomizedBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(RandomizedBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(RandomizedBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(RandomizedBettiCurves[i,3,:])\n",
    "    \n",
    "    averageTimes[i] = mean(randomModelsWithTimes[i,280,:])\n",
    "    \n",
    "    edgeDensityMeans[i,1] = mean(edgeDensities[i,:])\n",
    "    edgeDensityMeans[i,2] = std(edgeDensities[i,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd, averageTimes, edgeDensityMeans\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(averageTimes[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Additional Connections within 15% Average Wiring Cost\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(averageTimes[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , title = \"Additional Connections within 15% Average Wiring Cost\", label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "p3 = plot(averageTimes[:,1], edgeDensityMeans[:,1], color = [:gray], ribbon = edgeDensityMeans[:,2], line=(:dot,2),  ylabel = \"edge density\", xlims=(0,2400))\n",
    "\n",
    "plot(p1,p2,p3, layout = (3,1), size = (800,800))\n",
    "\n",
    "#savefig(\"within15.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find birth rate for each dimension also\n",
    "begin\n",
    "#perform computations\n",
    "randomRates = pmap(i -> computeRates(randomModelsWithTimes[:,1:279,i], randomModelsWithTimes[:,280,i]), 1:1000)\n",
    "\n",
    "        \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "RandomizedBettiRates = zeros(279,3,2,1000)\n",
    "\n",
    "for k in 1:1000\n",
    "    RandomizedBettiRates[:,:,:,k] = randomRates[k]\n",
    "end\n",
    "\n",
    "        \n",
    "    \n",
    "println(\"output: RandomizedBettiCurves\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(279,3,2)\n",
    "bettiStd = zeros(279,3,2)\n",
    "\n",
    "\n",
    "for i in 1:279\n",
    "    bettiMeans[i,1,1] = mean(RandomizedBettiRates[i,1,1,:])\n",
    "    bettiMeans[i,2,1] = mean(RandomizedBettiRates[i,2,1,:])\n",
    "    bettiMeans[i,3,1] = mean(RandomizedBettiRates[i,3,1,:])\n",
    "    bettiMeans[i,1,2] = mean(RandomizedBettiRates[i,1,2,:])\n",
    "    bettiMeans[i,2,2] = mean(RandomizedBettiRates[i,2,2,:])\n",
    "    bettiMeans[i,3,2] = mean(RandomizedBettiRates[i,3,2,:])\n",
    "    \n",
    "    bettiStd[i,1,1] = std(RandomizedBettiRates[i,1,1,:])\n",
    "    bettiStd[i,2,1] = std(RandomizedBettiRates[i,2,1,:])\n",
    "    bettiStd[i,3,1] = std(RandomizedBettiRates[i,3,1,:])\n",
    "    bettiStd[i,1,2] = std(RandomizedBettiRates[i,1,2,:])\n",
    "    bettiStd[i,2,2] = std(RandomizedBettiRates[i,2,2,:])\n",
    "    bettiStd[i,3,2] = std(RandomizedBettiRates[i,3,2,:])\n",
    "\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(averageTimes[:], bettiMeans[:,1,1], xlims = (0,2400), ribbon = bettiStd[:,1,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 0\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,1,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "p2 = plot(averageTimes[:], bettiMeans[:,2,1], xlims = (0,2400), ribbon = bettiStd[:,2,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 1\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,2,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "p3 = plot(averageTimes[:], bettiMeans[:,3,1], xlims = (0,2400), ribbon = bettiStd[:,3,1], color = [:red theme(:default)], legend=:topleft , title = \"Dimension 2\", label = [\"birth\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,3,1], line=(:dot,2), color = [:red theme(:default)], label = [ \"original\"])\n",
    "\n",
    "p4 = plot(averageTimes[:], bettiMeans[:,1,2], xlims = (0,2400), ribbon = bettiStd[:,1,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 0\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,1,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "p5 = plot(averageTimes[:], bettiMeans[:,2,2], xlims = (0,2400), ribbon = bettiStd[:,2,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 1\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,2,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "p6 = plot(averageTimes[:], bettiMeans[:,3,2], xlims = (0,2400), ribbon = bettiStd[:,3,2], color = [:blue theme(:default)], legend=:topleft , title = \"Dimension 2\", label = [\"death\"],  xlabel = \"time\", ylabel = \"number of cavities\")\n",
    "plot!(times[:,1], BettiRates[:,3,2], line=(:dot,2), color = [:blue theme(:default)], label = [ \"original\"])\n",
    "\n",
    "plot(p1,p4,p2,p5,p3,p6, layout = (3,2), size = (1400,800))\n",
    "\n",
    "#savefig(\"within15Rates.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#post plot cleanup\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "bettiMeans = nothing\n",
    "bettiStd = nothing\n",
    "edgeDensities = nothing\n",
    "edgeDensityMeans = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check to see if betti deltas correlate with biological categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make a 279x2 array, where [i,1] is the first neuron with the same birth time as i, and [i,2] is the last such\n",
    "firstAndLast = Array{Int64, 2}(undef,279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    foundFirst = false\n",
    "    foundLast = false\n",
    "    for j in 1:279\n",
    "        if labelsTimesPositionsType[j,2] == labelsTimesPositionsType[i,2]\n",
    "            if !foundFirst\n",
    "                foundFirst = true\n",
    "                firstAndLast[i,1] = j\n",
    "            end\n",
    "        else\n",
    "            if foundFirst && !foundLast\n",
    "                firstAndLast[i,2] = j-1\n",
    "                foundLast = true\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        if j == 279 && !foundLast\n",
    "            firstAndLast[i,2] = j\n",
    "        end\n",
    "    end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function randomLinearOrdering()\n",
    "    presentTimes = labelsTimesPositionsType[:,2]\n",
    "    \n",
    "    currentNeuron = 1\n",
    "    \n",
    "    while currentNeuron <= 279\n",
    "        n =  firstAndLast[currentNeuron, 2] - firstAndLast[currentNeuron, 1] + 1\n",
    "        \n",
    "        if n != 1\n",
    "            vector = Array{Int, 1}(undef, n)\n",
    "            for i in 1:n\n",
    "                vector[i] = i\n",
    "            end\n",
    "            p = nthperm(vector, rand(1:factorial(n)))\n",
    "        \n",
    "            for i in 1:n\n",
    "                presentTimes[currentNeuron] += p[i]/(n*100 + 1)\n",
    "                currentNeuron += 1\n",
    "            end\n",
    "            \n",
    "        else\n",
    "            currentNeuron += 1\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    return presentTimes\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make 1000 models by varying birth times\n",
    "numTrials = 1000\n",
    "randomModelsWithTimes = zeros(279,280,numTrials)\n",
    "\n",
    "\n",
    "\n",
    "#load all the models with the base\n",
    "for k in 1:numTrials\n",
    "    randomModelsWithTimes[:,1:279,k] = birthSortedAdjacencyMatrix[:,:]\n",
    "\n",
    "\n",
    "    #reconstruct each as a binary matrix to ease calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] < 3000\n",
    "                randomModelsWithTimes[i,j,k] = 1\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 3000\n",
    "                randomModelsWithTimes[i,j,k] = 0\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "\n",
    "\n",
    "    #alter the birth times of each neuron\n",
    "    bornTimes = randomLinearOrdering()\n",
    "    \n",
    "    #reorder the current matrix by birth time\n",
    "    for i in 1:279\n",
    "        movement = 0\n",
    "    \n",
    "        while i - movement > 1 && bornTimes[i - movement] < bornTimes[i-movement - 1]\n",
    "            shadowCache = bornTimes[i - movement]\n",
    "            bornTimes[i - movement] = bornTimes[i - movement - 1]\n",
    "            bornTimes[i - movement - 1] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[i - movement, :,k]\n",
    "            randomModelsWithTimes[i - movement, :,k] = randomModelsWithTimes[i - movement - 1, :,k]\n",
    "            randomModelsWithTimes[i - movement - 1, :,k] = shadowCache\n",
    "            \n",
    "            shadowCache = randomModelsWithTimes[:,i - movement, k]\n",
    "            randomModelsWithTimes[:,i - movement, k] = randomModelsWithTimes[:,i - movement - 1, k]\n",
    "            randomModelsWithTimes[:,i - movement - 1, k] = shadowCache\n",
    "        \n",
    "            movement += 1\n",
    "        end\n",
    "    end\n",
    "    randomModelsWithTimes[:,280,k] = bornTimes[:]\n",
    "    \n",
    "    #re-label the random models for eirene calculations\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if randomModelsWithTimes[i,j,k] == 1\n",
    "                randomModelsWithTimes[i,j,k] = maximum([bornTimes[i], bornTimes[j]])\n",
    "            end\n",
    "            if randomModelsWithTimes[i,j,k] == 0\n",
    "                randomModelsWithTimes[i,j,k] = 3000\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "println(\"output: randomModelsWithTimes, numTrials\")\n",
    "\n",
    "#cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#array to hold average + and - to each dimension of betti curve per node\n",
    "bettiDeltas = zeros(279,3,2,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "print(k,\" \")\n",
    "#computes the betti curves in dimensions 0-2\n",
    "eireneMatrix = eirene(randomModelsWithTimes[:,1:279,k], model = \"vr\", minrad = 0.5, maxrad = 3001, maxdim = 2)\n",
    " \n",
    "\n",
    "#determine number of points on x-axis\n",
    "numValues = 279\n",
    "\n",
    "\n",
    "\n",
    "#make arrays to store betti curves\n",
    "barcodes = barcode(eireneMatrix, dim=0)\n",
    "amount = size(barcodes,1)\n",
    "\n",
    "\n",
    "\n",
    "#calculate and store dimension 0 betti deltas\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if abs(barcodes[j,1] - barcodes[j,2]) >= 0.01\n",
    "            if barcodes[j,1] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,1,1,k] += 1\n",
    "            end\n",
    "            if barcodes[j,2] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,1,2,k] += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#calculate and store dimension 1 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=1)\n",
    "amount = size(barcodes,1)\n",
    "\n",
    "\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if abs(barcodes[j,1] - barcodes[j,2]) >= 0.01\n",
    "            if barcodes[j,1] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,2,1,k] += 1\n",
    "            end\n",
    "            if barcodes[j,2] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,2,2,k] += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "#calculate and store dimension 2 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=2)\n",
    "amount = size(barcodes,1)\n",
    "\n",
    "for i in 1:numValues\n",
    "    for j in 1:amount\n",
    "        if abs(barcodes[j,1] - barcodes[j,2]) >= 0.01\n",
    "            if barcodes[j,1] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,3,1,k] += 1\n",
    "            end\n",
    "            if barcodes[j,2] == randomModelsWithTimes[i,280,k]\n",
    "                bettiDeltas[i,3,2,k] += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "end\n",
    "println(\"output: bettiDeltas\")\n",
    "\n",
    "#cleanup\n",
    "barcodes = nothing\n",
    "amount = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets test +/- deltas against biomarkers\n",
    "\n",
    "#variables for connections within a type\n",
    "sensoryDeltas = zeros(85*numTrials,3,2) #85 sensory neurons\n",
    "motorDeltas = zeros(115*numTrials,3,2) #115 motor neurons\n",
    "interDeltas = zeros(79*numTrials,3,2) # 79 inter neurons\n",
    "headDeltas = zeros(160*numTrials,3,2) #160 head neurons\n",
    "bodyDeltas = zeros(92*numTrials,3,2) #92 body neurons\n",
    "tailDeltas = zeros(27*numTrials,3,2) #27 tail neurons\n",
    "sensoryCounter = 1\n",
    "motorCounter = 1\n",
    "interCounter = 1\n",
    "headCounter = 1\n",
    "bodyCounter = 1\n",
    "tailCounter = 1\n",
    "\n",
    "\n",
    "for k in 1:numTrials\n",
    "#loads all the various arrays\n",
    "for i in 1:279\n",
    "   \n",
    "    #see if the neuron is sensory\n",
    "    if labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "        \n",
    "        #load dim 0-2 + and -\n",
    "        sensoryDeltas[sensoryCounter,:,:] = bettiDeltas[birthOrder[i],:,:,k]\n",
    "        sensoryCounter += 1\n",
    "    end\n",
    "        \n",
    "     #see if the neuron is motor\n",
    "    if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "        \n",
    "        #load dim 0-2 + and -\n",
    "        motorDeltas[motorCounter,:,:] = bettiDeltas[birthOrder[i],:,:,k]\n",
    "        motorCounter += 1\n",
    "    end\n",
    "    \n",
    "    #see if the neuron is inter\n",
    "    if labelsTimesPositionsType[i,5] == \"inter\"\n",
    "        \n",
    "        #load dim 0-2 + and -\n",
    "        interDeltas[interCounter,:,:] = bettiDeltas[i,:,:,k]\n",
    "        interCounter += 1\n",
    "    end\n",
    "    \n",
    "    #see if the neuron is in the head\n",
    "    if labelsTimesPositionsType[i,3] < 0\n",
    "        headDeltas[headCounter,:,:] = bettiDeltas[i,:,:,k]\n",
    "        headCounter += 1\n",
    "    end\n",
    "    \n",
    "    #see if the neuron is in the body\n",
    "    if labelsTimesPositionsType[i,3] < 0.935 && labelsTimesPositionsType[i,3] > 0\n",
    "        bodyDeltas[bodyCounter,:,:] = bettiDeltas[i,:,:,k]\n",
    "        bodyCounter += 1\n",
    "    end\n",
    "    \n",
    "    #see if the neuron is in the tail\n",
    "    if labelsTimesPositionsType[i,3] > 0.935\n",
    "        tailDeltas[tailCounter,:,:] = bettiDeltas[i,:,:,k]\n",
    "        tailCounter += 1\n",
    "    end\n",
    "   \n",
    "            \n",
    "end\n",
    "end\n",
    "println(\"Output: sensoryDeltas and motorDeltas and interDeltas and headDeltas and bodyDeltas and tailDeltas\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "checkAgainst = zeros(92*numTrials+27*numTrials)\n",
    "checkAgainst[1:92*numTrials] = bodyDeltas[:,3,1]\n",
    "checkAgainst[92*numTrials + 1 : end] = tailDeltas[:,3,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(headDeltas[:,3,1], checkAgainst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensoryDeltas = zeros(85*numTrials,3,2) #85 sensory neurons\n",
    "motorDeltas = zeros(115*numTrials,3,2) #115 motor neurons\n",
    "interDeltas = zeros(79*numTrials,3,2) # 79 inter neurons\n",
    "headDeltas = zeros(160*numTrials,3,2) #160 head neurons\n",
    "bodyDeltas = zeros(92*numTrials,3,2) #92 body neurons\n",
    "tailDeltas = zeros(27*numTrials,3,2) #27 tail neurons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot positive deltas by type\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 12)\n",
    "plotArray[1] = sensoryDeltas[:,1,2]\n",
    "plotArray[2] = motorDeltas[:,1,2]\n",
    "plotArray[3] = interDeltas[:,1,2]\n",
    "plotArray[4] = sensoryDeltas[:,1,1]\n",
    "plotArray[5] = motorDeltas[:,1,1]\n",
    "plotArray[6] = interDeltas[:,1,1]\n",
    "plotArray[7] = headDeltas[:,1,2]\n",
    "plotArray[8] = bodyDeltas[:,1,2]\n",
    "plotArray[9] = tailDeltas[:,1,2]\n",
    "plotArray[10] = headDeltas[:,1,1]\n",
    "plotArray[11] = bodyDeltas[:,1,1]\n",
    "plotArray[12] = tailDeltas[:,1,1]\n",
    "\n",
    "#violin([ \"Sensory (-,0)\" \"Motor (-,0)\" \"Inter (-,0)\" \"Sensory (+,0)\" \"Motor (+,0)\" \"Inter (+,0)\" \"Head (-,0)\" \"Body (-,0)\" \"Tail (-,0)\" \"Head (+,0)\" \"Body (+,0)\" \"Tail (+,0)\"],plotArray[:],leg=false, #=color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)]=# title = \"Dimension Zero Betti Variation\", size = (1200, 400))\n",
    "boxplot([\"Sensory (-,0)\" \"Motor (-,0)\" \"Inter (-,0)\" \"Sensory (+,0)\" \"Motor (+,0)\" \"Inter (+,0)\" \"Head (-,0)\" \"Body (-,0)\" \"Tail (-,0)\" \"Head (+,0)\" \"Body (+,0)\" \"Tail (+,0)\"],plotArray[:],leg=false,  size = (1200, 400))\n",
    "#savefig(\"dimensionZeroBettiVariation.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot positive deltas by type\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 12)\n",
    "plotArray[1] = sensoryDeltas[:,2,2]\n",
    "plotArray[2] = motorDeltas[:,2,2]\n",
    "plotArray[3] = interDeltas[:,2,2]\n",
    "plotArray[4] = sensoryDeltas[:,2,1]\n",
    "plotArray[5] = motorDeltas[:,2,1]\n",
    "plotArray[6] = interDeltas[:,2,1]\n",
    "plotArray[7] = headDeltas[:,2,2]\n",
    "plotArray[8] = bodyDeltas[:,2,2]\n",
    "plotArray[9] = tailDeltas[:,2,2]\n",
    "plotArray[10] = headDeltas[:,2,1]\n",
    "plotArray[11] = bodyDeltas[:,2,1]\n",
    "plotArray[12] = tailDeltas[:,2,1]\n",
    "\n",
    "#violin([ \"Sensory (-,1)\" \"Motor (-,1)\" \"Inter (-,1)\" \"Sensory (+,1)\" \"Motor (+,1)\" \"Inter (+,1)\" \"Head (-,1)\" \"Body (-,1)\" \"Tail (-,1)\" \"Head (+,1)\" \"Body (+,1)\" \"Tail (+,1)\"],plotArray[:],leg=false, #=color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)]=# title = \"Dimension One Betti Variation\", size = (1200, 400))\n",
    "boxplot([\"Sensory (-,1)\" \"Motor (-,1)\" \"Inter (-,1)\" \"Sensory (+,1)\" \"Motor (+,1)\" \"Inter (+,1)\" \"Head (-,1)\" \"Body (-,1)\" \"Tail (-,1)\" \"Head (+,1)\" \"Body (+,1)\" \"Tail (+,1)\"],plotArray[:],leg=false#= color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)]=#, size = (1200, 400))\n",
    "#savefig(\"dimensionOneBettiVariation.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot positive deltas by type\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 12)\n",
    "plotArray[1] = sensoryDeltas[:,3,2]\n",
    "plotArray[2] = motorDeltas[:,3,2]\n",
    "plotArray[3] = interDeltas[:,3,2]\n",
    "plotArray[4] = sensoryDeltas[:,3,1]\n",
    "plotArray[5] = motorDeltas[:,3,1]\n",
    "plotArray[6] = interDeltas[:,3,1]\n",
    "plotArray[7] = headDeltas[:,3,2]\n",
    "plotArray[8] = bodyDeltas[:,3,2]\n",
    "plotArray[9] = tailDeltas[:,3,2]\n",
    "plotArray[10] = headDeltas[:,3,1]\n",
    "plotArray[11] = bodyDeltas[:,3,1]\n",
    "plotArray[12] = tailDeltas[:,3,1]\n",
    "\n",
    "#violin([ \"Sensory (-,2)\" \"Motor (-,2)\" \"Inter (-,2)\" \"Sensory (+,2)\" \"Motor (+,2)\" \"Inter (+,2)\" \"Head (-,2)\" \"Body (-,2)\" \"Tail (-,2)\" \"Head (+,2)\" \"Body (+,2)\" \"Tail (+,2)\"],plotArray[:],leg=false, #=color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)]=# title = \"Dimension Zero Betti Variation\", size = (1200, 400))\n",
    "boxplot([ \"Sensory (-,2)\" \"Motor (-,2)\" \"Inter (-,2)\" \"Sensory (+,2)\" \"Motor (+,2)\" \"Inter (+,2)\" \"Head (-,2)\" \"Body (-,2)\" \"Tail (-,2)\" \"Head (+,2)\" \"Body (+,2)\" \"Tail (+,2)\"],plotArray[:],leg=false, #=color = [:red :blue :green :red :red :red: :red :red :red :red :red :red theme(:dark)],=# size = (1200, 400))\n",
    "#savefig(\"dimensionTwoBettiVariation.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine the variances of each neuron\n",
    "stdDev = zeros(279,3,2)\n",
    "\n",
    "for k in 1:279\n",
    "    for i in 1:3\n",
    "        for j in 1:2\n",
    "            stdDev[k,i,j] = std(bettiDeltas[k,i,j,:])\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "posZeroStd = stdDev[:,1,1]\n",
    "posOneStd = stdDev[:,2,1]\n",
    "posTwoStd = stdDev[:,3,1]\n",
    "\n",
    "negZeroStd = variances[:,1,2]\n",
    "negOneStd = variances[:,2,2]\n",
    "negTwoStd = variances[:,3,2]\n",
    "\n",
    "\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = posZeroStd[:]\n",
    "plotArray[2] = negZeroStd[:]\n",
    "plotArray[3] = posOneStd[:]\n",
    "plotArray[4] = negOneStd[:]\n",
    "plotArray[5] = posTwoStd[:]\n",
    "plotArray[6] = negTwoStd[:]\n",
    "\n",
    "\n",
    "\n",
    "boxplot([ \"Dim Zero +\" \"Dim Zero -\" \"Dim One +\" \"Dim One -\" \"Dim Two +\" \"Dim Two -\"],plotArray[:],leg=false, size = (1200, 400))\n",
    "#savefig(\"bettiVariationStds.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = nothing\n",
    "tailDeltas = nothing\n",
    "bodyDeltas = nothing\n",
    "headDeltas = nothing\n",
    "interDeltas = nothing\n",
    "motorDeltas = nothing\n",
    "sensoryDeltas = nothing\n",
    "tailCounter = nothing\n",
    "bodyCounter = nothing\n",
    "headCounter = nothing\n",
    "interCounter = nothing\n",
    "motorCounter = nothing\n",
    "sensoryCounter = nothing\n",
    "bettiDeltas = nothing\n",
    "randomModelsWithTimes = nothing\n",
    "numTrials = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####end of betti curve analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#beginning of topological similarity analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes zeroth dimensional topological similarity\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(birthSortedAdjacencyMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 0)\n",
    "\n",
    "zeroDimBarcodes = barcode(eireneGraph, dim = 0)\n",
    "originalBarcodes = zeros(size(zeroDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(zeroDimBarcodes)[1] > 0\n",
    "    for i in 1:size(zeroDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 0\n",
    "        originalBarcodes[i,2] = zeroDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = zeroDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "zeroDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(birthSortedAdjacencyMatrix[:,:], x[1], x[2], labelsTimesPositionsType[:,2], 0)), dimension = 0)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        zeroDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "zeroDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            zeroDimemsionalTopologicalSimilarity[i,j] = 1 - (zeroDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        end\n",
    "        if labelsTimesPositionsType[i,2] == labelsTimesPositionsType[j,2]\n",
    "            zeroDimemsionalTopologicalSimilarity[i,j] = 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: zeroDimensionalBND, zeroDimemsionalTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "zeroDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes first dimensional topological similarity\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(birthSortedAdjacencyMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 1)\n",
    "\n",
    "oneDimBarcodes = barcode(eireneGraph, dim = 1)\n",
    "originalBarcodes = zeros(size(oneDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(oneDimBarcodes)[1] > 0\n",
    "    for i in 1:size(oneDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 1\n",
    "        originalBarcodes[i,2] = oneDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = oneDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "oneDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(birthSortedAdjacencyMatrix[:,:], x[1], x[2], labelsTimesPositionsType[:,2],1)), dimension = 1)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        oneDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "oneDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            oneDimemsionalTopologicalSimilarity[i,j] = 1 - (oneDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        end\n",
    "        if labelsTimesPositionsType[i,2] == labelsTimesPositionsType[j,2]\n",
    "            oneDimemsionalTopologicalSimilarity[i,j] = 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: oneDimensionalBND, oneDimemsionalTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "oneDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes second dimensional topological similarity\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(birthSortedAdjacencyMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 2)\n",
    "\n",
    "twoDimBarcodes = barcode(eireneGraph, dim = 2)\n",
    "originalBarcodes = zeros(size(twoDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(twoDimBarcodes)[1] > 0\n",
    "    for i in 1:size(twoDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 2\n",
    "        originalBarcodes[i,2] = twoDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = twoDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "twoDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(birthSortedAdjacencyMatrix[:,:], x[1], x[2], labelsTimesPositionsType[:,2],2)), dimension = 2)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        twoDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "twoDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            twoDimemsionalTopologicalSimilarity[i,j] = 1 - (twoDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        end\n",
    "        if labelsTimesPositionsType[i,2] == labelsTimesPositionsType[j,2]\n",
    "            twoDimemsionalTopologicalSimilarity[i,j] = 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: twoDimensionalBND, twoDimemsionalTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "twoDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute average similarity across dimensions 1-3\n",
    "averageTopologicalSimilarity = zeroDimemsionalTopologicalSimilarity[:,:]/3 + oneDimemsionalTopologicalSimilarity[:,:]/3 + twoDimemsionalTopologicalSimilarity[:,:]/3\n",
    "averageBND = zeroDimensionalBND[:,:]/3 + oneDimensionalBND[:,:]/3 + twoDimensionalBND[:,:]/3\n",
    "\n",
    "println(\"output: averageTopologicalSimilarity, averageBND\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#finds locations of identical birth times\n",
    "sameTime = zeros(279,279)\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        sameTime[i,j] = 1\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(zeroDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(zeroDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionZeroSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(oneDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension One Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(oneDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionOneSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(twoDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(twoDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionTwoSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(averageTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Averaged Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(averageBND, yflip = true, aspect_ratio=:equal, title = \"Averaged Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"averageSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save these as csv's to save computation time\n",
    "\n",
    "CSV.write(\"dimensionZeroSimilarity.csv\", DataFrame(zeroDimemsionalTopologicalSimilarity), writeheader = false)\n",
    "CSV.write(\"dimensionOneSimilarity.csv\", DataFrame(oneDimemsionalTopologicalSimilarity), writeheader = false)\n",
    "CSV.write(\"dimensionTwoSimilarity.csv\", DataFrame(twoDimemsionalTopologicalSimilarity), writeheader = false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pull from csv as another option\n",
    "zeroDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "oneDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "twoDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "\n",
    "zeroDimensionalBND = zeros(279,279)\n",
    "oneDimensionalBND = zeros(279,279)\n",
    "twoDimensionalBND = zeros(279,279)\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        zeroDimemsionalTopologicalSimilarity[i,j] = zeroDimTopSim[i,j]\n",
    "        oneDimemsionalTopologicalSimilarity[i,j] = oneDimTopSim[i,j]\n",
    "        twoDimemsionalTopologicalSimilarity[i,j] = twoDimTopSim[i,j]\n",
    "        \n",
    "        timeDiff = abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2])\n",
    "        if timeDiff != 0\n",
    "            zeroDimensionalBND[i,j] =  timeDiff * (1 - zeroDimTopSim[i,j])\n",
    "            oneDimensionalBND[i,j] =  timeDiff * (1 - oneDimTopSim[i,j])\n",
    "            twoDimensionalBND[i,j] =  timeDiff * (1 - twoDimTopSim[i,j])\n",
    "        else\n",
    "            zeroDimensionalBND[i,j] = 0\n",
    "            oneDimensionalBND[i,j] = 0\n",
    "            twoDimensionalBND[i,j] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#as all of these computations are identical, simply pick which dimension you want to analyze here\n",
    "topologicalSimilarity = zeroDimemsionalTopologicalSimilarity[:,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test on reorderability between types\n",
    "\n",
    "#variables for connections within a type\n",
    "intraSensorySimilarities = zeros(3570) #85 sensory neurons\n",
    "intraMotorSimilarities = zeros(6555) #115 motor neurons\n",
    "intraInterSimilarities = zeros(3081) # 79 inter neurons\n",
    "intraSensoryCounter = 1\n",
    "intraMotorCounter = 1\n",
    "intraInterCounter = 1\n",
    "\n",
    "\n",
    "#variables for connections between different types and a fixed type\n",
    "interSensorySimilarities = zeros(16490) \n",
    "interMotorSimilarities = zeros(18860) \n",
    "interInterSimilarities = zeros(15800)\n",
    "interSensoryCounter = 1\n",
    "interMotorCounter = 1\n",
    "interInterCounter = 1\n",
    "\n",
    "#variables for general within types connections\n",
    "intraTypeSimilariteis = zeros(13206)\n",
    "intraCounter = 1\n",
    "\n",
    "#variabels for general between types connections\n",
    "interTypeSimilariteis = zeros(25575)\n",
    "interCounter = 1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#loads all the various arrays\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        #see if the connection is intra-type\n",
    "        if labelsTimesPositionsType[i,5] == labelsTimesPositionsType[j,5]\n",
    "            \n",
    "            #determine which type the connection is, add the similarity to the list, and up the ticker\n",
    "            if labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "                intraSensorySimilarities[intraSensoryCounter] = topologicalSimilarity[i,j]\n",
    "                intraSensoryCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "                intraMotorSimilarities[intraMotorCounter] = topologicalSimilarity[i,j]\n",
    "                intraMotorCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"inter\"\n",
    "                intraInterSimilarities[intraInterCounter] = topologicalSimilarity[i,j]\n",
    "                intraInterCounter += 1\n",
    "            end\n",
    "            \n",
    "            intraTypeSimilariteis[intraCounter]  = topologicalSimilarity[i,j]\n",
    "            intraCounter += 1\n",
    "            \n",
    "        end\n",
    "        \n",
    "        #otherwise the connection is a general InterType\n",
    "        if labelsTimesPositionsType[i,5] != labelsTimesPositionsType[j,5]\n",
    "            \n",
    "            #determine which types the connection is between, add the similarity to the lists, and up the tickers\n",
    "            if labelsTimesPositionsType[i,5] == \"sensory\" || labelsTimesPositionsType[j,5] == \"sensory\"\n",
    "                interSensorySimilarities[interSensoryCounter] = topologicalSimilarity[i,j]\n",
    "                interSensoryCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"motor\" || labelsTimesPositionsType[j,5] == \"motor\"\n",
    "                interMotorSimilarities[interMotorCounter] = topologicalSimilarity[i,j]\n",
    "                interMotorCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"inter\" || labelsTimesPositionsType[j,5] == \"inter\"\n",
    "                interInterSimilarities[interInterCounter] = topologicalSimilarity[i,j]\n",
    "                interInterCounter += 1\n",
    "            end\n",
    "                      \n",
    "            interTypeSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "            interCounter += 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"many outputs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter sensory pair\n",
    "MannWhitneyUTest(intraSensorySimilarities, interSensorySimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter motor pair\n",
    "MannWhitneyUTest(intraMotorSimilarities, interMotorSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter inter pair\n",
    "MannWhitneyUTest(intraInterSimilarities, interInterSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(intraTypeSimilariteis, interTypeSimilariteis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we would also like to determine summed similarity for each type\n",
    "\n",
    "summedSensorySimilarities = zeros(85) #85 sensory neurons\n",
    "summedMotorSimilarities = zeros(115) #115 motor neurons\n",
    "summedInterSimilarities = zeros(79) # 79 inter neurons\n",
    "summedSensoryCounter = 1\n",
    "summedMotorCounter = 1\n",
    "summedInterCounter = 1\n",
    "\n",
    "for i in 1:279\n",
    "    \n",
    "    if labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "        summedSensorySimilarities[summedSensoryCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedSensoryCounter += 1\n",
    "    end\n",
    "    \n",
    "   if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "        summedMotorSimilarities[summedMotorCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedMotorCounter += 1\n",
    "    end\n",
    "    \n",
    "    if labelsTimesPositionsType[i,5] == \"inter\"\n",
    "        summedInterSimilarities[summedInterCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedInterCounter += 1\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedSensorySimilarities, summedMotorSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedSensorySimilarities, summedInterSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedMotorSimilarities, summedInterSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 8)\n",
    "plotArray[1] = interMotorSimilarities[:]\n",
    "plotArray[2] = intraMotorSimilarities[:]\n",
    "plotArray[3] = interSensorySimilarities[:]\n",
    "plotArray[4] = intraSensorySimilarities[:]\n",
    "plotArray[5] = interInterSimilarities[:]\n",
    "plotArray[6] = intraInterSimilarities[:]\n",
    "plotArray[7] = interTypeSimilariteis[:]\n",
    "plotArray[8] = intraTypeSimilariteis[:]\n",
    "\n",
    "#violin([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"inter-Type\"],plotArray,leg=false, color = theme(:default))\n",
    "boxplot([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"intra-Type\"],title = \"Dimension Zero Similarity\", plotArray[1:6],leg=false)\n",
    "#savefig(\"dimensionZeroSimilarityPlots.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 3)\n",
    "plotArray[1] = summedSensorySimilarities[:]\n",
    "plotArray[2] = summedMotorSimilarities[:]\n",
    "plotArray[3] = summedInterSimilarities[:]\n",
    "\n",
    "\n",
    "boxplot([\"sensory\" \"motor\" \"inter\"],title = \"Dimension Zero Summed Similarity\", plotArray,leg=false)\n",
    "#savefig(\"dimensionZeroSummedSimilarityPlots.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#as all of these computations are identical, simply pick which dimension you want to analyze here\n",
    "topologicalSimilarity = zeroDimemsionalTopologicalSimilarity[:,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test on reorderability between region\n",
    "\n",
    "#variables for connections within a region\n",
    "intraHeadSimilarities = zeros(12720) #160 head neurons\n",
    "intraBodySimilarities = zeros(4186) #92 body neurons\n",
    "intraTailSimilarities = zeros(351) # 27 tail neurons\n",
    "intraHeadCounter = 1\n",
    "intraBodyCounter = 1\n",
    "intraTailCounter = 1\n",
    "\n",
    "\n",
    "#variables for connections between different types and a fixed type\n",
    "interHeadSimilarities = zeros(19040) \n",
    "interBodySimilarities = zeros(17204) \n",
    "interTailSimilarities = zeros(6804)\n",
    "interHeadCounter = 1\n",
    "interBodyCounter = 1\n",
    "interTailCounter = 1\n",
    "\n",
    "#variables for general within types connections\n",
    "intraRegionSimilariteis = zeros(17257)\n",
    "intraCounter = 1\n",
    "\n",
    "#variabels for general between types connections\n",
    "interRegionSimilariteis = zeros(21524)\n",
    "interCounter = 1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#loads all the various arrays\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        #see which region i belongs to\n",
    "        if labelsTimesPositionsType[i,3] < 0\n",
    "            #i is in the head\n",
    "            \n",
    "            if labelsTimesPositionsType[j,3] < 0\n",
    "                #j is in the head\n",
    "                intraHeadSimilarities[intraHeadCounter] = topologicalSimilarity[i,j]\n",
    "                intraRegionSimilariteis[intraCounter] = topologicalSimilarity[i,j]\n",
    "                intraHeadCounter += 1\n",
    "                intraCounter += 1\n",
    "                \n",
    "            elseif 0.935 > labelsTimesPositionsType[j,3] > 0\n",
    "               #j is in the body\n",
    "                interHeadSimilarities[interHeadCounter] = topologicalSimilarity[i,j]\n",
    "                interBodySimilarities[interBodyCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interHeadCounter += 1\n",
    "                interBodyCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            else\n",
    "                #j is in the tail\n",
    "                interHeadSimilarities[interHeadCounter] = topologicalSimilarity[i,j]\n",
    "                interTailSimilarities[interTailCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interHeadCounter += 1\n",
    "                interTailCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            end\n",
    "            \n",
    "               \n",
    "        elseif 0.935 > labelsTimesPositionsType[i,3] > 0\n",
    "           #i is in the body\n",
    "            \n",
    "            if labelsTimesPositionsType[j,3] < 0\n",
    "                #j is in the head\n",
    "                interHeadSimilarities[interHeadCounter] = topologicalSimilarity[i,j]\n",
    "                interBodySimilarities[interBodyCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interHeadCounter += 1\n",
    "                interBodyCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            elseif 0.935 > labelsTimesPositionsType[j,3] > 0\n",
    "               #j is in the body\n",
    "                intraBodySimilarities[intraBodyCounter] = topologicalSimilarity[i,j]\n",
    "                intraRegionSimilariteis[intraCounter] = topologicalSimilarity[i,j]\n",
    "                intraBodyCounter += 1\n",
    "                intraCounter += 1\n",
    "                \n",
    "            else\n",
    "                #j is in the tail\n",
    "                interBodySimilarities[interBodyCounter] = topologicalSimilarity[i,j]\n",
    "                interTailSimilarities[interTailCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interBodyCounter += 1\n",
    "                interTailCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            end\n",
    "            \n",
    "        else\n",
    "            #i is in the tail\n",
    "            \n",
    "            if labelsTimesPositionsType[j,3] < 0\n",
    "                #j is in the head\n",
    "                interHeadSimilarities[interHeadCounter] = topologicalSimilarity[i,j]\n",
    "                interTailSimilarities[interTailCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interHeadCounter += 1\n",
    "                interTailCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            elseif 0.935 > labelsTimesPositionsType[j,3] > 0\n",
    "               #j is in the body\n",
    "                interBodySimilarities[interBodyCounter] = topologicalSimilarity[i,j]\n",
    "                interTailSimilarities[interTailCounter] = topologicalSimilarity[i,j]\n",
    "                interRegionSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "                interBodyCounter += 1\n",
    "                interTailCounter += 1\n",
    "                interCounter += 1\n",
    "                \n",
    "            else\n",
    "                #j is in the tail\n",
    "                intraTailSimilarities[intraTailCounter] = topologicalSimilarity[i,j]\n",
    "                intraRegionSimilariteis[intraCounter] = topologicalSimilarity[i,j]\n",
    "                intraTailCounter += 1\n",
    "                intraCounter += 1\n",
    "                \n",
    "            end\n",
    "            \n",
    "        end\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"many outputs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter sensory pair\n",
    "MannWhitneyUTest(intraHeadSimilarities, interHeadSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter motor pair\n",
    "MannWhitneyUTest(intraBodySimilarities, interBodySimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter inter pair\n",
    "MannWhitneyUTest(intraTailSimilarities, interTailSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(intraRegionSimilariteis, interRegionSimilariteis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we would also like to determine summed similarity for each type\n",
    "\n",
    "summedHeadSimilarities = zeros(160) #160 head neurons\n",
    "summedBodySimilarities = zeros(92) #92 body neurons\n",
    "summedTailSimilarities = zeros(27) #27 tail neurons\n",
    "summedHeadCounter = 1\n",
    "summedBodyCounter = 1\n",
    "summedTailCounter = 1\n",
    "\n",
    "for i in 1:279\n",
    "    \n",
    "    if labelsTimesPositionsType[i,3] < 0\n",
    "        summedHeadSimilarities[summedHeadCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedHeadCounter += 1\n",
    "    elseif labelsTimesPositionsType[i,3] < 0.935\n",
    "        summedBodySimilarities[summedBodyCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedBodyCounter += 1\n",
    "    else\n",
    "        summedTailSimilarities[summedTailCounter] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[i,:])\n",
    "        summedTailCounter += 1\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedHeadSimilarities, summedBodySimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedHeadSimilarities, summedTailSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(summedBodySimilarities, summedTailSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 8)\n",
    "plotArray[1] = interHeadSimilarities[:]\n",
    "plotArray[2] = intraHeadSimilarities[:]\n",
    "plotArray[3] = interBodySimilarities[:]\n",
    "plotArray[4] = intraBodySimilarities[:]\n",
    "plotArray[5] = interTailSimilarities[:]\n",
    "plotArray[6] = intraTailSimilarities[:]\n",
    "plotArray[7] = interRegionSimilariteis[:]\n",
    "plotArray[8] = intraRegionSimilariteis[:]\n",
    "\n",
    "#violin([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"inter-Type\"],plotArray,leg=false, color = theme(:default))\n",
    "boxplot([\"inter-Head\" \"intra-Head\" \"inter-Body\" \"intra-Body\" \"inter-Tail\" \"intra-Tail\" \"inter-Region\" \"intra-Region\"],title = \"Dimension Zero Similarity\", plotArray[1:6],leg=false)\n",
    "#savefig(\"dimensionZeroSimilarityPlotsRegion.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 3)\n",
    "plotArray[1] = summedHeadSimilarities[:]\n",
    "plotArray[2] = summedBodySimilarities[:]\n",
    "plotArray[3] = summedTailSimilarities[:]\n",
    "\n",
    "\n",
    "#violin([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"inter-Type\"],plotArray,leg=false, color = theme(:default))\n",
    "boxplot([\"Head\" \"Body\" \"Tail\"],title = \"Dimension Zero Similarity\", plotArray,leg=false)\n",
    "#savefig(\"dimensionZeroSummedSimilarityPlotsRegion.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixedPairValues = zeros(15678,3,2)\n",
    "unmixedPairValues = zeros(23103,3,2)\n",
    "mixedCounter = 0\n",
    "unmixedCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            unmixedCounter += 1\n",
    "            unmixedPairValues[unmixedCounter,1,1] = zeroDimemsionalTopologicalSimilarity[i,j]\n",
    "            unmixedPairValues[unmixedCounter,2,1] = oneDimemsionalTopologicalSimilarity[i,j]\n",
    "            unmixedPairValues[unmixedCounter,3,1] = twoDimemsionalTopologicalSimilarity[i,j]\n",
    "            \n",
    "            unmixedPairValues[unmixedCounter,1,2] = zeroDimensionalBND[i,j]\n",
    "            unmixedPairValues[unmixedCounter,2,2] = oneDimensionalBND[i,j]\n",
    "            unmixedPairValues[unmixedCounter,3,2] = twoDimensionalBND[i,j]\n",
    "        else\n",
    "            mixedCounter += 1\n",
    "            mixedPairValues[mixedCounter,1,1] = zeroDimemsionalTopologicalSimilarity[i,j]\n",
    "            mixedPairValues[mixedCounter,2,1] = oneDimemsionalTopologicalSimilarity[i,j]\n",
    "            mixedPairValues[mixedCounter,3,1] = twoDimemsionalTopologicalSimilarity[i,j]\n",
    "            \n",
    "            mixedPairValues[mixedCounter,1,2] = zeroDimensionalBND[i,j]\n",
    "            mixedPairValues[mixedCounter,2,2] = oneDimensionalBND[i,j]\n",
    "            mixedPairValues[mixedCounter,3,2] = twoDimensionalBND[i,j]\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(unmixedPairValues[:,3,1], mixedPairValues[:,3,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(unmixedPairValues[:,1,2], mixedPairValues[:,1,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = unmixedPairValues[:,1,1]\n",
    "plotArray[2] = mixedPairValues[:,1,1]\n",
    "plotArray[3] = unmixedPairValues[:,2,1]\n",
    "plotArray[4] = mixedPairValues[:,2,1]\n",
    "plotArray[5] = unmixedPairValues[:,3,1]\n",
    "plotArray[6] = mixedPairValues[:,3,1]\n",
    "\n",
    "boxplot([\"zero un-mixed\" \"zero mixed\" \"one un-mixed\" \"one mixed\" \"two un-mixed\" \"two mixed\"], title = \"Topological Similarity\", plotArray[1:6],leg=false)\n",
    "savefig(\"mixedUnmixedTopSim.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = unmixedPairValues[:,1,2]\n",
    "plotArray[2] = mixedPairValues[:,1,2]\n",
    "plotArray[3] = unmixedPairValues[:,2,2]\n",
    "plotArray[4] = mixedPairValues[:,2,2]\n",
    "plotArray[5] = unmixedPairValues[:,3,2]\n",
    "plotArray[6] = mixedPairValues[:,3,2]\n",
    "\n",
    "boxplot([\"zero un-mixed\" \"zero mixed\" \"one un-mixed\" \"one mixed\" \"two un-mixed\" \"two mixed\"], title = \"Bottleneck Distance\", plotArray[1:6],leg=false)\n",
    "savefig(\"mixedUnmixedBND.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#top sim with time gap deleted\n",
    "newTimesList = labelsTimesPositionsType[:,2]\n",
    "for i in 202:279\n",
    "    newTimesList[i] -= 765.55\n",
    "end\n",
    "\n",
    "\n",
    "gapDeletedMatrix = birthSortedAdjacencyMatrix[:,:]\n",
    "\n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        \n",
    "        if gapDeletedMatrix[i,j] < 3000 && ( i > 201  || j > 201)\n",
    "            \n",
    "            max = maximum([i,j])\n",
    "            gapDeletedMatrix[i,j] = newTimesList[max]\n",
    "        end\n",
    "    end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes zeroth dimensional topological similarity for all mixed pairs, as the value *cannot* change for all other pairs\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(gapDeletedMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 0)\n",
    "\n",
    "zeroDimBarcodes = barcode(eireneGraph, dim = 0)\n",
    "originalBarcodes = zeros(size(zeroDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(zeroDimBarcodes)[1] > 0\n",
    "    for i in 1:size(zeroDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 0\n",
    "        originalBarcodes[i,2] = zeroDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = zeroDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 15678)\n",
    "counter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "zeroDimensionalMixedPairBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(gapDeletedMatrix[:,:], x[1], x[2], newTimesList, 0)), dimension = 0)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        storageCounter += 1\n",
    "        zeroDimensionalMixedPairBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "zeroDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            zeroDimemsionalMixedPairTopologicalSimilarity[i,j] = zeroDimemsionalTopologicalSimilarity[i,j]\n",
    "            zeroDimensionalMixedPairBND[i,j] = zeroDimensionalBND[i,j]\n",
    "        else\n",
    "        \n",
    "            if newTimesList[i] != newTimesList[j]\n",
    "                zeroDimemsionalMixedPairTopologicalSimilarity[i,j] = 1 - (zeroDimensionalMixedPairBND[i,j] / (newTimesList[i] - newTimesList[j]))\n",
    "            end\n",
    "            if newTimesList[i] == newTimesList[j]\n",
    "                zeroDimemsionalMixedPairTopologicalSimilarity[i,j] = 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: zeroDimensionalMixedPairBND, zeroDimemsionalMixedPairTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "zeroDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes first dimensional topological similarity for all mixed pairs, as the value *cannot* change for all other pairs\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(gapDeletedMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 1)\n",
    "\n",
    "oneDimBarcodes = barcode(eireneGraph, dim = 1)\n",
    "originalBarcodes = zeros(size(oneDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(oneDimBarcodes)[1] > 0\n",
    "    for i in 1:size(oneDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 1\n",
    "        originalBarcodes[i,2] = oneDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = oneDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 15678)\n",
    "counter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "oneDimensionalMixedPairBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(gapDeletedMatrix[:,:], x[1], x[2], newTimesList[:], 1)), dimension = 1)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        storageCounter += 1\n",
    "        oneDimensionalMixedPairBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "oneDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            oneDimemsionalMixedPairTopologicalSimilarity[i,j] = oneDimemsionalTopologicalSimilarity[i,j]\n",
    "            oneDimensionalMixedPairBND[i,j] = oneDimensionalBND[i,j]\n",
    "        else\n",
    "        \n",
    "            if newTimesList[i] != newTimesList[j]\n",
    "                oneDimemsionalMixedPairTopologicalSimilarity[i,j] = 1 - (oneDimensionalMixedPairBND[i,j] / (newTimesList[i] - newTimesList[j]))\n",
    "            end\n",
    "            if newTimesList[i] == newTimesList[j]\n",
    "                oneDimemsionalMixedPairTopologicalSimilarity[i,j] = 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: oneDimensionalMixedPairBND, oneDimemsionalMixedPairTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "oneDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes second dimensional topological similarity for all mixed pairs, as the value *cannot* change for all other pairs\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(gapDeletedMatrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 2)\n",
    "\n",
    "twoDimBarcodes = barcode(eireneGraph, dim = 2)\n",
    "originalBarcodes = zeros(size(twoDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(twoDimBarcodes)[1] > 0\n",
    "    for i in 1:size(twoDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 2\n",
    "        originalBarcodes[i,2] = twoDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = twoDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 15678)\n",
    "counter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "twoDimensionalMixedPairBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderedBarcodes(gapDeletedMatrix[:,:], x[1], x[2], newTimesList[:], 2)), dimension = 2)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 202:279\n",
    "    for j in 1:201\n",
    "        storageCounter += 1\n",
    "        twoDimensionalMixedPairBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "twoDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            twoDimemsionalMixedPairTopologicalSimilarity[i,j] = twoDimemsionalTopologicalSimilarity[i,j]\n",
    "            twoDimensionalMixedPairBND[i,j] = twoDimensionalBND[i,j]\n",
    "        else\n",
    "        \n",
    "            if newTimesList[i] != newTimesList[j]\n",
    "                twoDimemsionalMixedPairTopologicalSimilarity[i,j] = 1 - (twoDimensionalMixedPairBND[i,j] / (newTimesList[i] - newTimesList[j]))\n",
    "            end\n",
    "            if newTimesList[i] == newTimesList[j]\n",
    "                twoDimemsionalMixedPairTopologicalSimilarity[i,j] = 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "println(\"output: twoDimensionalMixedPairBND, twoDimemsionalMixedPairTopologicalSimilarity\")\n",
    "\n",
    "storageCounter = nothing\n",
    "indicies = nothing\n",
    "storage = nothing\n",
    "counter = nothing\n",
    "eireneGraph = nothing\n",
    "twoDimBarcodes = nothing\n",
    "originalBarcodes = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute average similarity across dimensions 1-3\n",
    "averageMixedPairTopologicalSimilarity = zeroDimemsionalMixedPairTopologicalSimilarity[:,:]/3 + oneDimemsionalMixedPairTopologicalSimilarity[:,:]/3 + twoDimemsionalMixedPairTopologicalSimilarity[:,:]/3\n",
    "averageMixedPairBND = zeroDimensionalMixedPairBND[:,:]/3 + oneDimensionalMixedPairBND[:,:]/3 + twoDimensionalMixedPairBND[:,:]/3\n",
    "\n",
    "println(\"output: averageMixedPairTopologicalSimilarity, averageMixedPairBND\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(zeroDimemsionalMixedPairTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Gap Deleted Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(zeroDimensionalMixedPairBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Gap Deleted Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "savefig(\"dimensionZeroGapDeletedSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(oneDimemsionalMixedPairTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension One Gap Deleted Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(oneDimensionalMixedPairBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Gap Deleted Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "savefig(\"dimensionOneGapDeletedSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(twoDimemsionalMixedPairTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Gap Deleted Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(twoDimensionalMixedPairBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Gap Deleted Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "savefig(\"dimensionTwoGapDeletedSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(averageMixedPairTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Averaged Gap Deleted Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(averageMixedPairBND, yflip = true, aspect_ratio=:equal, title = \"Averaged Gap Deleted Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "savefig(\"averagedGapDeletedSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save these as csv's to save computation time\n",
    "\n",
    "CSV.write(\"dimensionZeroMixedPairSimilarity.csv\", DataFrame(zeroDimemsionalMixedPairTopologicalSimilarity), writeheader = false)\n",
    "CSV.write(\"dimensionOneMixedPairSimilarity.csv\", DataFrame(oneDimemsionalMixedPairTopologicalSimilarity), writeheader = false)\n",
    "CSV.write(\"dimensionTwoMixedPairSimilarity.csv\", DataFrame(twoDimemsionalMixedPairTopologicalSimilarity), writeheader = false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pull from csv as another option\n",
    "zeroDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "oneDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "twoDimemsionalMixedPairTopologicalSimilarity = zeros(279,279)\n",
    "\n",
    "zeroDimensionalMixedPairBND = zeros(279,279)\n",
    "oneDimensionalMixedPairBND = zeros(279,279)\n",
    "twoDimensionalMixedPairBND = zeros(279,279)\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        zeroDimemsionalMixedPairTopologicalSimilarity[i,j] = zeroDimTopSimGD[i,j]\n",
    "        oneDimemsionalMixedPairTopologicalSimilarity[i,j] = oneDimTopSimGD[i,j]\n",
    "        twoDimemsionalMixedPairTopologicalSimilarity[i,j] = twoDimTopSimGD[i,j]\n",
    "        \n",
    "        timeDiff = abs(newTimesList[i] - newTimesList[j])\n",
    "        if timeDiff != 0\n",
    "            zeroDimensionalMixedPairBND[i,j] =  timeDiff * (1 - zeroDimTopSimGD[i,j])\n",
    "            oneDimensionalMixedPairBND[i,j] =  timeDiff * (1 - oneDimTopSimGD[i,j])\n",
    "            twoDimensionalMixedPairBND[i,j] =  timeDiff * (1 - twoDimTopSimGD[i,j])\n",
    "        else\n",
    "            zeroDimensionalMixedPairBND[i,j] = 0\n",
    "            oneDimensionalMixedPairBND[i,j] = 0\n",
    "            twoDimensionalMixedPairBND[i,j] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixedPairValues = zeros(15678,3,2)\n",
    "unmixedPairValues = zeros(23103,3,2)\n",
    "mixedCounter = 0\n",
    "unmixedCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if i < 202 || j > 201\n",
    "            unmixedCounter += 1\n",
    "            unmixedPairValues[unmixedCounter,1,1] = zeroDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            unmixedPairValues[unmixedCounter,2,1] = oneDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            unmixedPairValues[unmixedCounter,3,1] = twoDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            \n",
    "            unmixedPairValues[unmixedCounter,1,2] = zeroDimensionalMixedPairBND[i,j]\n",
    "            unmixedPairValues[unmixedCounter,2,2] = oneDimensionalMixedPairBND[i,j]\n",
    "            unmixedPairValues[unmixedCounter,3,2] = twoDimensionalMixedPairBND[i,j]\n",
    "        else\n",
    "            mixedCounter += 1\n",
    "            mixedPairValues[mixedCounter,1,1] = zeroDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            mixedPairValues[mixedCounter,2,1] = oneDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            mixedPairValues[mixedCounter,3,1] = twoDimemsionalMixedPairTopologicalSimilarity[i,j]\n",
    "            \n",
    "            mixedPairValues[mixedCounter,1,2] = zeroDimensionalMixedPairBND[i,j]\n",
    "            mixedPairValues[mixedCounter,2,2] = oneDimensionalMixedPairBND[i,j]\n",
    "            mixedPairValues[mixedCounter,3,2] = twoDimensionalMixedPairBND[i,j]\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(unmixedPairValues[:,1,1], mixedPairValues[:,1,1])\n",
    "\n",
    "#0 - -33        0\n",
    "#1 - -99     - .0457\n",
    "#2 - -99     .347"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(unmixedPairValues[:,1,2], mixedPairValues[:,1,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = unmixedPairValues[:,1,1]\n",
    "plotArray[2] = mixedPairValues[:,1,1]\n",
    "plotArray[3] = unmixedPairValues[:,2,1]\n",
    "plotArray[4] = mixedPairValues[:,2,1]\n",
    "plotArray[5] = unmixedPairValues[:,3,1]\n",
    "plotArray[6] = mixedPairValues[:,3,1]\n",
    "\n",
    "boxplot([\"zero un-mixed\" \"zero mixed\" \"one un-mixed\" \"one mixed\" \"two un-mixed\" \"two mixed\"], title = \"Topological Similarity\", plotArray[1:6],leg=false)\n",
    "savefig(\"mixedUnmixedTopSim.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = unmixedPairValues[:,1,2]\n",
    "plotArray[2] = mixedPairValues[:,1,2]\n",
    "plotArray[3] = unmixedPairValues[:,2,2]\n",
    "plotArray[4] = mixedPairValues[:,2,2]\n",
    "plotArray[5] = unmixedPairValues[:,3,2]\n",
    "plotArray[6] = mixedPairValues[:,3,2]\n",
    "\n",
    "boxplot([\"zero un-mixed\" \"zero mixed\" \"one un-mixed\" \"one mixed\" \"two un-mixed\" \"two mixed\"], title = \"Bottleneck Distance\", plotArray[1:6],leg=false)\n",
    "savefig(\"mixedUnmixedBND.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine top sim vs distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find set of actually existing time differences, to cut down computation costs on the next thing\n",
    "allTimeDifferences = zeros(38781)\n",
    "counter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        allTimeDifferences[counter] =  round(abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]); digits=2) \n",
    "    end\n",
    "end\n",
    "\n",
    "sort!(allTimeDifferences)\n",
    "\n",
    "actualTimeDifferences = zeros(10910)\n",
    "counter = 1\n",
    "actualTimeDifferences[counter] = allTimeDifferences[1]\n",
    "\n",
    "for i in 2:38781\n",
    "    if allTimeDifferences[i-1] != allTimeDifferences[i]\n",
    "        counter += 1\n",
    "        actualTimeDifferences[counter] = allTimeDifferences[i]\n",
    "    end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do similarity over increasing window calculations\n",
    "growingWindows = zeros(10910,3,2)  #window size x dimension x mean/std\n",
    "\n",
    "runningSumZero = zeros(38781)\n",
    "runningSumOne = zeros(38781)\n",
    "runningSumTwo = zeros(38781)\n",
    "counter = 0\n",
    "\n",
    "for k in actualTimeDifferences\n",
    "    counter += 1\n",
    "    runningCount = 0\n",
    "    \n",
    "    \n",
    "    #find averages given a time window of k\n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            \n",
    "            if abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]) <= k\n",
    "                runningCount += 1\n",
    "                runningSumZero[runningCount] = zeroDimemsionalTopologicalSimilarity[i,j] \n",
    "                runningSumOne[runningCount] = oneDimemsionalTopologicalSimilarity[i,j]   \n",
    "                runningSumTwo[runningCount] = twoDimemsionalTopologicalSimilarity[i,j]  \n",
    "            end\n",
    "            \n",
    "        end\n",
    "    end\n",
    "    \n",
    "    growingWindows[counter,1,1] = mean(runningSumZero[1:runningCount])\n",
    "    growingWindows[counter,2,1] = mean(runningSumZero[1:runningCount])\n",
    "    growingWindows[counter,3,1] = mean(runningSumZero[1:runningCount])\n",
    "    \n",
    "    growingWindows[counter,1,2] = std(runningSumZero[1:runningCount])\n",
    "    growingWindows[counter,2,2] = std(runningSumZero[1:runningCount])\n",
    "    growingWindows[counter,3,2] = std(runningSumZero[1:runningCount])\n",
    "end\n",
    "\n",
    "print(\"output: growingWindows, runningSumZero, runningSumOne, runningSumTwo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for each pair of neurons, find their distance and their top sim\n",
    "dimZeroDistances = zeros(38781,2)\n",
    "dimOneDistances = zeros(38781,2)\n",
    "dimTwoDistances = zeros(38781,2)\n",
    "birthTimeColors = zeros(38781,2)\n",
    "randomColors = zeros(38781,2)\n",
    "birthNumber = zeros(38781,2)\n",
    "byNeuronType = zeros(38781,2)\n",
    "byRegion = zeros(38781,2)\n",
    "byxcoor = zeros(38781,2)\n",
    "movieColors = zeros(38781,2)\n",
    "\n",
    "counter = 0\n",
    "colors = rand(279)\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        counter += 1\n",
    "        time = round(abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]); digits=2) \n",
    "        \n",
    "        dimZeroDistances[counter,1] = time \n",
    "        dimOneDistances[counter,1] = time \n",
    "        dimTwoDistances[counter,1] = time\n",
    "        \n",
    "        dimZeroDistances[counter,2] = zeroDimemsionalTopologicalSimilarity[i,j] \n",
    "        dimOneDistances[counter,2] = oneDimemsionalTopologicalSimilarity[i,j]   \n",
    "        dimTwoDistances[counter,2] = twoDimemsionalTopologicalSimilarity[i,j]\n",
    "        \n",
    "        randomColors[counter,:] = [colors[i],colors[j]]\n",
    "        birthTimeColors[counter,:] = [labelsTimesPositionsType[i,2] , labelsTimesPositionsType[j,2]]\n",
    "        birthNumber[counter,:] = [i,j]\n",
    "        \n",
    "        #do region sorting\n",
    "        if labelsTimesPositionsType[i,3] < 0\n",
    "            byRegion[counter,1] = 0\n",
    "        elseif labelsTimesPositionsType[i,3] < 0.935\n",
    "            byRegion[counter,1] = 0.5\n",
    "        else\n",
    "            byRegion[counter,1] = 1\n",
    "        end\n",
    "        \n",
    "        if labelsTimesPositionsType[j,3] < 0\n",
    "            byRegion[counter,2] = 0\n",
    "        elseif labelsTimesPositionsType[j,3] < 0.935\n",
    "            byRegion[counter,2] = 0.5\n",
    "        else\n",
    "            byRegion[counter,2] = 1\n",
    "        end\n",
    "        \n",
    "        \n",
    "        #type sorting\n",
    "        if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "            byNeuronType[counter,1] = 0\n",
    "        elseif labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "            byNeuronType[counter,1] = 0.5\n",
    "        else\n",
    "            byNeuronType[counter,1] = 1\n",
    "        end\n",
    "        \n",
    "        if labelsTimesPositionsType[j,5] == \"motor\"\n",
    "            byNeuronType[counter,2] = 0\n",
    "        elseif labelsTimesPositionsType[j,5] == \"sensory\"\n",
    "            byNeuronType[counter,2] = 0.5\n",
    "        else\n",
    "            byNeuronType[counter,2] = 1\n",
    "        end\n",
    "        \n",
    "        byxcoor[counter,:] = [labelsTimesPositionsType[i,3] , labelsTimesPositionsType[j,3]]\n",
    "        \n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1 = scatter(dimZeroDistances[:,1], dimZeroDistances[:,2], color = [:green theme(:default)], legend=false , title = \"Topological Similarity vs Temporal Distance\", label = [\"dim 0\"],  xlabel = \"Temporal Distance\", ylabel = \"Topological Similarity\")\n",
    "p2 = scatter(dimOneDistances[:,1], dimOneDistances[:,2], color = [:red theme(:default)], legend=false , title = \"Topological Similarity vs Temporal Distance\", label = [\"dim 1\"],  xlabel = \"Temporal Distance\", ylabel = \"Topological Similarity\")\n",
    "p3 = scatter(dimTwoDistances[:,1], dimTwoDistances[:,2], color = [:blue theme(:default)], legend=false , title = \"Topological Similarity vs Temporal Distance\", label = [\"dim 2\"],  xlabel = \"Temporal Distance\", ylabel = \"Topological Similarity\")\n",
    "\n",
    "plot(p1,p2,p3,layout = (3,1), size = (1000,1000))\n",
    "#savefig(\"topSimByDistance.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the gworing window curves\n",
    "\n",
    "\n",
    "p1 = plot(actualTimeDifferences[:], growingWindows[:,1,1], ribbon = growingWindows[:,1,2], color = [:green theme(:default)], legend=:bottomright , title = \"Topological Similarity over Growing Time Window\", label = [\"dim 0\"],  xlabel = \"window size\", ylabel = \"average topological similarity\")\n",
    "\n",
    "p2 = plot(actualTimeDifferences[:], growingWindows[:,2,1], ribbon = growingWindows[:,2,2], color = [:red theme(:default)], legend=:bottomright , title = \"Topological Similarity over Growing Time Window\", label = [\"dim 1\"],  xlabel = \"window size\", ylabel = \"average topological similarity\")\n",
    "\n",
    "p3 = plot(actualTimeDifferences[:], growingWindows[:,3,1], ribbon = growingWindows[:,3,2], color = [:blue theme(:default)], legend=:bottomright , title = \"Topological Similarity over Growing Time Window\", label = [\"dim 2\"],  xlabel = \"window size\", ylabel = \"average topological similarity\")\n",
    "\n",
    "\n",
    "plot(p1,p2,p3, layout = (3,1), size = (800,800))\n",
    "\n",
    "#savefig(\"growingTimeWindow.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p2 = plot(actualTimeDifferences[:], growingWindows[:,2,1], ribbon = growingWindows[:,2,2], col = rand(38781), legend=:bottomright , title = \"Topological Similarity over Growing Time Window\", label = [\"dim 1\"],  xlabel = \"window size\", ylabel = \"average topological similarity\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(1:38781, birthBoundsForColor[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "birthTimeColors\n",
    "randomColors\n",
    "birthNumber\n",
    "byNeuronType\n",
    "byRegion\n",
    "byxcoor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyplot(size=(800,800), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(dimOneDistances[:,1], dimOneDistances[:,2], markerstrokewidth=0,\n",
    "        m = :blues,\n",
    "        zcolor = randomColors[:,1] # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyplot(size=(800,800), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(dimTwoDistances[:,1], dimTwoDistances[:,2], markerstrokewidth=0,\n",
    "        m = :blues,\n",
    "        zcolor = randomColors[:,1] # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )\n",
    "scatter!(dimOne[4,:,1], dimOne[4,:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets visualize each node's chart one at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for each pair of neurons, find their distance and their top sim\n",
    "dimOne = zeros(279,278,2)\n",
    "dimTwo = zeros(279,278,2)\n",
    " = zeros(279,)\n",
    "\n",
    "\n",
    "\n",
    "for current in 1:279\n",
    "    counter = 0\n",
    "    colors[current] = current\n",
    "for i in 1:279\n",
    "    \n",
    "    if i != current\n",
    "        counter += 1\n",
    "        time = round(abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[current,2]); digits=2) \n",
    "        \n",
    "        dimOne[current,counter,1] = time \n",
    "        dimTwo[current,counter,1] = time\n",
    "        \n",
    "        dimOne[current,counter,2] = oneDimemsionalTopologicalSimilarity[i,current]   \n",
    "        dimTwo[current,counter,2] = twoDimemsionalTopologicalSimilarity[i,current]\n",
    "        \n",
    "    end\n",
    "end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scatter(dimOne[i,:,1], dimOne[i,:,2], markerstrokewidth=0,\n",
    "        m = :blues,\n",
    "        zcolor = current[i] # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "movieColors = zeros(279,38781)\n",
    "\n",
    "for k in 1:279\n",
    "    counter = 0\n",
    "    \n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            counter += 1\n",
    "            if i == k || j == k\n",
    "                movieColors[k,counter] = 1\n",
    "            else\n",
    "                movieColors[k,counter] = 0\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anim = @animate for i in 1:279\n",
    "    \n",
    "    pyplot(size=(800,800), legend=false)  \n",
    "\n",
    "    scatter(dimTwoDistances[:,1], dimTwoDistances[:,2], markerstrokewidth=0, m = :blues, zcolor = movieColors[i,:], title = string(\"Dimension Two: Neuron #\", i,))\n",
    "    scatter!(dimTwo[i,:,1], dimTwo[i,:,2], m = :blues, zcolor = ones(278))\n",
    "\n",
    "end\n",
    "\n",
    "gif(anim, \"dimTwoIndividualsHighlighted.gif\", fps = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 1\n",
    "p1 = scatter([0],[0], markerstrokewidth=0, m = :blues, zcolor = movieColors[i,:], title = string(\"Dimension One: Neuron #\", i,))\n",
    "scatter!(dimOne[2,:,1], dimOne[i,:,2], m = :blues, zcolor = ones(278))\n",
    "scatter!(dimOne[2,:,1], dimOne[2,:,2], m = :blues, zcolor = ones(278))\n",
    "for k in 1:200\n",
    "    scatter!(dimOne[k,:,1], dimOne[k,:,2], m = :blues, zcolor = ones(278))\n",
    "end\n",
    "plot(p1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anim = @animate for i in 1:279\n",
    "    \n",
    "    pyplot(size=(800,800), legend=false)  \n",
    "\n",
    "    p1 = scatter([0], [0], markerstrokewidth=0, m = :blues, zcolor = movieColors[i,:], title = string(\"Dimension One: Neuron #\", i,))\n",
    "    \n",
    "    for k in 1:i\n",
    "        color = zeros(278)\n",
    "        fill!(color, rand())\n",
    "        scatter!(dimOne[i,:,1], dimOne[i,:,2], m = :blues, zcolor = ones(278))\n",
    "    end\n",
    "    \n",
    "    plot(p1)\n",
    "\n",
    "end\n",
    "\n",
    "gif(anim, \"dimOneGrowing.gif\", fps = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anim = @animate for i in 1:279\n",
    "    scatter(dimOne[i,:,1], dimOne[i,:,2], xlims = (0,2400), ylims = (0,1), title = string(\"Neuron #\", i,))\n",
    "end\n",
    "\n",
    "gif(anim, \"demo.gif\", fps = 1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plots(size=(800,800), legend=false)  # set backend and set some session defaults\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyplot(size=(800,800), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(dimOne[now,:,1], dimOne[now,:,2], markerstrokewidth=0,\n",
    "        m = :blues,\n",
    "        zcolor = rand(278) # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyplot(size=(800,800), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(dimTwo[now,:,1], dimTwo[now,:,2], markerstrokewidth=0,\n",
    "        m = :blues, zcolor = rand(278) # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plotlyjs(size=(400,200), legend=false)  # set backend and set some session defaults\n",
    "\n",
    "scatter(rand(30), alpha = 1, markerstrokewidth=0,\n",
    "        m= :reds,\n",
    "        zcolor = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30] # sample from the gradient, cycling through: 0, 0.5, 1\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scatter(fill(randn(10), 6), fill(randn(10), 6), framestyle=[:box :semi :origin :zerolines :grid :none], title=[\":box\" \":semi\" \":origin\" \":zerolines\" \":grid\" \":none\"], color=permutedims(1:6), layout=6, label=\"\", markerstrokewidth=0, ticks=-2:2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try to categorize unfriendly nodes by adjacent location on matrix deltas\n",
    "topologicalSimilarityDeltas  = zeros(279,276) #nnodes x ndeltas size\n",
    "reductThisIntoThat = zeros(279,278)\n",
    "\n",
    "#iterates over the nodes\n",
    "for i in 1:279\n",
    "    adjustment = 0\n",
    "    \n",
    "    #iterate over all pairs of nodes including i\n",
    "    for j in 1:278\n",
    "           \n",
    "        #to avoid having to 'switch directions,' take the max of the symmetric pair\n",
    "        first = maximum([topologicalSimilarity[i,j], topologicalSimilarity[j,i]])\n",
    "        second = maximum([topologicalSimilarity[i,j+1], topologicalSimilarity[j+1,i]])\n",
    "        delta = abs(first - second)\n",
    "        reductThisIntoThat[i,j] = delta\n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "#remove deltas in which (i,i) contributed\n",
    "for i in 2:278\n",
    "    counter = 0\n",
    "    \n",
    "    for j in 1:278\n",
    "        if j < i - 1\n",
    "            counter += 1\n",
    "            topologicalSimilarityDeltas[i,counter] = reductThisIntoThat[i,j]\n",
    "        end\n",
    "        \n",
    "        if j > i \n",
    "            counter += 1\n",
    "            topologicalSimilarityDeltas[i,counter] = reductThisIntoThat[i,j]\n",
    "        end\n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "#the above construction is problematic for nodes 1 and 279, so we store them in a separate matrix\n",
    "for k in 1:276\n",
    "    topologicalSimilarityDeltas[1,k] = reductThisIntoThat[1,k+1]\n",
    "    topologicalSimilarityDeltas[279,k] = reductThisIntoThat[279,k]\n",
    "end\n",
    "\n",
    "\n",
    "println(\"output: topologicalSimilarityDeltas and deltasForExtrema\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute variance and mean for each node's set of deltas\n",
    "nodeDeltaMeanAndVaraince = zeros(279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    nodeDeltaMeanAndVaraince[i,1] = mean(topologicalSimilarityDeltas[i,:])\n",
    "    nodeDeltaMeanAndVaraince[i,2] = var(topologicalSimilarityDeltas[i,:])\n",
    "end\n",
    "\n",
    "println(\"output: nodeDeltaMeanAndVaraince\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots delta mean and delta varaince\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 2)\n",
    "plotArray[1] = nodeDeltaMeanAndVaraince[:,1]\n",
    "plotArray[2] = nodeDeltaMeanAndVaraince[:,2]\n",
    "\n",
    "\n",
    "\n",
    "violin([\"Delta Mean\" \"Delta Variance\" ],plotArray,leg=false, color = theme(:dark))\n",
    "boxplot!([\"Delta Mean\" \"Delta Variance\"],plotArray,leg=false, color = theme(:dark))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot distance/similarity pairs\n",
    "scatter(nodeDeltaMeanAndVaraince[:,1], nodeDeltaMeanAndVaraince[:,2], legend = false, title = \"Delta Mean vs Delta Variance\", xlabel = \"delta mean\", ylabel = \"delta variance\", color = :grey)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute correlation between distance and topological similarity\n",
    "cor(nodeDeltaMeanAndVaraince[:,1], nodeDeltaMeanAndVaraince[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newTopSim = topologicalSimilarity[:,:]\n",
    "newDeltaMean = nodeDeltaMeanAndVaraince[:,2]\n",
    "\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && newDeltaMean[i - movement] < newDeltaMean[i-movement - 1]\n",
    "        shadowCache = newDeltaMean[i - movement]\n",
    "        newDeltaMean[i - movement] = newDeltaMean[i - movement - 1]\n",
    "        newDeltaMean[i - movement - 1] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[i - movement, :]\n",
    "        newTopSim[i - movement, :] = newTopSim[i - movement - 1, :]\n",
    "        newTopSim[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[:, i - movement]\n",
    "        newTopSim[:, i - movement] = newTopSim[:, i - movement - 1]\n",
    "        newTopSim[:, i - movement - 1] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if newTopSim[j,i] > 0\n",
    "            newTopSim[i,j] = newTopSim[j,i]\n",
    "            newTopSim[j,i] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "heatmap(newTopSim, yflip = false, aspect_ratio=:equal, title = \"One-Two Dimensional Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newDeltaMean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do summed similarity calculations\n",
    "summedSimilarity = zeros(279)\n",
    "\n",
    "for i in 1:279\n",
    "    summedSimilarity[i] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[:,i])\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots delta mean and delta varaince\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 1)\n",
    "plotArray[1] = summedSimilarity\n",
    "\n",
    "\n",
    "\n",
    "violin([\"Summed Similarity\" ],plotArray,leg=false, color = theme(:dark))\n",
    "boxplot!([\"Summed Similarity\"],plotArray,leg=false, color = theme(:dark))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newTopSim = topologicalSimilarity[:,:]\n",
    "newDeltaMean = summedSimilarity[:]\n",
    "\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && newDeltaMean[i - movement] < newDeltaMean[i-movement - 1]\n",
    "        shadowCache = newDeltaMean[i - movement]\n",
    "        newDeltaMean[i - movement] = newDeltaMean[i - movement - 1]\n",
    "        newDeltaMean[i - movement - 1] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[i - movement, :]\n",
    "        newTopSim[i - movement, :] = newTopSim[i - movement - 1, :]\n",
    "        newTopSim[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[:, i - movement]\n",
    "        newTopSim[:, i - movement] = newTopSim[:, i - movement - 1]\n",
    "        newTopSim[:, i - movement - 1] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if newTopSim[j,i] > 0\n",
    "            newTopSim[i,j] = newTopSim[j,i]\n",
    "            newTopSim[j,i] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "heatmap(newTopSim, yflip = false, aspect_ratio=:equal, title = \"One-Two Dimensional Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#end of reorderability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#beginning of null models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate 1000 erdos renyi \n",
    "nTrials = 1000\n",
    "prob = 2287 / 38781\n",
    "\n",
    "erdosRenyiTrails = zeros(279,279,nTrials)\n",
    "fill!(erdosRenyiTrails, 3000)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    \n",
    "    for i in 1:279\n",
    "       erdosRenyiTrails[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "    end\n",
    "    \n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            if rand() <= prob\n",
    "                erdosRenyiTrails[i,j,k] = labelsTimesPositionsType[i,2]\n",
    "                erdosRenyiTrails[j,i,k] = labelsTimesPositionsType[i,2]\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "print(\"output: erdosRenyiTrials, prob, nTrials\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine persistent homology of erdos renyi\n",
    "\n",
    "begin\n",
    "    randomCurves = pmap(i -> computeBarcodesNullModels(erdosRenyiTrails[:,:,i], times), 1:nTrials)\n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "erdosRenyiBettiCurves = zeros(numValues,3,nTrials)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    erdosRenyiBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "            \n",
    "    \n",
    "println(\"output: erdosRenyiBettiCurves, randomCurves\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of nTrials copies\n",
    "bettiMeans = zeros(numValues,3)\n",
    "bettiStd = zeros(numValues,3)\n",
    "\n",
    "for i in 1:numValues\n",
    "    bettiMeans[i,1] = mean(erdosRenyiBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(erdosRenyiBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(erdosRenyiBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(erdosRenyiBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(erdosRenyiBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(erdosRenyiBettiCurves[i,3,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Erdos Renyi\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0,15))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "\n",
    "plot(p1,p2, layout = (2,1), size = (800,800))\n",
    "\n",
    "#savefig(\"erdosRenyiTrials.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "computeRates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reorderability of erdos renyi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute reorderability for nTrials trials\n",
    "nTrials = 5\n",
    "\n",
    "\n",
    "begin\n",
    "    trials = pmap(i -> nullModelTopologicalSimilarityStorage(erdosRenyiTrails[:,:,i], i, \"Erdos Renyi\"), 1:nTrials)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute reorderability for nTrials trials\n",
    "nTrials = 5\n",
    "\n",
    "ERdimZeroTrials = zeros(279,279,nTrials)\n",
    "ERdimOneTrials = zeros(279,279,nTrials)\n",
    "ERdimTwoTrials = zeros(279,279,nTrials)\n",
    "\n",
    "begin\n",
    "    trials = pmap(i -> nullModelTopologicalSimilarityStorage(erdosRenyiTrails[:,:,i], i, \"Erdos Renyi\"), 1:nTrials)\n",
    "end\n",
    "\n",
    "\n",
    "#time to unpack this stuff\n",
    "for k in 1:nTrials\n",
    "    ERdimZeroTrials[:,:,k] = trials[k][1]\n",
    "    ERdimOneTrials[:,:,k] = trials[k][2]\n",
    "    ERdimTwoTrials[:,:,k] = trials[k][3]\n",
    "end\n",
    "\n",
    "aveDimZeroTopSim = zeros(279,279)\n",
    "aveDimOneTopSim = zeros(279,279)\n",
    "aveDimTwoTopSim = zeros(279,279)\n",
    "\n",
    "stdDimZeroTopSim = zeros(279,279)\n",
    "stdDimOneTopSim = zeros(279,279)\n",
    "stdDimTwoTopSim = zeros(279,279)\n",
    "\n",
    "aveDimZeroBND = zeros(279,279)\n",
    "aveDimOneBND = zeros(279,279)\n",
    "aveDimTwoBND = zeros(279,279)\n",
    "\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        aveDimZeroTopSim[i,j] = mean(ERdimZeroTrials[i,j,:])\n",
    "        aveDimOneTopSim[i,j] = mean(ERdimOneTrials[i,j,:])\n",
    "        aveDimTwoTopSim[i,j] = mean(ERdimTwoTrials[i,j,:])\n",
    "        \n",
    "        stdDimZeroTopSim[i,j] = std(ERdimZeroTrials[i,j,:])\n",
    "        stdDimOneTopSim[i,j] = std(ERdimOneTrials[i,j,:])\n",
    "        stdDimTwoTopSim[i,j] = std(ERdimTwoTrials[i,j,:])\n",
    "        \n",
    "        timeDiff = abs(labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2])\n",
    "        \n",
    "        if timeDiff == 0\n",
    "            aveDimZeroBND[i,j] = 0\n",
    "            aveDimOneBND[i,j] = 0\n",
    "            aveDimTwoBND[i,j] = 0\n",
    "        else\n",
    "            aveDimZeroBND[i,j] = timeDiff * (1 - aveDimZeroTopSim[i,j])\n",
    "            aveDimOneBND[i,j] = timeDiff * (1 - aveDimOneTopSim[i,j])\n",
    "            aveDimTwoBND[i,j] = timeDiff * (1 - aveDimTwoTopSim[i,j])\n",
    "        end\n",
    "        \n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save these cause it takes a big while to compute these things\n",
    "\n",
    "CSV.write(\"OneErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,1]), writeheader = false)\n",
    "CSV.write(\"TwoErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,2]), writeheader = false)\n",
    "CSV.write(\"ThreeErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,3]), writeheader = false)\n",
    "CSV.write(\"FourErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,4]), writeheader = false)\n",
    "CSV.write(\"FiveErdosRenyiDimZero.csv\", DataFrame( ERdimZeroTrials[:,:,5]), writeheader = false)\n",
    "\n",
    "CSV.write(\"OneErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,1]), writeheader = false)\n",
    "CSV.write(\"TwoErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,2]), writeheader = false)\n",
    "CSV.write(\"ThreeErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,3]), writeheader = false)\n",
    "CSV.write(\"FourErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,4]), writeheader = false)\n",
    "CSV.write(\"FiveErdosRenyiDimOne.csv\", DataFrame(ERdimOneTrials[:,:,5]), writeheader = false)\n",
    "\n",
    "CSV.write(\"OneErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,1]), writeheader = false)\n",
    "CSV.write(\"TwoErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,2]), writeheader = false)\n",
    "CSV.write(\"ThreeErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,3]), writeheader = false)\n",
    "CSV.write(\"FourErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,4]), writeheader = false)\n",
    "CSV.write(\"FiveErdosRenyiDimTwo.csv\", DataFrame(ERdimTwoTrials[:,:,5]), writeheader = false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimZeroTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Erdos Renyi\", color = :tempo)\n",
    "p2 = heatmap(aveDimZeroBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Erdos Renyi\", color = :grays)\n",
    "p3 = heatmap(stdDimZeroTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Erdos Renyi\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionZeroSimilarityErdosRenyi.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimOneTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension One Erdos Renyi\", color = :tempo)\n",
    "p2 = heatmap(aveDimOneBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Erdos Renyi\", color = :grays)\n",
    "p3 = heatmap(stdDimOneTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension One Erdos Renyi\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionOneSimilarityErdosRenyi.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimTwoTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Erdos Renyi\", color = :tempo)\n",
    "p2 = heatmap(aveDimTwoBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Erdos Renyi\", color = :grays)\n",
    "p3 = heatmap(stdDimTwoTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Erdos Renyi\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionTwoSimilarityErdosRenyi.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimzero, dimone, dimtwo = nullModelTopologicalSimilarity(birthSortedAdjacencyMatrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap(dimzero, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap(dimone, yflip = true, aspect_ratio=:equal, title = \"Dimension One Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap(dimtwo, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cleanup\n",
    "nTrials = nothing\n",
    "prob = nothing\n",
    "erdosRenyiTrails = nothing\n",
    "randomCurves = nothing\n",
    "erdosRenyiBettiCurves = nothing\n",
    "bettiMeans = nothing\n",
    "bettiStd = nothing\n",
    "ERdimZeroTrials = nothing\n",
    "ERdimOneTrials = nothing\n",
    "ERdimTwoTrials = nothing\n",
    "trials = nothing\n",
    "aveDimZeroTopSim = nothing\n",
    "aveDimOneTopSim = nothing\n",
    "aveDimTwoTopSim = nothing\n",
    "stdDimZeroTopSim = nothing\n",
    "stdDimOneTopSim = nothing\n",
    "stdDimTwoTopSim = nothing\n",
    "aveDimZeroBND = nothing\n",
    "aveDimOneBND = nothing\n",
    "aveDimTwoBND = nothing\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "p3 = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate 1000 spatial growth\n",
    "nTrials = 1000\n",
    "threshold = 2287\n",
    "spatialGrowthTrials = zeros(279,279,nTrials)\n",
    "fill!(spatialGrowthTrials, 3000)\n",
    "\n",
    "\n",
    "for k in 1:nTrials\n",
    "    \n",
    "    for i in 1:279\n",
    "       spatialGrowthTrials[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "    end\n",
    "    \n",
    "    #spawn 279 random points\n",
    "    points = zeros(279,2)\n",
    "    for i in 1:279\n",
    "        points[i,:] = [rand(0:1000), rand(0:40)]\n",
    "    end\n",
    "    \n",
    "    #find distances between all points\n",
    "    fullDistances = zeros(38781, 3)\n",
    "    counter = 0\n",
    "    for i in 2:279\n",
    "        for j in 1:i-1\n",
    "            counter += 1\n",
    "            \n",
    "            a = points[i,1]\n",
    "            b = points[i,2]\n",
    "            c = points[j,1]\n",
    "            d = points[j,2]\n",
    "            \n",
    "            fullDistances[counter,:] = [evaluate(Euclidean(),[a,b], [c,d]), i, j]\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    #sort distances\n",
    "    fullDistances = sortslices(fullDistances, dims = 1)\n",
    "    \n",
    "    #add connections between closest 2287 connections\n",
    "    for l in 1:2287\n",
    "        i = trunc(Int, fullDistances[l,2])\n",
    "        j = trunc(Int, fullDistances[l,3])\n",
    "\n",
    "        spatialGrowthTrials[i,j,k] = labelsTimesPositionsType[i,2]\n",
    "        spatialGrowthTrials[j,i,k] = labelsTimesPositionsType[i,2]      \n",
    "    end\n",
    "            \n",
    "end\n",
    "print(\"output: spatialGrowthTrials, threshold, nTrials\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine persistent homology of spatial growth\n",
    "\n",
    "begin\n",
    "randomCurves = pmap(i -> computeBarcodesNullModels(spatialGrowthTrials[:,:,i], times), 1:nTrials)\n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "spatialGrowthBettiCurves = zeros(numValues,3,nTrials)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    spatialGrowthBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "            \n",
    "    \n",
    "println(\"output: spatialGrowthBettiCurves\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of nTrials copies\n",
    "bettiMeans = zeros(numValues,3)\n",
    "bettiStd = zeros(numValues,3)\n",
    "\n",
    "for i in 1:numValues\n",
    "    bettiMeans[i,1] = mean(spatialGrowthBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(spatialGrowthBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(spatialGrowthBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(spatialGrowthBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(spatialGrowthBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(spatialGrowthBettiCurves[i,3,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Spatial Growth\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0, 20))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "\n",
    "plot(p1,p2, layout = (2,1), size = (800,800))\n",
    "\n",
    "#savefig(\"spatialGrowthTrials.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute reorderability for nTrials trials\n",
    "nTrials = 5\n",
    "\n",
    "\n",
    "begin\n",
    "    trials = pmap(i -> nullModelTopologicalSimilarityStorage(spatialGrowthTrials[:,:,i], i, \"Spatial Growth\"), 1:nTrials)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimZeroTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Spatial Growth\", color = :tempo)\n",
    "p2 = heatmap(aveDimZeroBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Spatial Growth\", color = :grays)\n",
    "p3 = heatmap(stdDimZeroTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Spatial Growth\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionZeroSimilaritySpatialGrowth.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimOneTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension One Spatial Growth\", color = :tempo)\n",
    "p2 = heatmap(aveDimOneBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Spatial Growth\", color = :grays)\n",
    "p3 = heatmap(stdDimOneTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension One Spatial Growth\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionOneSimilaritySpatialGrowth.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(aveDimTwoTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Spatial Growth\", color = :tempo)\n",
    "p2 = heatmap(aveDimTwoBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Spatial Growth\", color = :grays)\n",
    "p3 = heatmap(stdDimTwoTopSim, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Spatial Growth\", color = :purples)\n",
    "plot(p1, p2, p3,  layout = (1,3))\n",
    "#savefig(\"dimensionTwoSimilaritySpatialGrowth.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cleanup\n",
    "nTrials = nothing\n",
    "prob = nothing\n",
    "spatialGrowthTrials = nothing\n",
    "randomCurves = nothing\n",
    "spatialGrowthBettiCurves = nothing\n",
    "bettiMeans = nothing\n",
    "bettiStd = nothing\n",
    "SGdimZeroTrials = nothing\n",
    "SGdimOneTrials = nothing\n",
    "SGdimTwoTrials = nothing\n",
    "trials = nothing\n",
    "aveDimZeroTopSim = nothing\n",
    "aveDimOneTopSim = nothing\n",
    "aveDimTwoTopSim = nothing\n",
    "stdDimZeroTopSim = nothing\n",
    "stdDimOneTopSim = nothing\n",
    "stdDimTwoTopSim = nothing\n",
    "aveDimZeroBND = nothing\n",
    "aveDimOneBND = nothing\n",
    "aveDimTwoBND = nothing\n",
    "p1 = nothing\n",
    "p2 = nothing\n",
    "p3 = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate 1000 rewiring trials\n",
    "nTrials = 1000\n",
    "threshold = 2287\n",
    "rewiredTrials = zeros(279,279,nTrials)\n",
    "fill!(rewiredTrials, 3000)\n",
    "\n",
    "#find set of all edges in the graph\n",
    "edges = Array{Int64, 2}(undef, 2287, 2)\n",
    "edgeCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        if birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "            edgeCounter += 1\n",
    "            edges[edgeCounter,:] = [i,j]\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "for k in 1:nTrials\n",
    "    print(k, \" \")\n",
    "    for i in 1:279\n",
    "       rewiredTrials[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "    end\n",
    "    \n",
    "    #perform 22870 random rewirings\n",
    "    for i in 1:22870\n",
    "        \n",
    "        redundant = true\n",
    "        \n",
    "        while redundant\n",
    "            redundant = false\n",
    "            \n",
    "            initial = rand(1:2287)\n",
    "            terminal = rand(1:2287)\n",
    "            parity = rand(1:2)\n",
    "        \n",
    "           while (initial == terminal)\n",
    "                terminal = rand(1:2287)\n",
    "            end\n",
    "           \n",
    "            if parity == 0\n",
    "                a = edges[initial, 1]\n",
    "                d = edges[initial, 2]\n",
    "                c = edges[terminal, 1]\n",
    "                b = edges[terminal, 2]\n",
    "            else\n",
    "                a = edges[initial, 1]\n",
    "                c = edges[initial, 2]\n",
    "                b = edges[terminal, 1]\n",
    "                d = edges[terminal, 2]\n",
    "            end\n",
    "            \n",
    "            for i in 1:2287\n",
    "                if edges[i,:] == [initial, terminal] || edges[i,:] == [terminal, initial]\n",
    "                    redundant = true\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            if !redundant\n",
    "               edges[initial, 1] = a\n",
    "               edges[initial, 2] = b\n",
    "               edges[terminal, 1] = c\n",
    "               edges[terminal, 2] = d\n",
    "            end\n",
    "        end\n",
    "            \n",
    "    end\n",
    "    \n",
    "    #add edges\n",
    "    for i in 1:2287\n",
    "        max = maximum(edges[i,:])\n",
    "        \n",
    "        rewiredTrials[edges[i,1],edges[i,2],k] = labelsTimesPositionsType[max,2]\n",
    "        rewiredTrials[edges[i,2],edges[i,1],k] = labelsTimesPositionsType[max,2] \n",
    "    end\n",
    "end\n",
    "print(\"output: rewiredTrials, threshold, nTrials\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine persistent homology of spatial growth\n",
    "\n",
    "begin\n",
    "randomCurves = pmap(i -> computeBarcodesNullModels(rewiredTrials[:,:,i], times), 1:nTrials)\n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "rewiredBettiCurves = zeros(numValues,3,nTrials)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    rewiredBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "            \n",
    "    \n",
    "println(\"output: rewiredBettiCurves\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of nTrials copies\n",
    "bettiMeans = zeros(numValues,3)\n",
    "bettiStd = zeros(numValues,3)\n",
    "\n",
    "for i in 1:numValues\n",
    "    bettiMeans[i,1] = mean(rewiredBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(rewiredBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(rewiredBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(rewiredBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(rewiredBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(rewiredBettiCurves[i,3,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Identical Degree Distribution\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0, 20))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "\n",
    "plot(p1,p2, layout = (2,1), size = (800,800))\n",
    "\n",
    "#savefig(\"rewiredTrials.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate 1000 rewiring trials\n",
    "nTrials = 1000\n",
    "configurationTrials = zeros(279,279,nTrials)\n",
    "fill!(configurationTrials, 3000)\n",
    "\n",
    "#find set of all edges in the graph\n",
    "degreeDistribution = Array{Int64, 1}(undef, 279)\n",
    "for i in 1:279\n",
    "    degree = 0\n",
    "    \n",
    "    for j in 1:279\n",
    "        if j != i && birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "            degree += 1\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    degreeDistribution[i] = degree\n",
    "end\n",
    "\n",
    "\n",
    "for k in 1:nTrials\n",
    "    print(k, \"  \")\n",
    "    for i in 1:279\n",
    "       configurationTrials[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "    end\n",
    "    \n",
    "    ddHolder = degreeDistribution[:]\n",
    "    \n",
    "    errorCounter = 0\n",
    "    \n",
    "   while sum(ddHolder) > 0\n",
    "        \n",
    "        initial = rand(1:279)\n",
    "        terminal = rand(1:279)\n",
    "        \n",
    "        if ddHolder[initial] != 0 && ddHolder[terminal] != 0\n",
    "            if configurationTrials[initial, terminal,k] == 3000\n",
    "                \n",
    "                max = maximum([initial, terminal])\n",
    "                ddHolder[initial] -= 1\n",
    "                ddHolder[terminal] -= 1\n",
    "                configurationTrials[initial, terminal,k] = labelsTimesPositionsType[max, 2]\n",
    "                configurationTrials[terminal, initial,k] = labelsTimesPositionsType[max, 2]\n",
    "            end\n",
    "            \n",
    "        end\n",
    "        \n",
    "        errorCounter += 1\n",
    "        \n",
    "        if errorCounter == 10000\n",
    "            errorCounter = 0\n",
    "            fill!(configurationTrials[:,:,k], 3000)\n",
    "            ddHolder = degreeDistribution[:]\n",
    "            \n",
    "            for i in 1:279\n",
    "               configurationTrials[i,i,k] = labelsTimesPositionsType[i,2]\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "print(\"output: configurationTrials, nTrials\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine persistent homology of spatial growth\n",
    "\n",
    "begin\n",
    "randomCurves = pmap(i -> computeBarcodesNullModels(configurationTrials[:,:,i], times), 1:nTrials)\n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "rewiredBettiCurves = zeros(numValues,3,nTrials)\n",
    "\n",
    "for k in 1:nTrials\n",
    "    configurationBettiCurves[:,:,k] = randomCurves[k]\n",
    "end\n",
    "            \n",
    "    \n",
    "println(\"output: configurationBettiCurves\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes average and standard deviation of nTrials copies\n",
    "bettiMeans = zeros(numValues,3)\n",
    "bettiStd = zeros(numValues,3)\n",
    "\n",
    "for i in 1:numValues\n",
    "    bettiMeans[i,1] = mean(configurationBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(configurationBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(configurationBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(configurationBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(configurationBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(configurationBettiCurves[i,3,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"Configurationj\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0, 20))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "\n",
    "plot(p1,p2, layout = (2,1), size = (800,800))\n",
    "\n",
    "savefig(\"configurationTrials.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": "5d83d9ce-a431-48f1-a173-dd2eed910831",
   "lastKernelId": "28baed09-383e-4905-acba-f8382fcbba90"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
