{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "// Immediately-invoked-function-expression to avoid global variables.\n",
       "(function() {\n",
       "    var warning_div = document.getElementById(\"webio-warning-10217028534412859812\");\n",
       "    var hide = function () {\n",
       "        var script = document.getElementById(\"webio-setup-13167766282558689781\");\n",
       "        var parent = script && script.parentElement;\n",
       "        var grandparent = parent && parent.parentElement;\n",
       "        if (grandparent) {\n",
       "            grandparent.style.display = \"none\";\n",
       "        }\n",
       "        warning_div.style.display = \"none\";\n",
       "    };\n",
       "    if (typeof Jupyter !== \"undefined\") {\n",
       "        console.log(\"WebIO detected Jupyter notebook environment.\");\n",
       "        // Jupyter notebook.\n",
       "        var extensions = (\n",
       "            Jupyter\n",
       "            && Jupyter.notebook.config.data\n",
       "            && Jupyter.notebook.config.data.load_extensions\n",
       "        );\n",
       "        if (extensions && extensions[\"webio-jupyter-notebook\"]) {\n",
       "            // Extension already loaded.\n",
       "            console.log(\"Jupyter WebIO nbextension detected; not loading ad-hoc.\");\n",
       "            hide();\n",
       "            return;\n",
       "        }\n",
       "    } else if (window.location.pathname.includes(\"/lab\")) {\n",
       "        // Guessing JupyterLa\n",
       "        console.log(\"Jupyter Lab detected; make sure the @webio/jupyter-lab-provider labextension is installed.\");\n",
       "        hide();\n",
       "        return;\n",
       "    }\n",
       "})();\n",
       "\n",
       "</script>\n",
       "<p\n",
       "    id=\"webio-warning-10217028534412859812\"\n",
       "    class=\"output_text output_stderr\"\n",
       "    style=\"padding: 1em; font-weight: bold;\"\n",
       ">\n",
       "    Unable to load WebIO. Please make sure WebIO works for your Jupyter client.\n",
       "    For troubleshooting, please see <a href=\"https://juliagizmos.github.io/WebIO.jl/latest/providers/ijulia/\">\n",
       "    the WebIO/IJulia documentation</a>.\n",
       "    <!-- TODO: link to installation docs. -->\n",
       "</p>\n"
      ],
      "text/plain": [
       "HTML{String}(\"<script>\\n// Immediately-invoked-function-expression to avoid global variables.\\n(function() {\\n    var warning_div = document.getElementById(\\\"webio-warning-10217028534412859812\\\");\\n    var hide = function () {\\n        var script = document.getElementById(\\\"webio-setup-13167766282558689781\\\");\\n        var parent = script && script.parentElement;\\n        var grandparent = parent && parent.parentElement;\\n        if (grandparent) {\\n            grandparent.style.display = \\\"none\\\";\\n        }\\n        warning_div.style.display = \\\"none\\\";\\n    };\\n    if (typeof Jupyter !== \\\"undefined\\\") {\\n        console.log(\\\"WebIO detected Jupyter notebook environment.\\\");\\n        // Jupyter notebook.\\n        var extensions = (\\n            Jupyter\\n            && Jupyter.notebook.config.data\\n            && Jupyter.notebook.config.data.load_extensions\\n        );\\n        if (extensions && extensions[\\\"webio-jupyter-notebook\\\"]) {\\n            // Extension already loaded.\\n            console.log(\\\"Jupyter WebIO nbextension detected; not loading ad-hoc.\\\");\\n            hide();\\n            return;\\n        }\\n    } else if (window.location.pathname.includes(\\\"/lab\\\")) {\\n        // Guessing JupyterLa\\n        console.log(\\\"Jupyter Lab detected; make sure the @webio/jupyter-lab-provider labextension is installed.\\\");\\n        hide();\\n        return;\\n    }\\n})();\\n\\n</script>\\n<p\\n    id=\\\"webio-warning-10217028534412859812\\\"\\n    class=\\\"output_text output_stderr\\\"\\n    style=\\\"padding: 1em; font-weight: bold;\\\"\\n>\\n    Unable to load WebIO. Please make sure WebIO works for your Jupyter client.\\n    For troubleshooting, please see <a href=\\\"https://juliagizmos.github.io/WebIO.jl/latest/providers/ijulia/\\\">\\n    the WebIO/IJulia documentation</a>.\\n    <!-- TODO: link to installation docs. -->\\n</p>\\n\")"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "[ Info: waiting for lock on pidfile at \"C:\\Users\\alech\\.jlassetregistry.lock\"\n",
      "┌ Warning: Error requiring WebSockets from WebIO:\n",
      "│ LoadError: IOError: open: operation not permitted (EPERM)\n",
      "│ Stacktrace:\n",
      "│  [1] uv_error at .\\libuv.jl:85 [inlined]\n",
      "│  [2] open(::String, ::UInt16, ::UInt16) at .\\filesystem.jl:81\n",
      "│  [3] tryopen_exclusive(::String, ::UInt16) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:152\n",
      "│  [4] #open_exclusive#4(::UInt16, ::Int64, ::Int64, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:185\n",
      "│  [5] #open_exclusive at .\\none:0 [inlined]\n",
      "│  [6] (::getfield(Pidfile, Symbol(\"##mkpidlock#1#2\")))(::Base.Iterators.Pairs{Symbol,Int64,Tuple{Symbol},NamedTuple{(:stale_age,),Tuple{Int64}}}, ::Function, ::String, ::Int32) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:42\n",
      "│  [7] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [8] #mkpidlock#3 at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:56 [inlined]\n",
      "│  [9] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [10] withlock at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:10 [inlined]\n",
      "│  [11] #register#1(::String, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:49\n",
      "│  [12] register(::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:29\n",
      "│  [13] top-level scope at none:0\n",
      "│  [14] include at .\\boot.jl:317 [inlined]\n",
      "│  [15] include_relative(::Module, ::String) at .\\loading.jl:1044\n",
      "│  [16] include at .\\sysimg.jl:29 [inlined]\n",
      "│  [17] include(::String) at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1\n",
      "│  [18] top-level scope at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:111\n",
      "│  [19] eval at .\\boot.jl:319 [inlined]\n",
      "│  [20] eval at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1 [inlined]\n",
      "│  [21] (::getfield(WebIO, Symbol(\"##79#91\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:67\n",
      "│  [22] err(::getfield(WebIO, Symbol(\"##79#91\")), ::Module, ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:38\n",
      "│  [23] #78 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:66 [inlined]\n",
      "│  [24] withpath(::getfield(WebIO, Symbol(\"##78#90\")), ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:28\n",
      "│  [25] (::getfield(WebIO, Symbol(\"##77#89\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:65\n",
      "│  [26] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [27] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [28] #3 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19 [inlined]\n",
      "│  [29] iterate at .\\generator.jl:47 [inlined]\n",
      "│  [30] _collect(::Array{Function,1}, ::Base.Generator{Array{Function,1},getfield(Requires, Symbol(\"##3#4\"))}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at .\\array.jl:632\n",
      "│  [31] map at .\\array.jl:561 [inlined]\n",
      "│  [32] loadpkg(::Base.PkgId) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19\n",
      "│  [33] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [34] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [35] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:651\n",
      "│  [36] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [37] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [38] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [39] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [40] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [41] _require(::Base.PkgId) at .\\loading.jl:937\n",
      "│  [42] require(::Base.PkgId) at .\\loading.jl:858\n",
      "│  [43] (::getfield(Distributed, Symbol(\"##2#4\")){Base.PkgId})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\Distributed.jl:77\n",
      "│  [44] (::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:276\n",
      "│  [45] run_work_thunk(::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}, ::Bool) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:56\n",
      "│  [46] run_work_thunk(::Distributed.RemoteValue, ::Function) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:65\n",
      "│  [47] (::getfield(Distributed, Symbol(\"##102#104\")){Distributed.RemoteValue,getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}})() at .\\task.jl:259\n",
      "│ in expression starting at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\providers\\generic_http.jl:21\n",
      "└ @ Requires C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:40\n",
      "┌ Warning: Error requiring WebSockets from WebIO:\n",
      "│ LoadError: IOError: open: operation not permitted (EPERM)\n",
      "│ Stacktrace:\n",
      "│  [1] uv_error at .\\libuv.jl:85 [inlined]\n",
      "│  [2] open(::String, ::UInt16, ::UInt16) at .\\filesystem.jl:81\n",
      "│  [3] tryopen_exclusive(::String, ::UInt16) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:152\n",
      "│  [4] #open_exclusive#4(::UInt16, ::Int64, ::Int64, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:185\n",
      "│  [5] #open_exclusive at .\\none:0 [inlined]\n",
      "│  [6] (::getfield(Pidfile, Symbol(\"##mkpidlock#1#2\")))(::Base.Iterators.Pairs{Symbol,Int64,Tuple{Symbol},NamedTuple{(:stale_age,),Tuple{Int64}}}, ::Function, ::String, ::Int32) at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:42\n",
      "│  [7] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [8] #mkpidlock#3 at C:\\Users\\alech\\.julia\\packages\\Pidfile\\q0iHJ\\src\\Pidfile.jl:56 [inlined]\n",
      "│  [9] #mkpidlock at .\\none:0 [inlined]\n",
      "│  [10] withlock at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:10 [inlined]\n",
      "│  [11] #register#1(::String, ::Function, ::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:49\n",
      "│  [12] register(::String) at C:\\Users\\alech\\.julia\\packages\\AssetRegistry\\4TyKv\\src\\AssetRegistry.jl:29\n",
      "│  [13] top-level scope at none:0\n",
      "│  [14] include at .\\boot.jl:317 [inlined]\n",
      "│  [15] include_relative(::Module, ::String) at .\\loading.jl:1044\n",
      "│  [16] include at .\\sysimg.jl:29 [inlined]\n",
      "│  [17] include(::String) at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1\n",
      "│  [18] top-level scope at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:111\n",
      "│  [19] eval at .\\boot.jl:319 [inlined]\n",
      "│  [20] eval at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\WebIO.jl:1 [inlined]\n",
      "│  [21] (::getfield(WebIO, Symbol(\"##79#91\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:67\n",
      "│  [22] err(::getfield(WebIO, Symbol(\"##79#91\")), ::Module, ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:38\n",
      "│  [23] #78 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:66 [inlined]\n",
      "│  [24] withpath(::getfield(WebIO, Symbol(\"##78#90\")), ::String) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:28\n",
      "│  [25] (::getfield(WebIO, Symbol(\"##77#89\")))() at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:65\n",
      "│  [26] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [27] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [28] #3 at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19 [inlined]\n",
      "│  [29] iterate at .\\generator.jl:47 [inlined]\n",
      "│  [30] _collect(::Array{Function,1}, ::Base.Generator{Array{Function,1},getfield(Requires, Symbol(\"##3#4\"))}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at .\\array.jl:632\n",
      "│  [31] map at .\\array.jl:561 [inlined]\n",
      "│  [32] loadpkg(::Base.PkgId) at C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:19\n",
      "│  [33] #invokelatest#1 at .\\essentials.jl:697 [inlined]\n",
      "│  [34] invokelatest at .\\essentials.jl:696 [inlined]\n",
      "│  [35] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:651\n",
      "│  [36] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [37] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [38] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [39] _tryrequire_from_serialized(::Base.PkgId, ::UInt64, ::String) at .\\loading.jl:648\n",
      "│  [40] _require_search_from_serialized(::Base.PkgId, ::String) at .\\loading.jl:702\n",
      "│  [41] _require(::Base.PkgId) at .\\loading.jl:937\n",
      "│  [42] require(::Base.PkgId) at .\\loading.jl:858\n",
      "│  [43] (::getfield(Distributed, Symbol(\"##2#4\")){Base.PkgId})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\Distributed.jl:77\n",
      "│  [44] (::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg})() at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:276\n",
      "│  [45] run_work_thunk(::getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}, ::Bool) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:56\n",
      "│  [46] run_work_thunk(::Distributed.RemoteValue, ::Function) at C:\\cygwin\\home\\Administrator\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.0\\Distributed\\src\\process_messages.jl:65\n",
      "│  [47] (::getfield(Distributed, Symbol(\"##102#104\")){Distributed.RemoteValue,getfield(Distributed, Symbol(\"##116#118\")){Distributed.CallWaitMsg}})() at .\\task.jl:259\n",
      "│ in expression starting at C:\\Users\\alech\\.julia\\packages\\WebIO\\2mZPb\\src\\providers\\generic_http.jl:21\n",
      "└ @ Requires C:\\Users\\alech\\.julia\\packages\\Requires\\9Jse8\\src\\require.jl:40\n",
      "┌ Info: Recompiling stale cache file C:\\Users\\alech\\.julia\\compiled\\v1.0\\cElegansFunctions.ji for cElegansFunctions [top-level]\n",
      "└ @ Base loading.jl:1190\n",
      "WARNING: Method definition savehtml(IO, PlotlyBase.Plot{TT} where TT<:PlotlyBase.AbstractTrace) in module PlotlyBase at C:\\Users\\alech\\.julia\\packages\\PlotlyBase\\80KwD\\src\\output.jl:40 overwritten in module PlotlyJS at C:\\Users\\alech\\.julia\\packages\\PlotlyJS\\b9Efu\\src\\display.jl:334.\n",
      "WARNING: Method definition savehtml(PlotlyBase.Plot{TT} where TT<:PlotlyBase.AbstractTrace, AbstractString) in module PlotlyBase at C:\\Users\\alech\\.julia\\packages\\PlotlyBase\\80KwD\\src\\output.jl:52 overwritten in module PlotlyJS at C:\\Users\\alech\\.julia\\packages\\PlotlyJS\\b9Efu\\src\\display.jl:361.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "R version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "Copyright (C) 2019 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "\n",
      "R version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "Copyright (C) 2019 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "      From worker 8:\t\n",
      "      From worker 3:\t\n",
      "      From worker 3:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 3:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 3:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 3:\t\n",
      "      From worker 3:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 3:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 3:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 3:\t\n",
      "      From worker 3:\tR is a collaborative project with many contributors.\n",
      "      From worker 3:\tType 'contributors()' for more information and\n",
      "      From worker 3:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 3:\t\n",
      "      From worker 3:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 3:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 3:\tType 'q()' to quit R.\n",
      "      From worker 3:\t\n",
      "      From worker 8:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 8:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 8:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 8:\t\n",
      "      From worker 8:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 8:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 8:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 8:\t\n",
      "      From worker 8:\tR is a collaborative project with many contributors.\n",
      "      From worker 8:\tType 'contributors()' for more information and\n",
      "      From worker 8:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 8:\t\n",
      "      From worker 8:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 8:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 8:\tType 'q()' to quit R.\n",
      "      From worker 8:\t\n",
      "      From worker 7:\t\n",
      "      From worker 7:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 7:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 7:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 7:\t\n",
      "      From worker 7:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 7:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 7:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 7:\t\n",
      "      From worker 7:\tR is a collaborative project with many contributors.\n",
      "      From worker 7:\tType 'contributors()' for more information and\n",
      "      From worker 7:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 7:\t\n",
      "      From worker 7:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 7:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 7:\tType 'q()' to quit R.\n",
      "      From worker 7:\t\n",
      "      From worker 2:\t\n",
      "      From worker 5:\t\n",
      "      From worker 5:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 5:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 5:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 5:\t\n",
      "      From worker 5:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 5:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 5:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 5:\t\n",
      "      From worker 5:\tR is a collaborative project with many contributors.\n",
      "      From worker 5:\tType 'contributors()' for more information and\n",
      "      From worker 5:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 5:\t\n",
      "      From worker 5:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 5:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 5:\tType 'q()' to quit R.\n",
      "      From worker 5:\t\n",
      "      From worker 6:\t\n",
      "      From worker 6:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 6:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 6:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 6:\t\n",
      "      From worker 6:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 6:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 6:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 6:\t\n",
      "      From worker 6:\tR is a collaborative project with many contributors.\n",
      "      From worker 6:\tType 'contributors()' for more information and\n",
      "      From worker 6:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 6:\t\n",
      "      From worker 6:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 6:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 6:\tType 'q()' to quit R.\n",
      "      From worker 6:\t\n",
      "      From worker 4:\t\n",
      "      From worker 4:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 4:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 4:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 4:\t\n",
      "      From worker 4:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 4:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 4:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 4:\t\n",
      "      From worker 4:\tR is a collaborative project with many contributors.\n",
      "      From worker 4:\tType 'contributors()' for more information and\n",
      "      From worker 4:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 4:\t\n",
      "      From worker 4:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 4:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 4:\tType 'q()' to quit R.\n",
      "      From worker 4:\t\n",
      "      From worker 2:\tR version 3.6.1 (2019-07-05) -- \"Action of the Toes\"\n",
      "      From worker 2:\tCopyright (C) 2019 The R Foundation for Statistical Computing\n",
      "      From worker 2:\tPlatform: x86_64-w64-mingw32/x64 (64-bit)\n",
      "      From worker 2:\t\n",
      "      From worker 2:\tR is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "      From worker 2:\tYou are welcome to redistribute it under certain conditions.\n",
      "      From worker 2:\tType 'license()' or 'licence()' for distribution details.\n",
      "      From worker 2:\t\n",
      "      From worker 2:\tR is a collaborative project with many contributors.\n",
      "      From worker 2:\tType 'contributors()' for more information and\n",
      "      From worker 2:\t'citation()' on how to cite R or R packages in publications.\n",
      "      From worker 2:\t\n",
      "      From worker 2:\tType 'demo()' for some demos, 'help()' for on-line help, or\n",
      "      From worker 2:\t'help.start()' for an HTML browser interface to help.\n",
      "      From worker 2:\tType 'q()' to quit R.\n",
      "      From worker 2:\t\n"
     ]
    }
   ],
   "source": [
    "#load neccesary packages\n",
    "using CSV\n",
    "using DataFrames\n",
    "using Plots\n",
    "using Statistics\n",
    "using StatsPlots\n",
    "using RDatasets\n",
    "using HypothesisTests\n",
    "using StatsBase\n",
    "using VegaLite\n",
    "using Distances\n",
    "using Distributed\n",
    "addprocs(7)\n",
    "@everywhere using Eirene   \n",
    "@everywhere using SharedArrays\n",
    "\n",
    "\n",
    "@everywhere push!(LOAD_PATH, $\"E:functions/\") \n",
    "@everywhere using cElegansFunctions\n",
    "@everywhere using RCall\n",
    "\n",
    "\n",
    "@everywhere R\"library(TDA)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "#load neccesary data files\n",
    "positions = CSV.read(\"E://From Gene Expression File//279positions.csv\", copycols = true, header = false)\n",
    "positionsLabels = CSV.read(\"E://From Gene Expression File//279positionsLabels.csv\", copycols = true, header = false)\n",
    "birthTimes = CSV.read(\"E://From Gene Expression File//279birthTimes.csv\", copycols = true, header = false)\n",
    "birthTimesLabels = CSV.read(\"E://From Gene Expression File//279birthTimesLabels.csv\", copycols = true, header = false)\n",
    "connectionList = CSV.read(\"E://From Gene Expression File//NeuronConnect_corrected.csv\", copycols = true, header = false)\n",
    "labeledTypeList = CSV.read(\"E://From Gene Expression File//279labeledTypes.csv\", copycols = true, header = false)\n",
    "\n",
    "#done\n",
    "println(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[15]:7\n",
      "└ @ Core .\\In[15]:7\n",
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[15]:16\n",
      "└ @ Core .\\In[15]:16\n",
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[15]:17\n",
      "└ @ Core .\\In[15]:17\n",
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[15]:24\n",
      "└ @ Core .\\In[15]:24\n",
      "┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.\n",
      "│   caller = top-level scope at In[15]:25\n",
      "└ @ Core .\\In[15]:25\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: labelsTimesPositionsType, birthSortedAdjacencyMatrix, birthOrder, and times\n"
     ]
    }
   ],
   "source": [
    "#process all the data into a single array sorted by birth time\n",
    "\n",
    "#creates an array for positions with their labels\n",
    "positionsMatrix = Array{Any, 2}(UndefInitializer(),279,3)\n",
    "\n",
    "for i in 1:279\n",
    "    positionsMatrix[i,1] = positionsLabels[1][i]\n",
    "    positionsMatrix[i,2] = positions[i,1]\n",
    "    positionsMatrix[i,3] = positions[i,2]\n",
    "end\n",
    "\n",
    "#creates an array for positions with their labels\n",
    "birthTimesMatrix = Array{Any, 2}(UndefInitializer(),279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    birthTimesMatrix[i,1] = birthTimesLabels[1][i]\n",
    "    birthTimesMatrix[i,2] = birthTimes[1][i]\n",
    "end\n",
    "\n",
    "#creates an array for types with their labels\n",
    "typeMatrix = Array{Any, 2}(UndefInitializer(), 279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    typeMatrix[i,1] = labeledTypeList[1][i]\n",
    "    typeMatrix[i,2] = labeledTypeList[2][i]\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#sorts all the above matricies by neuron label to make them compatible\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && positionsMatrix[i - movement, 1] < positionsMatrix[i-movement - 1, 1]\n",
    "        shadowCache = positionsMatrix[i - movement, :]\n",
    "        positionsMatrix[i - movement, :] = positionsMatrix[i - movement - 1, :]\n",
    "        positionsMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "    \n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && birthTimesMatrix[i - movement, 1] < birthTimesMatrix[i-movement - 1, 1]\n",
    "        shadowCache = birthTimesMatrix[i - movement, :]\n",
    "        birthTimesMatrix[i - movement, :] = birthTimesMatrix[i - movement - 1, :]\n",
    "        birthTimesMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "    \n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && typeMatrix[i - movement, 1] < typeMatrix[i-movement - 1, 1]\n",
    "        shadowCache = typeMatrix[i - movement, :]\n",
    "        typeMatrix[i - movement, :] = typeMatrix[i - movement - 1, :]\n",
    "        typeMatrix[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#stores all these data points into a single concise matrix\n",
    "labelsTimesPositionsType = Array{Any, 2}(UndefInitializer(),279,5)\n",
    "\n",
    "labelsTimesPositionsType[:,1:2] = birthTimesMatrix[:,:]\n",
    "labelsTimesPositionsType[:,3:4] = positionsMatrix[:,2:3]\n",
    "labelsTimesPositionsType[:,5] = typeMatrix[:,2]\n",
    "\n",
    "#alter the positions isometrically to look nicer \n",
    "labelsTimesPositionsType[:,3] *= -1\n",
    "labelsTimesPositionsType[:,4] *= -1\n",
    "\n",
    "\n",
    "#sorts new matrix by birth time\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && labelsTimesPositionsType[i - movement, 2] < labelsTimesPositionsType[i-movement - 1, 2]\n",
    "        shadowCache = labelsTimesPositionsType[i - movement, :]\n",
    "        labelsTimesPositionsType[i - movement, :] = labelsTimesPositionsType[i - movement - 1, :]\n",
    "        labelsTimesPositionsType[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#processes connection list and stores that in its own matrix\n",
    "\n",
    "connectionBase = Array{Any, 2}(UndefInitializer(),6417,4)\n",
    "\n",
    "for i in 1:6417\n",
    "    for j in 1:4\n",
    "        connectionBase[i,j] = connectionList[i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "#create adjacency matrix sorted by birth time weighted for a node filtered persistent homology calculation\n",
    "birthSortedAdjacencyMatrix = zeros(279,279)\n",
    "       \n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        birthSortedAdjacencyMatrix[i,j] = 3000\n",
    "    end\n",
    "end\n",
    "\n",
    "for k in 1:6417\n",
    "    for i in 1:279\n",
    "        for j in 1:279\n",
    "            if connectionBase[k,1] == labelsTimesPositionsType[i,1] && connectionBase[k,2] == labelsTimesPositionsType[j,1]\n",
    "                max = maximum([labelsTimesPositionsType[i,2],labelsTimesPositionsType[j,2]])\n",
    "                birthSortedAdjacencyMatrix[i,j] = max\n",
    "            end\n",
    "            if connectionBase[k,1] == labelsTimesPositionsType[j,1] && connectionBase[k,2] == labelsTimesPositionsType[i,1]\n",
    "                max = maximum([labelsTimesPositionsType[i,2],labelsTimesPositionsType[j,2]])\n",
    "                birthSortedAdjacencyMatrix[i,j] = max\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "#add all the mostly trivial self loops, which are neccesary for dimension 0 calculations\n",
    "for i in 1:279\n",
    "    birthSortedAdjacencyMatrix[i,i] = labelsTimesPositionsType[i,2]\n",
    "end       \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#construct an array that is a list of times at which neurons are born, and how many born at that time. also constructs a list providing the number on this list for each neuron\n",
    "birthOrder = Array{Int64, 1}(undef, 279)\n",
    "times = Array{Float64, 2}(undef, 2,2)\n",
    "times[1,:] = [labelsTimesPositionsType[1,2],1]\n",
    "birthOrder[1] = 1\n",
    "times[2,:] = [labelsTimesPositionsType[2,2],1]\n",
    "birthOrder[2] = 2\n",
    "birthOrderCounter = 2\n",
    "\n",
    "for i in 3:279\n",
    "    if labelsTimesPositionsType[i,2] != times[end,1]\n",
    "        times = cat(dims = 1, times[:,:], transpose([labelsTimesPositionsType[i,2],1]))\n",
    "        birthOrderCounter += 1\n",
    "    else\n",
    "        times[end,2] += 1\n",
    "    end\n",
    "    birthOrder[i] = birthOrderCounter\n",
    "end\n",
    "\n",
    "\n",
    "#end step\n",
    "println(\"output: labelsTimesPositionsType, birthSortedAdjacencyMatrix, birthOrder, and times\")\n",
    "\n",
    "#cleanup\n",
    "positionsMatrix = nothing\n",
    "positions = nothing\n",
    "positionsLabels = nothing\n",
    "birthTimes = nothing\n",
    "birthTimesLabels = nothing\n",
    "connectionList = nothing\n",
    "labeledTypeList = nothing\n",
    "birthTimesMatrix = nothing\n",
    "typeMatrix = nothing\n",
    "connectionBase = nothing\n",
    "birthOrderCounter = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: numValues, bettiDeltas, BettiCurves\n"
     ]
    }
   ],
   "source": [
    "#computes the original betti curves in dimensions 0-2\n",
    "eireneMatrix = eirene(birthSortedAdjacencyMatrix, model = \"vr\", minrad = 0.5, maxrad = 3001, maxdim = 2)\n",
    " \n",
    "\n",
    "#determine number of points on x-axis\n",
    "numValues = size(times,1)\n",
    "\n",
    "#make arrays to store betti curves\n",
    "BettiCurves = zeros(numValues,3)\n",
    "barcodes = barcode(eireneMatrix, dim=0)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "\n",
    "#calculate and store dimension 0 betti values\n",
    "for i in 1:numValues\n",
    "    \n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,1] = bettiNumber\n",
    "end\n",
    "\n",
    "\n",
    "#calculate and store dimension 1 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=1)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "\n",
    "for i in 1:numValues\n",
    "        \n",
    "\n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,2] = bettiNumber\n",
    "end\n",
    "\n",
    "#calculate and store dimension 2 betti values (resetting variables first)\n",
    "barcodes = barcode(eireneMatrix, dim=2)\n",
    "amount = size(barcodes,1)\n",
    "bettiNumber = 0\n",
    "\n",
    "for i in 1:numValues\n",
    "        \n",
    "\n",
    "    for j in 1:amount\n",
    "        if barcodes[j,1] == times[i,1]\n",
    "            bettiNumber += 1\n",
    "        end\n",
    "        if barcodes[j,2] == times[i,1]\n",
    "            bettiNumber -= 1\n",
    "        end\n",
    "    end\n",
    "    BettiCurves[i,3] = bettiNumber\n",
    "end\n",
    "\n",
    "\n",
    "#end step\n",
    "println(\"output: numValues, bettiDeltas, BettiCurves\")\n",
    "\n",
    "#cleanup\n",
    "barcodes = nothing\n",
    "bettiNumber = nothing\n",
    "amount = nothing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2-element Array{Int64,1}:\n",
       " 3900\n",
       " 1060"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#set up variables\n",
    "numNodes = 279\n",
    "H = Array{Int, 1}(undef, 279)\n",
    "fill!(H, 0)\n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        if i != j\n",
    "            if birthSortedAdjacencyMatrix[i,j] < 3000 || birthSortedAdjacencyMatrix[j,i] < 3000\n",
    "                H[i] += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "Hmax = maximum(H)\n",
    "delta = 0.0126\n",
    "timeLength = Array{Float64, 2}(undef, 21, 2)\n",
    "timeLength[1,:] = [0,50]\n",
    "timeLength[2,:] = [270,50]\n",
    "timeLength[3,:] = [360,60]\n",
    "timeLength[4,:] = [390,75]\n",
    "timeLength[5,:] = [420,100]\n",
    "timeLength[6,:] = [450,125]\n",
    "timeLength[7,:] = [480,150]\n",
    "timeLength[8,:] = [510,175]\n",
    "timeLength[9,:] = [540,190]\n",
    "timeLength[10,:] = [570,190]\n",
    "timeLength[11,:] = [600,192.5]\n",
    "timeLength[12,:] = [630,195]\n",
    "timeLength[13,:] = [660,190]\n",
    "timeLength[14,:] = [690,192.5]\n",
    "timeLength[15,:] = [720,200]\n",
    "\n",
    "timeLength[16,:] = [1200,250]\n",
    "timeLength[17,:] = [1560,370]\n",
    "timeLength[18,:] = [2070,450]\n",
    "timeLength[19,:] = [2610,640]\n",
    "timeLength[20,:] = [3360,850]\n",
    "timeLength[21,:] = [3900,1060]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"600\" height=\"400\" viewBox=\"0 0 2400 1600\">\n",
       "<defs>\n",
       "  <clipPath id=\"clip6000\">\n",
       "    <rect x=\"0\" y=\"0\" width=\"2400\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip6000)\" d=\"\n",
       "M0 1600 L2400 1600 L2400 0 L0 0  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip6001\">\n",
       "    <rect x=\"480\" y=\"0\" width=\"1681\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip6000)\" d=\"\n",
       "M194.041 1487.47 L2352.76 1487.47 L2352.76 47.2441 L194.041 47.2441  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip6002\">\n",
       "    <rect x=\"194\" y=\"47\" width=\"2160\" height=\"1441\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  255.137,1487.47 255.137,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  777.323,1487.47 777.323,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1299.51,1487.47 1299.51,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1821.69,1487.47 1821.69,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  2343.88,1487.47 2343.88,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  194.041,1244.93 2352.76,1244.93 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  194.041,975.874 2352.76,975.874 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  194.041,706.823 2352.76,706.823 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  194.041,437.772 2352.76,437.772 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  194.041,168.721 2352.76,168.721 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  194.041,1487.47 2352.76,1487.47 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  194.041,1487.47 194.041,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.137,1487.47 255.137,1465.87 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  777.323,1487.47 777.323,1465.87 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1299.51,1487.47 1299.51,1465.87 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1821.69,1487.47 1821.69,1465.87 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2343.88,1487.47 2343.88,1465.87 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  194.041,1244.93 226.422,1244.93 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  194.041,975.874 226.422,975.874 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  194.041,706.823 226.422,706.823 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  194.041,437.772 226.422,437.772 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  194.041,168.721 226.422,168.721 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 255.137, 1541.47)\" x=\"255.137\" y=\"1541.47\">0</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 777.323, 1541.47)\" x=\"777.323\" y=\"1541.47\">1000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1299.51, 1541.47)\" x=\"1299.51\" y=\"1541.47\">2000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1821.69, 1541.47)\" x=\"1821.69\" y=\"1541.47\">3000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 2343.88, 1541.47)\" x=\"2343.88\" y=\"1541.47\">4000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 170.041, 1262.43)\" x=\"170.041\" y=\"1262.43\">200</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 170.041, 993.374)\" x=\"170.041\" y=\"993.374\">400</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 170.041, 724.323)\" x=\"170.041\" y=\"724.323\">600</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 170.041, 455.272)\" x=\"170.041\" y=\"455.272\">800</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 170.041, 186.221)\" x=\"170.041\" y=\"186.221\">1000</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip6002)\" style=\"stroke:#009af9; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.137,1446.71 396.127,1446.71 443.124,1433.26 458.789,1413.08 474.455,1379.45 490.121,1345.82 505.786,1312.19 521.452,1278.56 537.117,1258.38 552.783,1258.38 \n",
       "  568.448,1255.01 584.114,1251.65 599.78,1258.38 615.445,1255.01 631.111,1244.93 881.76,1177.66 1069.75,1016.23 1336.06,908.611 1618.04,653.013 2009.68,370.509 \n",
       "  2291.66,88.0053 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip6000)\" d=\"\n",
       "M1989.93 251.724 L2280.76 251.724 L2280.76 130.764 L1989.93 130.764  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1989.93,251.724 2280.76,251.724 2280.76,130.764 1989.93,130.764 1989.93,251.724 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip6000)\" style=\"stroke:#009af9; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2013.93,191.244 2157.93,191.244 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip6000)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 2181.93, 208.744)\" x=\"2181.93\" y=\"208.744\">y1</text>\n",
       "</g>\n",
       "</svg>\n"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot(timeLength[:,1], timeLength[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scaledDistance (generic function with 1 method)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make a function to compute distance between a pair of nodes \n",
    "function scaledDistance(second, first)\n",
    "    \n",
    "    #find the adult distance\n",
    "    a = labelsTimesPositionsType[first,3]\n",
    "    b = labelsTimesPositionsType[first,4]\n",
    "    c = labelsTimesPositionsType[second,3]\n",
    "    d = labelsTimesPositionsType[second,4]\n",
    "    adultDistance = evaluate(Euclidean(),[a,b], [c,d])\n",
    "    \n",
    "    \n",
    "    #find the scaling factor\n",
    "    position = 21\n",
    "    currentTime = labelsTimesPositionsType[second, 2]\n",
    "    while currentTime < timeLength[position,1]\n",
    "        position -= 1\n",
    "    end\n",
    "    \n",
    "    scalingNumerator = (((timeLength[position+1,2] - timeLength[position,2])/(timeLength[position+1,1] - timeLength[position,1])) * (currentTime - timeLength[position,1])) + timeLength[position,2]\n",
    "    scaleFactor = scalingNumerator / 3900\n",
    "    \n",
    "    return scaleFactor * adultDistance\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make numTrials copies of ESTG\n",
    "numTrials = 1000\n",
    "\n",
    "trials = zeros(numNodes, numNodes, numTrials)\n",
    "fill!(trials, 3000)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    \n",
    "    for i in 1:numNodes\n",
    "        trials[i,i,k] = labelsTimesPositionsType[i, 2]\n",
    "    end\n",
    "    \n",
    "    for i in 2:numNodes\n",
    "        for j in 1:i-1\n",
    "            \n",
    "            prob = (H[j]/Hmax) * ((MathConstants.e)^ (-1 * (scaledDistance(i, j)/delta)))\n",
    "            if rand() < prob\n",
    "                trials[i,j,k] = labelsTimesPositionsType[i, 2]\n",
    "                trials[j,i,k] = labelsTimesPositionsType[i, 2]\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: ESTGBettiCurves, edgeDensities\n"
     ]
    }
   ],
   "source": [
    "begin\n",
    "#perform betti computations\n",
    "trialCurves = pmap(i -> computeBarcodesNullModels(trials[:,:,i], times), 1:numTrials)\n",
    "\n",
    "        \n",
    "end\n",
    "\n",
    "quant = size(times, 1)\n",
    "\n",
    "#organize this into a nicer format\n",
    "ESTGBettiCurves = zeros(quant,3,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    ESTGBettiCurves[:,:,k] = trialCurves[k]\n",
    "end\n",
    "\n",
    "\n",
    "edgeDensities = zeros(quant, numTrials)\n",
    "for k in 1:numTrials\n",
    "    numberEdges = 0\n",
    "    \n",
    "    for i in 2:quant\n",
    "        for j in 1:i-1\n",
    "            if trials[i,j,k] < 3000\n",
    "                numberEdges += 1\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        edgeDensities[i,k] += (numberEdges/binomial(i,2))\n",
    "    end\n",
    "end \n",
    "            \n",
    "    \n",
    "println(\"output: ESTGBettiCurves, edgeDensities\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output: bettiMeans, bettiStd, averageTimes, edgeDensityMeans"
     ]
    }
   ],
   "source": [
    "#computes average and standard deviation of 1000 copies\n",
    "bettiMeans = zeros(quant,3)\n",
    "bettiStd = zeros(quant,3)\n",
    "averageTimes = zeros(quant)\n",
    "edgeDensityMeans = zeros(quant,2)\n",
    "\n",
    "for i in 1:quant\n",
    "    bettiMeans[i,1] = mean(ESTGBettiCurves[i,1,:])\n",
    "    bettiMeans[i,2] = mean(ESTGBettiCurves[i,2,:])\n",
    "    bettiMeans[i,3] = mean(ESTGBettiCurves[i,3,:])\n",
    "    \n",
    "    bettiStd[i,1] = std(ESTGBettiCurves[i,1,:])\n",
    "    bettiStd[i,2] = std(ESTGBettiCurves[i,2,:])\n",
    "    bettiStd[i,3] = std(ESTGBettiCurves[i,3,:])\n",
    "        \n",
    "    edgeDensityMeans[i,1] = mean(edgeDensities[i,:])\n",
    "    edgeDensityMeans[i,2] = std(edgeDensities[i,:])\n",
    "end\n",
    "\n",
    "print(\"output: bettiMeans, bettiStd, averageTimes, edgeDensityMeans\")\n",
    "\n",
    "#cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine edge density with time\n",
    "edgeDensity = zeros(279)\n",
    "numberEdges = 0\n",
    "\n",
    "for i in 1:279\n",
    "    for j in 1:i-1\n",
    "        if birthSortedAdjacencyMatrix[i,j] < 3000\n",
    "            numberEdges += 1\n",
    "        end\n",
    "    end\n",
    "    edgeDensity[i] = numberEdges/binomial(i,2)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots the average of the random curves, which a strip covering +/- 1 std\n",
    "\n",
    "\n",
    "p1 = plot(times[:,1], bettiMeans[:,2:3], ribbon = bettiStd[:,2:3], color = [:red :blue theme(:default)], legend=:bottomright , title = \"ESTG\", label = [\"dim 1\", \"dim 2\"],  xlabel = \"time\", ylabel = \"betti number\")\n",
    "plot!(times[:,1], BettiCurves[:,2:3], line=(:dot,2), color = [:red :blue theme(:default)], label = [ \"original dim 1\" \"original dim 2\"])\n",
    "p2 = plot(times[:,1], bettiMeans[:,1], ribbon = bettiStd[:,1], color = [:green theme(:default)], legend=:bottomright , title = \"ESTG\", label = [\"dim 0\"],  xlabel = \"time\", ylabel = \"betti number\", ylims = (0,10))\n",
    "plot!(times[:,1], BettiCurves[:,1], line=(:dot,2), label = \"original dim 0\", color = [:green theme(:default)])\n",
    "#p3 = plot(times[:,1], edgeDensityMeans[:,1], color = [:gray], ribbon = edgeDensityMeans[:,2], line=(:dot,2),  ylabel = \"edge density\")\n",
    "#plot!(labelsTimesPositionsType[:,1], edgeDensity[:], lc = [:grey], line=(:dot,2))\n",
    "\n",
    "\n",
    "plot(p1,p2,layout = (2,1), size = (800,800))\n",
    "\n",
    "savefig(\"ESTG.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#beginning of topological similarity analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes zeroth dimensional topological similarity\n",
    "@everywhere function zeroDimTopSim(matrix)\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(matrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 0)\n",
    "\n",
    "zeroDimBarcodes = barcode(eireneGraph, dim = 0)\n",
    "originalBarcodes = zeros(size(zeroDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(zeroDimBarcodes)[1] > 0\n",
    "    for i in 1:size(zeroDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 0\n",
    "        originalBarcodes[i,2] = zeroDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = zeroDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "zeroDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderability(matrix[:,:], x[1], x[2], originalBarcodes, labelsTimesPositionsType[:,2], 0)), dimension = 0)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        zeroDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "zeroDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            zeroDimemsionalTopologicalSimilarity[i,j] = 1 - (zeroDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        else\n",
    "            zeroDimemsionalTopologicalSimilarity[i,j] = 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "return zeroDimemsionalTopologicalSimilarity\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "begin\n",
    "#perform betti computations\n",
    "trialSims = pmap(i -> zeroDimTopSim(trials[:,:,i]), 1:numTrials)      \n",
    "end\n",
    "\n",
    "print(\"finished hard part\")\n",
    "\n",
    "#organize this into a nicer format\n",
    "ESTGtopSimDimZeroTrials = zeros(279,279,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    ESTGtopSimDimZeroTrials[:,:,k] = trialSims[k]\n",
    "end\n",
    " \n",
    "#then take the average\n",
    "ESTGtopSimDimZero = zeros(279,279)\n",
    "\n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        ESTGtopSimDimZero[i,j] = mean(ESTGtopSimDimZeroTrials[i,j,:])\n",
    "    end\n",
    "end\n",
    "    \n",
    "println(\"output: ESTGtopSimDimZero\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes first dimensional topological similarity\n",
    "@everywhere function oneDimTopSim(matrix)\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(matrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 1)\n",
    "\n",
    "oneDimBarcodes = barcode(eireneGraph, dim = 1)\n",
    "originalBarcodes = zeros(size(oneDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(oneDimBarcodes)[1] > 0\n",
    "    for i in 1:size(oneDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 1\n",
    "        originalBarcodes[i,2] = oneDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = oneDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "oneDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderability(matrix[:,:], x[1], x[2], originalBarcodes, labelsTimesPositionsType[:,2],1)), dimension = 1)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        oneDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "oneDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            oneDimemsionalTopologicalSimilarity[i,j] = 1 - (oneDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "return oneDimemsionalTopologicalSimilarity\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "begin\n",
    "#perform betti computations\n",
    "trialSims = pmap(i -> oneDimTopSim(trials[:,:,i]), 1:numTrials)      \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "ESTGtopSimDimOneTrials = zeros(279,279,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    ESTGtopSimDimOneTrials[:,:,k] = trialSims[k]\n",
    "end\n",
    " \n",
    "#then take the average\n",
    "ESTGtopSimDimOne = zeros(279,279)\n",
    "\n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        ESTGtopSimDimOne[i,j] = mean(ESTGtopSimDimOneTrials[i,j,:])\n",
    "    end\n",
    "end\n",
    "    \n",
    "println(\"output: ESTGtopSimDimOne\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes second dimensional topological similarity\n",
    "@everywhere function twoDimTopSim(matrix)\n",
    "\n",
    "#hold barcodes for topological similarity calculation  \n",
    "eireneGraph = eirene(matrix[:,:], minrad = 0.5, maxrad = 3001, maxdim = 2)\n",
    "\n",
    "twoDimBarcodes = barcode(eireneGraph, dim = 2)\n",
    "originalBarcodes = zeros(size(twoDimBarcodes)[1],3)\n",
    "\n",
    "\n",
    "#stores the information for the non-reordered matrix\n",
    "if size(twoDimBarcodes)[1] > 0\n",
    "    for i in 1:size(twoDimBarcodes)[1]\n",
    "        originalBarcodes[i,1] = 2\n",
    "        originalBarcodes[i,2] = twoDimBarcodes[i,1]\n",
    "        originalBarcodes[i,3] = twoDimBarcodes[i,2]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#construct indicies for upcoming map\n",
    "indicies = Array{Any, 1}(undef, 38781)\n",
    "counter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        counter += 1\n",
    "        indicies[counter] = [i,j]\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "#compute topological similarity for this graph\n",
    "twoDimensionalBND = zeros(279,279)\n",
    "begin\n",
    "storage = pmap(x -> rcopy(R\"bottleneck($originalBarcodes, $(reorderability(matrix[:,:], x[1], x[2], originalBarcodes, labelsTimesPositionsType[:,2],2)), dimension = 2)\") , indicies)\n",
    "end\n",
    "\n",
    "storageCounter = 0\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        storageCounter += 1\n",
    "        twoDimensionalBND[i,j] = storage[storageCounter]\n",
    "    end\n",
    "end\n",
    "\n",
    "twoDimemsionalTopologicalSimilarity = zeros(279,279)\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if labelsTimesPositionsType[i,2] != labelsTimesPositionsType[j,2]\n",
    "            twoDimemsionalTopologicalSimilarity[i,j] = 1 - (twoDimensionalBND[i,j] / (labelsTimesPositionsType[i,2] - labelsTimesPositionsType[j,2]))\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "return twoDimemsionalTopologicalSimilarity\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "begin\n",
    "#perform betti computations\n",
    "trialSims = pmap(i -> twoDimTopSim(trials[:,:,i]), 1:numTrials)      \n",
    "end\n",
    "    \n",
    "#organize this into a nicer format\n",
    "ESTGtopSimDimTwoTrials = zeros(279,279,numTrials)\n",
    "\n",
    "for k in 1:numTrials\n",
    "    ESTGtopSimDimTwoTrials[:,:,k] = trialSims[k]\n",
    "end\n",
    " \n",
    "#then take the average\n",
    "ESTGtopSimDimTwo = zeros(279,279)\n",
    "\n",
    "for i in 1:279\n",
    "    for j in 1:279\n",
    "        ESTGtopSimDimTwo[i,j] = mean(ESTGtopSimDimTwoTrials[i,j,:])\n",
    "    end\n",
    "end\n",
    "    \n",
    "println(\"output: ESTGtopSimDimTwo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute average similarity across dimensions 1-3\n",
    "averageTopologicalSimilarity = zeroDimemsionalTopologicalSimilarity[:,:]/3 + oneDimemsionalTopologicalSimilarity[:,:]/3 + twoDimemsionalTopologicalSimilarity[:,:]/3\n",
    "averageBND = zeroDimensionalBND[:,:]/3 + oneDimensionalBND[:,:]/3 + twoDimensionalBND[:,:]/3\n",
    "\n",
    "println(\"output: averageTopologicalSimilarity, averageBND\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(zeroDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(zeroDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Zero Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionZeroSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(oneDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension One Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(oneDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension One Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionOneSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(twoDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(twoDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"dimensionTwoSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "p1 = heatmap(twoDimemsionalTopologicalSimilarity, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Topological Similarity\", color = :tempo)\n",
    "p2 = heatmap(twoDimensionalBND, yflip = true, aspect_ratio=:equal, title = \"Dimension Two Bottleneck Distance\", color = :grays)\n",
    "plot(p1, p2, layout = (1,2))\n",
    "#savefig(\"averageSimilarity.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test on reorderability between types\n",
    "\n",
    "#variables for connections within a type\n",
    "intraSensorySimilarities = zeros(3570) #85 sensory neurons\n",
    "intraMotorSimilarities = zeros(6555) #115 motor neurons\n",
    "intraInterSimilarities = zeros(3081) # 79 inter neurons\n",
    "intraSensoryCounter = 1\n",
    "intraMotorCounter = 1\n",
    "intraInterCounter = 1\n",
    "\n",
    "\n",
    "#variables for connections between different types and a fixed type\n",
    "interSensorySimilarities = zeros(16490) \n",
    "interMotorSimilarities = zeros(18860) \n",
    "interInterSimilarities = zeros(15800)\n",
    "interSensoryCounter = 1\n",
    "interMotorCounter = 1\n",
    "interInterCounter = 1\n",
    "\n",
    "#variables for general within types connections\n",
    "intraTypeSimilariteis = zeros(13206)\n",
    "intraCounter = 1\n",
    "\n",
    "#variabels for general between types connections\n",
    "interTypeSimilariteis = zeros(25575)\n",
    "interCounter = 1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#loads all the various arrays\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        #see if the connection is intra-type\n",
    "        if labelsTimesPositionsType[i,5] == labelsTimesPositionsType[j,5]\n",
    "            \n",
    "            #determine which type the connection is, add the similarity to the list, and up the ticker\n",
    "            if labelsTimesPositionsType[i,5] == \"sensory\"\n",
    "                intraSensorySimilarities[intraSensoryCounter] = topologicalSimilarity[i,j]\n",
    "                intraSensoryCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"motor\"\n",
    "                intraMotorSimilarities[intraMotorCounter] = topologicalSimilarity[i,j]\n",
    "                intraMotorCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"inter\"\n",
    "                intraInterSimilarities[intraInterCounter] = topologicalSimilarity[i,j]\n",
    "                intraInterCounter += 1\n",
    "            end\n",
    "            \n",
    "            intraTypeSimilariteis[intraCounter]  = topologicalSimilarity[i,j]\n",
    "            intraCounter += 1\n",
    "            \n",
    "        end\n",
    "        \n",
    "        #otherwise the connection is a general InterType\n",
    "        if labelsTimesPositionsType[i,5] != labelsTimesPositionsType[j,5]\n",
    "            \n",
    "            #determine which types the connection is between, add the similarity to the lists, and up the tickers\n",
    "            if labelsTimesPositionsType[i,5] == \"sensory\" || labelsTimesPositionsType[j,5] == \"sensory\"\n",
    "                interSensorySimilarities[interSensoryCounter] = topologicalSimilarity[i,j]\n",
    "                interSensoryCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"motor\" || labelsTimesPositionsType[j,5] == \"motor\"\n",
    "                interMotorSimilarities[interMotorCounter] = topologicalSimilarity[i,j]\n",
    "                interMotorCounter += 1\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,5] == \"inter\" || labelsTimesPositionsType[j,5] == \"inter\"\n",
    "                interInterSimilarities[interInterCounter] = topologicalSimilarity[i,j]\n",
    "                interInterCounter += 1\n",
    "            end\n",
    "                      \n",
    "            interTypeSimilariteis[interCounter] = topologicalSimilarity[i,j]\n",
    "            interCounter += 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"many outputs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter sensory pair\n",
    "MannWhitneyUTest(intraSensorySimilarities, interSensorySimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter motor pair\n",
    "MannWhitneyUTest(intraMotorSimilarities, interMotorSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra vs inter inter pair\n",
    "MannWhitneyUTest(intraInterSimilarities, interInterSimilarities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for each intra-inter pair\n",
    "MannWhitneyUTest(intraTypeSimilariteis, interTypeSimilariteis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotArray = Array{Any,1}(UndefInitializer(), 8)\n",
    "plotArray[1] = interMotorSimilarities[:]\n",
    "plotArray[2] = intraMotorSimilarities[:]\n",
    "plotArray[3] = interSensorySimilarities[:]\n",
    "plotArray[4] = intraSensorySimilarities[:]\n",
    "plotArray[5] = interInterSimilarities[:]\n",
    "plotArray[6] = intraInterSimilarities[:]\n",
    "plotArray[7] = interTypeSimilariteis[:]\n",
    "plotArray[8] = intraTypeSimilariteis[:]\n",
    "\n",
    "violin([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"inter-Type\"],plotArray,leg=false, color = theme(:dark))\n",
    "boxplot!([\"inter-Motor\" \"intra-Motor\" \"inter-Sensory\" \"intra-Sensory\" \"inter-Inter\" \"intra-Inter\" \"inter-Type\" \"inter-Type\"],plotArray,leg=false, color = theme(:dark))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#store an array of all 92 indices of L-R pairs\n",
    "LRindices = zeros(92,2)\n",
    "counter = 0\n",
    "\n",
    "for i in 1:279\n",
    "    if (labelsTimesPositionsType[i,1])[end] == 'L' && labelsTimesPositionsType[i,1] != \"AVL\"\n",
    "        counter += 1\n",
    "        LRindices[counter, 1] = i\n",
    "        \n",
    "        rIndex = 0\n",
    "        while LRindices[counter,2] == 0\n",
    "            rIndex += 1\n",
    "            \n",
    "            if (labelsTimesPositionsType[rIndex, 1])[end] == 'R' && (labelsTimesPositionsType[rIndex, 1])[1:end-1] == (labelsTimesPositionsType[i,1])[1:end-1]\n",
    "                LRindices[counter,2] = rIndex\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"output: LRindices\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#store the topological similarities between LR pairs and otherwise\n",
    "topSimsBetweenLR = zeros(92)\n",
    "topSimsOutsideLR = zeros(38689) #279C2 - 92\n",
    "intraCounter = 0\n",
    "interCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        paired = false\n",
    "        \n",
    "        #determines if i-j is a L-R pair\n",
    "        for k in 1:92\n",
    "            if (LRindices[k,1] == i && LRindices[k,2] == j) || (LRindices[k,2] == i && LRindices[k,1] == j)\n",
    "                paired = true\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        #stores the similarity between i and j in the corresponding list\n",
    "        if paired\n",
    "            intraCounter += 1\n",
    "            topSimsBetweenLR[intraCounter] = topologicalSimilarity[i,j]\n",
    "        end\n",
    "        \n",
    "        if !paired\n",
    "            interCounter += 1\n",
    "            topSimsOutsideLR[interCounter] = topologicalSimilarity[i,j]\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"output: topSimsBetweenLR and topSimsOutsideLR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity for L-R and non L-R pairs\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 2)\n",
    "plotArray[1] = topSimsBetweenLR[:]\n",
    "plotArray[2] = topSimsOutsideLR[:]\n",
    "\n",
    "\n",
    "\n",
    "violin([\"L-R Pair\" \"Non L-R Pair\" ],plotArray,leg=false, color = theme(:dark), title = \"topological similarity\")\n",
    "boxplot!([\"L-R Pair\" \"Non L-R Pair\"],plotArray,leg=false, color = theme(:dark), title = \"topological similarity\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for difference within/out LR pairs\n",
    "MannWhitneyUTest(topSimsBetweenLR, topSimsOutsideLR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#store array of LR similarity that only takes in head, body, tail, pre, and post\n",
    "withinHead = zeros(74)\n",
    "withinBody = zeros(7)\n",
    "withinTail = zeros(11)\n",
    "withinPre = zeros(82)\n",
    "withinPost = zeros(10)\n",
    "\n",
    "withoutHead = zeros(16514)\n",
    "withoutBody = zeros(17590)\n",
    "withoutTail = zeros(4585)\n",
    "withoutPre = zeros(20018)\n",
    "withoutPost = zeros(18671)\n",
    "\n",
    "\n",
    "withinHeadCounter = 0\n",
    "withinBodyCounter = 0\n",
    "withinTailCounter = 0\n",
    "withinPreCounter = 0\n",
    "withinPostCounter = 0\n",
    "\n",
    "withoutHeadCounter = 0\n",
    "withoutBodyCounter = 0\n",
    "withoutTailCounter = 0\n",
    "withoutPreCounter = 0\n",
    "withoutPostCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        \n",
    "        #determine if i-j is a L-R pair\n",
    "        paired = false\n",
    "        for k in 1:92\n",
    "            if (LRindices[k,1] == i && LRindices[k,2] == j) || (LRindices[k,2] == i && LRindices[k,1] == j)\n",
    "                paired = true\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        if paired\n",
    "            #perform the pre-post storage\n",
    "            if i < 202\n",
    "                withinPreCounter += 1\n",
    "                withinPre[withinPreCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "            if i > 201\n",
    "                 withinPostCounter += 1\n",
    "                 withinPost[withinPostCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "            \n",
    "            #perform region storage\n",
    "            if labelsTimesPositionsType[i,3] < 0\n",
    "                withinHeadCounter += 1\n",
    "                withinHead[withinHeadCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,3] > 0 && labelsTimesPositionsType[i,3] < 0.935\n",
    "                withinBodyCounter += 1\n",
    "                withinBody[withinBodyCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,3] > 0.935\n",
    "                withinTailCounter += 1\n",
    "                withinTail[withinTailCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        if !paired\n",
    "            #perform the pre-post storage\n",
    "            if i < 202\n",
    "                withoutPreCounter += 1\n",
    "                withoutPre[withoutPreCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "            if i > 201\n",
    "                 withoutPostCounter += 1\n",
    "                 withoutPost[withoutPostCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "            \n",
    "            #perform region storage\n",
    "            if labelsTimesPositionsType[i,3] < 0\n",
    "                withoutHeadCounter += 1\n",
    "                withoutHead[withoutHeadCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,3] > 0 && labelsTimesPositionsType[i,3] < 0.935\n",
    "                withoutBodyCounter += 1\n",
    "                withoutBody[withoutBodyCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "            \n",
    "            if labelsTimesPositionsType[i,3] > 0.935\n",
    "                withoutTailCounter += 1\n",
    "                withoutTail[withoutTailCounter] = topologicalSimilarity[i,j]\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"output: many\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots L-R similarity parsed by location\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 6)\n",
    "plotArray[1] = withinHead[:]\n",
    "plotArray[2] = withoutHead[:]\n",
    "plotArray[3] = withinBody[:]\n",
    "plotArray[4] = withoutBody[:]\n",
    "plotArray[5] = withinTail[:]\n",
    "plotArray[6] = withoutTail[:]\n",
    "\n",
    "\n",
    "violin([\"Head pair\" \"Head non-pair\" \"Body pair\" \"Body non-pair\" \"Tail pair\" \"Tail non-pair\"],plotArray,leg=false, color = theme(:dark), title = \"LR similarity by location\")\n",
    "boxplot!([\"Head pair\" \"Head non-pair\" \"Body pair\" \"Body non-pair\" \"Tail pair\" \"Tail non-pair\"],plotArray,leg=false, color = theme(:dark), title = \"LR similarity by location\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots L-R similarity parsed by time\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 4)\n",
    "plotArray[1] = withinPre[:]\n",
    "plotArray[2] = withoutPre[:]\n",
    "plotArray[3] = withinPost[:]\n",
    "plotArray[4] = withoutPost[:]\n",
    "\n",
    "\n",
    "\n",
    "violin([\"Pre pair\" \"Pre non-pair\" \"Post pair\" \"Post non-pair\"],plotArray,leg=false, color = theme(:dark), title = \"LR similarity by time\")\n",
    "boxplot!([\"Pre pair\" \"Pre non-pair\" \"Post pair\" \"Post non-pair\"],plotArray,leg=false, color = theme(:dark), title = \"LR similarity by time\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute topological overlap of all LR and non-LR pairs\n",
    "topologicalOverlapWithin = zeros(92)\n",
    "topologicalOverlapWithout = zeros(38689)\n",
    "withinCounter = 0\n",
    "withoutCounter = 0\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        #generate and clear variables\n",
    "        iNeighbors = zeros(279)\n",
    "        jNeighbors = zeros(279)\n",
    "        sharedNeighbors = 0\n",
    "        numerator = 0\n",
    "        denominator = 0\n",
    "        topologicalOverlap = 0\n",
    "        \n",
    "        #determine the topological overlap\n",
    "        for k in 1:279\n",
    "            if birthSortedAdjacencyMatrix[i,k] < 300\n",
    "                iNeighbors[k] = 1\n",
    "            end\n",
    "            if birthSortedAdjacencyMatrix[j,k] < 300\n",
    "                jNeighbors[k] = 1\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        for k in 1:279\n",
    "            if iNeighbors[k] == 1 && jNeighbors[k] == 1\n",
    "                sharedNeighbors += 1\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        numerator = sharedNeighbors\n",
    "        if iNeighbors[j] == 1\n",
    "            numerator += 1\n",
    "        end\n",
    "        \n",
    "        denominator = minimum([sum(iNeighbors[:]), sum(jNeighbors[:])]) + 1\n",
    "        if iNeighbors[j] == 1\n",
    "            denominator -= 1\n",
    "        end\n",
    "        \n",
    "        topologicalOverlap = numerator/denominator\n",
    "        \n",
    "        \n",
    "        #now store the topological overlap in the correct bin\n",
    "        paired = false\n",
    "        for k in 1:92\n",
    "            if (LRindices[k,1] == i && LRindices[k,2] == j) || (LRindices[k,2] == i && LRindices[k,1] == j)\n",
    "                paired = true\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        if paired\n",
    "            withinCounter += 1\n",
    "            topologicalOverlapWithin[withinCounter] = topologicalOverlap       \n",
    "        end\n",
    "        \n",
    "        if !paired\n",
    "            withoutCounter += 1\n",
    "            topologicalOverlapWithout[withoutCounter] = topologicalOverlap \n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "println(\"output: topologicalOverlapWithin and topologicalOverlapWithout\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot the topological overlaps\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 2)\n",
    "plotArray[1] = topologicalOverlapWithin[:]\n",
    "plotArray[2] = topologicalOverlapWithout[:]\n",
    "\n",
    "\n",
    "\n",
    "violin([\"between L-R Pair\" \"Non L-R Pair\" ],plotArray,leg=false, color = theme(:dark), title = \"topological overlap\")\n",
    "boxplot!([\"between L-R Pair\" \"Non L-R Pair\"],plotArray,leg=false, color = theme(:dark), title = \"topological overlap\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do test of signifigance for difference within/out LR pairs\n",
    "MannWhitneyUTest(topologicalOverlapWithin, topologicalOverlapWithout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#store matrix of topopogical similarity vs distance\n",
    "#column 2 is distance, column 1 is similarity\n",
    "pairsWithTopSimandDistance = zeros(38781,2)\n",
    "counter = 1\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        pairsWithTopSimandDistance[counter,2] = sqrt( (labelsTimesPositionsType[i,3] - labelsTimesPositionsType[j,3])^2 + (labelsTimesPositionsType[i,4] - labelsTimesPositionsType[j,4])^2 )\n",
    "        pairsWithTopSimandDistance[counter,1] = topologicalSimilarity[i,j]\n",
    "        counter += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "#sorts the matrix by distance\n",
    "for i in 1:38781\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && pairsWithTopSimandDistance[i - movement, 2] < pairsWithTopSimandDistance[i-movement - 1, 2]\n",
    "        shadowCache = pairsWithTopSimandDistance[i - movement, :]\n",
    "        pairsWithTopSimandDistance[i - movement, :] = pairsWithTopSimandDistance[i - movement - 1, :]\n",
    "        pairsWithTopSimandDistance[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "println(\"output: pairsWithTopSimandDistance\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute correlation between distance and topological similarity\n",
    "cor(pairsWithTopSimandDistance[:,1], pairsWithTopSimandDistance[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try to categorize unfriendly nodes by adjacent location on matrix deltas\n",
    "topologicalSimilarityDeltas  = zeros(279,276) #nnodes x ndeltas size\n",
    "reductThisIntoThat = zeros(279,278)\n",
    "\n",
    "#iterates over the nodes\n",
    "for i in 1:279\n",
    "    adjustment = 0\n",
    "    \n",
    "    #iterate over all pairs of nodes including i\n",
    "    for j in 1:278\n",
    "           \n",
    "        #to avoid having to 'switch directions,' take the max of the symmetric pair\n",
    "        first = maximum([topologicalSimilarity[i,j], topologicalSimilarity[j,i]])\n",
    "        second = maximum([topologicalSimilarity[i,j+1], topologicalSimilarity[j+1,i]])\n",
    "        delta = abs(first - second)\n",
    "        reductThisIntoThat[i,j] = delta\n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "#remove deltas in which (i,i) contributed\n",
    "for i in 2:278\n",
    "    counter = 0\n",
    "    \n",
    "    for j in 1:278\n",
    "        if j < i - 1\n",
    "            counter += 1\n",
    "            topologicalSimilarityDeltas[i,counter] = reductThisIntoThat[i,j]\n",
    "        end\n",
    "        \n",
    "        if j > i \n",
    "            counter += 1\n",
    "            topologicalSimilarityDeltas[i,counter] = reductThisIntoThat[i,j]\n",
    "        end\n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "#the above construction is problematic for nodes 1 and 279, so we store them in a separate matrix\n",
    "for k in 1:276\n",
    "    topologicalSimilarityDeltas[1,k] = reductThisIntoThat[1,k+1]\n",
    "    topologicalSimilarityDeltas[279,k] = reductThisIntoThat[279,k]\n",
    "end\n",
    "\n",
    "\n",
    "println(\"output: topologicalSimilarityDeltas and deltasForExtrema\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute variance and mean for each node's set of deltas\n",
    "nodeDeltaMeanAndVaraince = zeros(279,2)\n",
    "\n",
    "for i in 1:279\n",
    "    nodeDeltaMeanAndVaraince[i,1] = mean(topologicalSimilarityDeltas[i,:])\n",
    "    nodeDeltaMeanAndVaraince[i,2] = var(topologicalSimilarityDeltas[i,:])\n",
    "end\n",
    "\n",
    "println(\"output: nodeDeltaMeanAndVaraince\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots delta mean and delta varaince\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 2)\n",
    "plotArray[1] = nodeDeltaMeanAndVaraince[:,1]\n",
    "plotArray[2] = nodeDeltaMeanAndVaraince[:,2]\n",
    "\n",
    "\n",
    "\n",
    "violin([\"Delta Mean\" \"Delta Variance\" ],plotArray,leg=false, color = theme(:dark))\n",
    "boxplot!([\"Delta Mean\" \"Delta Variance\"],plotArray,leg=false, color = theme(:dark))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot distance/similarity pairs\n",
    "scatter(nodeDeltaMeanAndVaraince[:,1], nodeDeltaMeanAndVaraince[:,2], legend = false, title = \"Delta Mean vs Delta Variance\", xlabel = \"delta mean\", ylabel = \"delta variance\", color = :grey)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute correlation between distance and topological similarity\n",
    "cor(nodeDeltaMeanAndVaraince[:,1], nodeDeltaMeanAndVaraince[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newTopSim = topologicalSimilarity[:,:]\n",
    "newDeltaMean = nodeDeltaMeanAndVaraince[:,2]\n",
    "\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && newDeltaMean[i - movement] < newDeltaMean[i-movement - 1]\n",
    "        shadowCache = newDeltaMean[i - movement]\n",
    "        newDeltaMean[i - movement] = newDeltaMean[i - movement - 1]\n",
    "        newDeltaMean[i - movement - 1] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[i - movement, :]\n",
    "        newTopSim[i - movement, :] = newTopSim[i - movement - 1, :]\n",
    "        newTopSim[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[:, i - movement]\n",
    "        newTopSim[:, i - movement] = newTopSim[:, i - movement - 1]\n",
    "        newTopSim[:, i - movement - 1] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if newTopSim[j,i] > 0\n",
    "            newTopSim[i,j] = newTopSim[j,i]\n",
    "            newTopSim[j,i] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "heatmap(newTopSim, yflip = false, aspect_ratio=:equal, title = \"One-Two Dimensional Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newDeltaMean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do summed similarity calculations\n",
    "summedSimilarity = zeros(279)\n",
    "\n",
    "for i in 1:279\n",
    "    summedSimilarity[i] = sum(topologicalSimilarity[i,:]) + sum(topologicalSimilarity[:,i])\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots delta mean and delta varaince\n",
    "plotArray = Array{Any,1}(UndefInitializer(), 1)\n",
    "plotArray[1] = summedSimilarity\n",
    "\n",
    "\n",
    "\n",
    "violin([\"Summed Similarity\" ],plotArray,leg=false, color = theme(:dark))\n",
    "boxplot!([\"Summed Similarity\"],plotArray,leg=false, color = theme(:dark))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newTopSim = topologicalSimilarity[:,:]\n",
    "newDeltaMean = summedSimilarity[:]\n",
    "\n",
    "for i in 1:279\n",
    "    movement = 0\n",
    "    \n",
    "    while i - movement > 1 && newDeltaMean[i - movement] < newDeltaMean[i-movement - 1]\n",
    "        shadowCache = newDeltaMean[i - movement]\n",
    "        newDeltaMean[i - movement] = newDeltaMean[i - movement - 1]\n",
    "        newDeltaMean[i - movement - 1] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[i - movement, :]\n",
    "        newTopSim[i - movement, :] = newTopSim[i - movement - 1, :]\n",
    "        newTopSim[i - movement - 1, :] = shadowCache\n",
    "        \n",
    "        shadowCache = newTopSim[:, i - movement]\n",
    "        newTopSim[:, i - movement] = newTopSim[:, i - movement - 1]\n",
    "        newTopSim[:, i - movement - 1] = shadowCache\n",
    "        \n",
    "        movement += 1\n",
    "    end\n",
    "end\n",
    "\n",
    "for i in 2:279\n",
    "    for j in 1:i-1\n",
    "        if newTopSim[j,i] > 0\n",
    "            newTopSim[i,j] = newTopSim[j,i]\n",
    "            newTopSim[j,i] = 0\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots topological similarity\n",
    "heatmap(newTopSim, yflip = false, aspect_ratio=:equal, title = \"One-Two Dimensional Topological Similarity\", color = :tempo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": "63e046a8-6b48-42e6-9707-c6982ec17318",
   "lastKernelId": "7b603f2d-27b9-4e98-acc2-a6168a34eb0f"
  },
  "kernelspec": {
   "display_name": "Julia 1.0.5",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
